// @deity/falcon-shop-data/dist/Cart/AddToCartMutation.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.useAddToCartMutation = exports.AddToCartMutation = exports.ADD_TO_CART = void 0;

var _inheritsLoose2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/inheritsLoose"));

var _client = /*#__PURE__*/require("@apollo/client");

var _falconData = /*#__PURE__*/require("@deity/falcon-data");

var ADD_TO_CART = {
  "kind": "Document",
  "definitions": [{
    "kind": "OperationDefinition",
    "operation": "mutation",
    "name": {
      "kind": "Name",
      "value": "AddToCart"
    },
    "variableDefinitions": [{
      "kind": "VariableDefinition",
      "variable": {
        "kind": "Variable",
        "name": {
          "kind": "Name",
          "value": "input"
        }
      },
      "type": {
        "kind": "NonNullType",
        "type": {
          "kind": "NamedType",
          "name": {
            "kind": "Name",
            "value": "AddToCartInput"
          }
        }
      },
      "directives": []
    }],
    "directives": [],
    "selectionSet": {
      "kind": "SelectionSet",
      "selections": [{
        "kind": "Field",
        "name": {
          "kind": "Name",
          "value": "addToCart"
        },
        "arguments": [{
          "kind": "Argument",
          "name": {
            "kind": "Name",
            "value": "input"
          },
          "value": {
            "kind": "Variable",
            "name": {
              "kind": "Name",
              "value": "input"
            }
          }
        }],
        "directives": [],
        "selectionSet": {
          "kind": "SelectionSet",
          "selections": [{
            "kind": "Field",
            "name": {
              "kind": "Name",
              "value": "id"
            },
            "arguments": [],
            "directives": []
          }, {
            "kind": "Field",
            "name": {
              "kind": "Name",
              "value": "sku"
            },
            "arguments": [],
            "directives": []
          }, {
            "kind": "Field",
            "name": {
              "kind": "Name",
              "value": "qty"
            },
            "arguments": [],
            "directives": []
          }, {
            "kind": "Field",
            "name": {
              "kind": "Name",
              "value": "name"
            },
            "arguments": [],
            "directives": []
          }, {
            "kind": "Field",
            "name": {
              "kind": "Name",
              "value": "price"
            },
            "arguments": [],
            "directives": []
          }, {
            "kind": "Field",
            "name": {
              "kind": "Name",
              "value": "productType"
            },
            "arguments": [],
            "directives": []
          }]
        }
      }]
    }
  }],
  "loc": {
    "start": 0,
    "end": 160,
    "source": {
      "body": "\n  mutation AddToCart($input: AddToCartInput!) {\n    addToCart(input: $input) {\n      id\n      sku\n      qty\n      name\n      price\n      productType\n    }\n  }\n",
      "name": "GraphQL request",
      "locationOffset": {
        "line": 1,
        "column": 1
      }
    }
  }
};
exports.ADD_TO_CART = ADD_TO_CART;

var AddToCartMutation = /*#__PURE__*/function (_Mutation) {
  (0, _inheritsLoose2.default)(AddToCartMutation, _Mutation);

  function AddToCartMutation() {
    return _Mutation.apply(this, arguments) || this;
  }

  return AddToCartMutation;
}(_falconData.Mutation);

exports.AddToCartMutation = AddToCartMutation;
AddToCartMutation.defaultProps = {
  mutation: ADD_TO_CART,
  refetchQueries: ['MiniCart', 'Cart', 'CheckoutState'],
  awaitRefetchQueries: true
};

var useAddToCartMutation = function useAddToCartMutation(options) {
  if (options === void 0) {
    options = {};
  }

  return (0, _client.useMutation)(ADD_TO_CART, Object.assign({
    refetchQueries: ['MiniCart', 'Cart', 'CheckoutState'],
    awaitRefetchQueries: true
  }, options));
};

exports.useAddToCartMutation = useAddToCartMutation;
