// @deity/falcon-shop-data/dist/Cart/CancelCouponMutation.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.useCancelCouponMutation = exports.CancelCouponMutation = exports.CANCEL_COUPON = void 0;

var _inheritsLoose2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/inheritsLoose"));

var _client = /*#__PURE__*/require("@apollo/client");

var _falconData = /*#__PURE__*/require("@deity/falcon-data");

var CANCEL_COUPON = {
  "kind": "Document",
  "definitions": [{
    "kind": "OperationDefinition",
    "operation": "mutation",
    "name": {
      "kind": "Name",
      "value": "CancelCoupon"
    },
    "variableDefinitions": [{
      "kind": "VariableDefinition",
      "variable": {
        "kind": "Variable",
        "name": {
          "kind": "Name",
          "value": "input"
        }
      },
      "type": {
        "kind": "NonNullType",
        "type": {
          "kind": "NamedType",
          "name": {
            "kind": "Name",
            "value": "CouponInput"
          }
        }
      },
      "directives": []
    }],
    "directives": [],
    "selectionSet": {
      "kind": "SelectionSet",
      "selections": [{
        "kind": "Field",
        "name": {
          "kind": "Name",
          "value": "cancelCoupon"
        },
        "arguments": [{
          "kind": "Argument",
          "name": {
            "kind": "Name",
            "value": "input"
          },
          "value": {
            "kind": "Variable",
            "name": {
              "kind": "Name",
              "value": "input"
            }
          }
        }],
        "directives": []
      }]
    }
  }],
  "loc": {
    "start": 0,
    "end": 85,
    "source": {
      "body": "\n  mutation CancelCoupon($input: CouponInput!) {\n    cancelCoupon(input: $input)\n  }\n",
      "name": "GraphQL request",
      "locationOffset": {
        "line": 1,
        "column": 1
      }
    }
  }
};
exports.CANCEL_COUPON = CANCEL_COUPON;
var defaultOptions = {
  mutation: CANCEL_COUPON,
  refetchQueries: ['Cart', 'CheckoutState'],
  awaitRefetchQueries: true
};

var CancelCouponMutation = /*#__PURE__*/function (_Mutation) {
  (0, _inheritsLoose2.default)(CancelCouponMutation, _Mutation);

  function CancelCouponMutation() {
    return _Mutation.apply(this, arguments) || this;
  }

  return CancelCouponMutation;
}(_falconData.Mutation);

exports.CancelCouponMutation = CancelCouponMutation;
CancelCouponMutation.defaultProps = defaultOptions;

var useCancelCouponMutation = function useCancelCouponMutation(options) {
  if (options === void 0) {
    options = {};
  }

  return (0, _client.useMutation)(defaultOptions.mutation, Object.assign({}, defaultOptions, options));
};

exports.useCancelCouponMutation = useCancelCouponMutation;
