// @deity/falcon-shop-data/dist/Cart/EmptyCartMutation.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.useEmptyCartMutation = exports.EmptyCartMutation = exports.EMPTY_CART = void 0;

var _inheritsLoose2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/inheritsLoose"));

var _client = /*#__PURE__*/require("@apollo/client");

var _falconData = /*#__PURE__*/require("@deity/falcon-data");

var EMPTY_CART = {
  "kind": "Document",
  "definitions": [{
    "kind": "OperationDefinition",
    "operation": "mutation",
    "name": {
      "kind": "Name",
      "value": "EmptyCart"
    },
    "variableDefinitions": [],
    "directives": [],
    "selectionSet": {
      "kind": "SelectionSet",
      "selections": [{
        "kind": "Field",
        "name": {
          "kind": "Name",
          "value": "emptyCart"
        },
        "arguments": [],
        "directives": []
      }]
    }
  }],
  "loc": {
    "start": 0,
    "end": 42,
    "source": {
      "body": "\n  mutation EmptyCart {\n    emptyCart\n  }\n",
      "name": "GraphQL request",
      "locationOffset": {
        "line": 1,
        "column": 1
      }
    }
  }
};
exports.EMPTY_CART = EMPTY_CART;
var defaultOptions = {
  refetchQueries: ['Cart', 'MiniCart', 'CheckoutState'],
  awaitRefetchQueries: false
};

var EmptyCartMutation = /*#__PURE__*/function (_Mutation) {
  (0, _inheritsLoose2.default)(EmptyCartMutation, _Mutation);

  function EmptyCartMutation() {
    return _Mutation.apply(this, arguments) || this;
  }

  return EmptyCartMutation;
}(_falconData.Mutation);

exports.EmptyCartMutation = EmptyCartMutation;
EmptyCartMutation.defaultProps = /*#__PURE__*/Object.assign({
  mutation: EMPTY_CART
}, defaultOptions);

var useEmptyCartMutation = function useEmptyCartMutation(options) {
  if (options === void 0) {
    options = {};
  }

  return (0, _client.useMutation)(EMPTY_CART, Object.assign({}, defaultOptions, options));
};

exports.useEmptyCartMutation = useEmptyCartMutation;
