// @deity/falcon-shop-data/dist/Cart/RemoveCartItemMutation.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.RemoveCartItemMutation = exports.REMOVE_CART_ITEM = void 0;

var _inheritsLoose2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/inheritsLoose"));

var _falconData = /*#__PURE__*/require("@deity/falcon-data");

var _CartQuery = /*#__PURE__*/require("./CartQuery");

var REMOVE_CART_ITEM = {
  "kind": "Document",
  "definitions": [{
    "kind": "OperationDefinition",
    "operation": "mutation",
    "name": {
      "kind": "Name",
      "value": "RemoveCartItem"
    },
    "variableDefinitions": [{
      "kind": "VariableDefinition",
      "variable": {
        "kind": "Variable",
        "name": {
          "kind": "Name",
          "value": "input"
        }
      },
      "type": {
        "kind": "NonNullType",
        "type": {
          "kind": "NamedType",
          "name": {
            "kind": "Name",
            "value": "RemoveCartItemInput"
          }
        }
      },
      "directives": []
    }],
    "directives": [],
    "selectionSet": {
      "kind": "SelectionSet",
      "selections": [{
        "kind": "Field",
        "name": {
          "kind": "Name",
          "value": "removeCartItem"
        },
        "arguments": [{
          "kind": "Argument",
          "name": {
            "kind": "Name",
            "value": "input"
          },
          "value": {
            "kind": "Variable",
            "name": {
              "kind": "Name",
              "value": "input"
            }
          }
        }],
        "directives": [],
        "selectionSet": {
          "kind": "SelectionSet",
          "selections": [{
            "kind": "Field",
            "name": {
              "kind": "Name",
              "value": "id"
            },
            "arguments": [],
            "directives": []
          }]
        }
      }]
    }
  }],
  "loc": {
    "start": 0,
    "end": 114,
    "source": {
      "body": "\n  mutation RemoveCartItem($input: RemoveCartItemInput!) {\n    removeCartItem(input: $input) {\n      id\n    }\n  }\n",
      "name": "GraphQL request",
      "locationOffset": {
        "line": 1,
        "column": 1
      }
    }
  }
};
exports.REMOVE_CART_ITEM = REMOVE_CART_ITEM;

var RemoveCartItemMutation = /*#__PURE__*/function (_Mutation) {
  (0, _inheritsLoose2.default)(RemoveCartItemMutation, _Mutation);

  function RemoveCartItemMutation() {
    return _Mutation.apply(this, arguments) || this;
  }

  return RemoveCartItemMutation;
}(_falconData.Mutation);

exports.RemoveCartItemMutation = RemoveCartItemMutation;
RemoveCartItemMutation.defaultProps = {
  mutation: REMOVE_CART_ITEM,
  refetchQueries: ['Cart', 'MiniCart', 'CheckoutState'],
  awaitRefetchQueries: true,
  update: function update(store, _ref) {
    var removeCartItem = _ref.data.removeCartItem;
    var data = store.readQuery({
      query: _CartQuery.GET_CART
    });
    var updatedCart = {};

    if (removeCartItem) {
      var id = removeCartItem.id;
      updatedCart['items'] = data.cart.items.filter(function (item) {
        return item.id !== id;
      });
      updatedCart['itemsQty'] = data.cart.items.length;
    } else {
      // Something isn't right.. clear the entire cart.
      // `removeCartItem` can be `null` if the user adds something to their cart and then logs out
      updatedCart['items'] = [];
      updatedCart['itemsQty'] = data.cart.items.length;
    }

    store.writeQuery({
      query: _CartQuery.GET_CART,
      data: Object.assign({}, data, {
        cart: Object.assign({}, data.cart, updatedCart)
      })
    });
  }
};
