// @deity/falcon-ui-kit/dist/ShippingMethod/ShippingMethodDetails.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.ShippingMethodDetails = void 0;

var _react = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("react"));

var _propTypes = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("prop-types"));

var _falconUi = /*#__PURE__*/require("@deity/falcon-ui");

var _Price = /*#__PURE__*/require("../Price");

var ShippingMethodDetails = function ShippingMethodDetails(_ref) {
  var carrierTitle = _ref.carrierTitle,
      methodTitle = _ref.methodTitle,
      amount = _ref.amount;
  return /*#__PURE__*/_react.default.createElement(_falconUi.FlexLayout, {
    justifyContent: "space-between"
  }, /*#__PURE__*/_react.default.createElement(_falconUi.FlexLayout, null, /*#__PURE__*/_react.default.createElement(_falconUi.Text, {
    fontWeight: "bold"
  }, "" + carrierTitle), methodTitle && /*#__PURE__*/_react.default.createElement(_falconUi.Text, {
    pl: "xs",
    color: "secondaryText"
  }, "(" + methodTitle + ")")), /*#__PURE__*/_react.default.createElement(_Price.Price, {
    value: amount
  }));
};

exports.ShippingMethodDetails = ShippingMethodDetails;
ShippingMethodDetails.propTypes = {
  carrierTitle: _propTypes.default.string.isRequired,
  methodTitle: _propTypes.default.string,
  amount: _propTypes.default.number.isRequired
};
