// @deity/falcon-ui-kit/dist/Product/ProductTierPrices.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.ProductTierPrices = void 0;

var _objectWithoutPropertiesLoose2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("react"));

var _propTypes = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("prop-types"));

var _falconI18n = /*#__PURE__*/require("@deity/falcon-i18n");

var _falconUi = /*#__PURE__*/require("@deity/falcon-ui");

var _falconFrontKit = /*#__PURE__*/require("@deity/falcon-front-kit");

var ProductTierPrices = function ProductTierPrices(_ref) {
  var tierPrices = _ref.items,
      rest = (0, _objectWithoutPropertiesLoose2.default)(_ref, ["items"]);

  var _useCurrency = (0, _falconFrontKit.useCurrency)(),
      priceFormat = _useCurrency.priceFormat;

  if (!Array.isArray(tierPrices) || !tierPrices.length) {
    return null;
  }

  return /*#__PURE__*/_react.default.createElement(_falconUi.Box, rest, tierPrices.map(function (x) {
    return /*#__PURE__*/_react.default.createElement(_falconUi.Text, {
      key: x.qty
    }, /*#__PURE__*/_react.default.createElement(_falconI18n.T, {
      id: "product.tierPriceDescription",
      qty: x.qty,
      price: priceFormat(x.value),
      discount: x.discount
    }));
  }));
};

exports.ProductTierPrices = ProductTierPrices;
ProductTierPrices.propTypes = {
  // @ts-ignore https://github.com/DefinitelyTyped/DefinitelyTyped/issues/34692
  items: /*#__PURE__*/_propTypes.default.arrayOf( /*#__PURE__*/_propTypes.default.shape({
    qty: _propTypes.default.number.isRequired,
    value: _propTypes.default.number.isRequired,
    discount: _propTypes.default.number.isRequired
  }))
};
