// @deity/falcon-ui-kit/dist/Product/ProductOptionList.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.ProductOptionList = void 0;

var _react = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("react"));

var _falconUi = /*#__PURE__*/require("@deity/falcon-ui");

var _ProductOptionRadioGroupField = /*#__PURE__*/require("./ProductOptionRadioGroupField");

var ProductOptionList = function ProductOptionList(props) {
  return /*#__PURE__*/_react.default.createElement(_falconUi.GridLayout, {
    gridGap: "sm",
    gridArea: props.gridArea
  }, props.items.map(function (_ref) {
    var label = _ref.label,
        id = _ref.id,
        values = _ref.values;
    return /*#__PURE__*/_react.default.createElement(_ProductOptionRadioGroupField.ProductOptionRadioGroupField, {
      key: id,
      label: label,
      name: props.name + "." + id,
      values: values,
      disabled: props.disabled
    });
  }));
};

exports.ProductOptionList = ProductOptionList;
ProductOptionList.defaultProps = {
  items: []
};
