// @deity/falcon-ui-kit/dist/Product/ProductConfigurator.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.ProductConfigurator = void 0;

var _inheritsLoose2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/inheritsLoose"));

var _react = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("react"));

var _falconFrontKit = /*#__PURE__*/require("@deity/falcon-front-kit");

/**
 * ProductConfigurator takes care of handling data relaed to product options available to be selected before adding to cart.
 * Currently only configurable options are supported, in the future that class will handle bundled products as well as custom product attributes.
 */
var ProductConfigurator = /*#__PURE__*/function (_React$Component) {
  (0, _inheritsLoose2.default)(ProductConfigurator, _React$Component);

  function ProductConfigurator() {
    var _this;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _React$Component.call.apply(_React$Component, [this].concat(args)) || this;
    _this.state = {
      selectedConfigurableOptions: {}
    };

    _this.handleProductConfigurationChange = function (type, ev) {
      var _ev$target = ev.target,
          name = _ev$target.name,
          value = _ev$target.value;

      if (type === 'configurableOption') {
        _this.handleConfigurationOptionChange(name, value);
      }
    };

    _this.isValueSelected = function (type, name, value) {
      if (type === 'configurableOption') {
        return _this.state.selectedConfigurableOptions.get(name) === value;
      }

      return false;
    };

    return _this;
  }

  var _proto = ProductConfigurator.prototype;

  /**
   * Handles change of configurable product option
   * @param {string} name name of the option
   * @param {any} value value of changed option
   */
  _proto.handleConfigurationOptionChange = function handleConfigurationOptionChange(name, value) {
    var _this2 = this;

    this.setState(function (state) {
      var _Object$assign;

      return {
        selectedConfigurableOptions: Object.assign({}, state.selectedConfigurableOptions, (_Object$assign = {}, _Object$assign[name] = value, _Object$assign))
      };
    }, function () {
      // when state is set then update form manager
      _this2.props.onChange('options', (0, _falconFrontKit.formProductOptionsToInput)(_this2.state.selectedConfigurableOptions));
    });
  };

  _proto.render = function render() {
    return this.props.children({
      handleProductConfigurationChange: this.handleProductConfigurationChange,
      isValueSelected: this.isValueSelected
    });
  };

  return ProductConfigurator;
}(_react.default.Component);

exports.ProductConfigurator = ProductConfigurator;
