// @deity/falcon-ui-kit/dist/Product/ProductCard.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.ProductCard = void 0;

var _extends2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/extends"));

var _react = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("react"));

var _propTypes = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("prop-types"));

var _falconUi = /*#__PURE__*/require("@deity/falcon-ui");

var _Image = /*#__PURE__*/require("../Image");

var _ProductPrice = /*#__PURE__*/require("./ProductPrice");

var _ProductCardLayout = /*#__PURE__*/require("./ProductCardLayout");

var ProductCard = function ProductCard(_ref) {
  var name = _ref.name,
      thumbnail = _ref.thumbnail,
      urlPath = _ref.urlPath,
      price = _ref.price;
  return /*#__PURE__*/_react.default.createElement(_ProductCardLayout.ProductCardLayout, {
    to: urlPath
  }, thumbnail && /*#__PURE__*/_react.default.createElement(_Image.Image, {
    css: {
      height: '100%',
      width: '100%',
      flex: '1 1 100%',
      minHeight: '0%'
    },
    src: thumbnail,
    alt: name,
    loading: "lazy"
  }), /*#__PURE__*/_react.default.createElement(_falconUi.Text, {
    py: "xs",
    ellipsis: true
  }, name), /*#__PURE__*/_react.default.createElement(_falconUi.FlexLayout, null, /*#__PURE__*/_react.default.createElement(_ProductPrice.ProductPrice, (0, _extends2.default)({}, price, {
    fontSize: "md"
  }))));
};

exports.ProductCard = ProductCard;
ProductCard.propTypes = {
  name: _propTypes.default.string.isRequired,
  urlPath: _propTypes.default.string.isRequired,
  // @ts-ignore https://github.com/DefinitelyTyped/DefinitelyTyped/issues/34692
  price: /*#__PURE__*/_propTypes.default.shape({
    regular: _propTypes.default.number.isRequired,
    special: _propTypes.default.number,
    min: _propTypes.default.number
  }).isRequired,
  thumbnail: _propTypes.default.string
};
