// @deity/falcon-ui-kit/dist/Forms/CheckboxFormField.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.CheckboxFormField = exports.CheckboxFormFieldLayout = void 0;

var _extends2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("react"));

var _falconFrontKit = /*#__PURE__*/require("@deity/falcon-front-kit");

var _falconUi = /*#__PURE__*/require("@deity/falcon-ui");

var _helpers = /*#__PURE__*/require("../helpers");

var _FormFieldError = /*#__PURE__*/require("./FormFieldError");

var _FormFieldLayout = /*#__PURE__*/require("./FormFieldLayout");

var CheckboxFormFieldLayout = /*#__PURE__*/(0, _falconUi.themed)({
  tag: _falconUi.Box,
  defaultTheme: {
    checkboxFormFieldLayout: {
      display: 'grid',
      gridGap: 'xs',
      // prettier-ignore
      gridTemplate: /*#__PURE__*/(0, _helpers.toGridTemplate)([['auto', '1fr'], [_FormFieldLayout.FormFieldArea.input, _FormFieldLayout.FormFieldArea.label], [_FormFieldLayout.FormFieldArea.error, _FormFieldLayout.FormFieldArea.error, '0px']])
    }
  }
});
exports.CheckboxFormFieldLayout = CheckboxFormFieldLayout;

var CheckboxFormField = function CheckboxFormField(props) {
  var name = props.name,
      validate = props.validate,
      required = props.required,
      children = props.children,
      restProps = (0, _objectWithoutPropertiesLoose2.default)(props, ["name", "validate", "required", "children"]);

  var _extractThemableProps = (0, _falconUi.extractThemableProps)(restProps),
      themableProps = _extractThemableProps.themableProps,
      rest = _extractThemableProps.rest;

  return /*#__PURE__*/_react.default.createElement(_falconFrontKit.Field, (0, _extends2.default)({
    name: name,
    validate: (0, _falconFrontKit.getDefaultInputValidators)(props)
  }, rest), function (_ref) {
    var form = _ref.form,
        field = _ref.field,
        label = _ref.label,
        error = _ref.error;
    return /*#__PURE__*/_react.default.createElement(CheckboxFormFieldLayout, themableProps, /*#__PURE__*/_react.default.createElement(_falconUi.Checkbox, (0, _extends2.default)({}, field, {
      gridArea: _FormFieldLayout.FormFieldArea.input,
      checked: field.value,
      onChange: function onChange(e) {
        return form.setFieldValue(field.name, e.target.checked);
      }
    })), /*#__PURE__*/_react.default.createElement(_falconUi.FlexLayout, {
      alignItems: "center",
      gridArea: _FormFieldLayout.FormFieldArea.label
    }, /*#__PURE__*/_react.default.createElement(_falconUi.Label, {
      htmlFor: field.id
    }, label)), /*#__PURE__*/_react.default.createElement(_FormFieldError.FormFieldError, null, field.invalid ? error : null));
  });
};

exports.CheckboxFormField = CheckboxFormField;
