// @deity/falcon-shop-data/dist/Address/AddAddressMutation.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.useAddAddressMutation = exports.AddAddressMutation = exports.ADD_ADDRESS = void 0;

var _inheritsLoose2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/inheritsLoose"));

var _client = /*#__PURE__*/require("@apollo/client");

var _falconData = /*#__PURE__*/require("@deity/falcon-data");

var ADD_ADDRESS = {
  "kind": "Document",
  "definitions": [{
    "kind": "OperationDefinition",
    "operation": "mutation",
    "name": {
      "kind": "Name",
      "value": "AddAddress"
    },
    "variableDefinitions": [{
      "kind": "VariableDefinition",
      "variable": {
        "kind": "Variable",
        "name": {
          "kind": "Name",
          "value": "input"
        }
      },
      "type": {
        "kind": "NonNullType",
        "type": {
          "kind": "NamedType",
          "name": {
            "kind": "Name",
            "value": "AddAddressInput"
          }
        }
      },
      "directives": []
    }],
    "directives": [],
    "selectionSet": {
      "kind": "SelectionSet",
      "selections": [{
        "kind": "Field",
        "name": {
          "kind": "Name",
          "value": "addAddress"
        },
        "arguments": [{
          "kind": "Argument",
          "name": {
            "kind": "Name",
            "value": "input"
          },
          "value": {
            "kind": "Variable",
            "name": {
              "kind": "Name",
              "value": "input"
            }
          }
        }],
        "directives": [],
        "selectionSet": {
          "kind": "SelectionSet",
          "selections": [{
            "kind": "Field",
            "name": {
              "kind": "Name",
              "value": "id"
            },
            "arguments": [],
            "directives": []
          }]
        }
      }]
    }
  }],
  "loc": {
    "start": 0,
    "end": 102,
    "source": {
      "body": "\n  mutation AddAddress($input: AddAddressInput!) {\n    addAddress(input: $input) {\n      id\n    }\n  }\n",
      "name": "GraphQL request",
      "locationOffset": {
        "line": 1,
        "column": 1
      }
    }
  }
};
exports.ADD_ADDRESS = ADD_ADDRESS;

var AddAddressMutation = /*#__PURE__*/function (_Mutation) {
  (0, _inheritsLoose2.default)(AddAddressMutation, _Mutation);

  function AddAddressMutation() {
    return _Mutation.apply(this, arguments) || this;
  }

  return AddAddressMutation;
}(_falconData.Mutation);

exports.AddAddressMutation = AddAddressMutation;
AddAddressMutation.defaultProps = {
  mutation: ADD_ADDRESS,
  refetchQueries: ['AddressList'],
  awaitRefetchQueries: true
};

var useAddAddressMutation = function useAddAddressMutation(options) {
  if (options === void 0) {
    options = {};
  }

  return (0, _client.useMutation)(ADD_ADDRESS, Object.assign({
    refetchQueries: ['AddressList'],
    awaitRefetchQueries: true
  }, options));
};

exports.useAddAddressMutation = useAddAddressMutation;
