// @deity/falcon-ui-kit/dist/Sidebar/Sidebar.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.Sidebar = void 0;

var _react = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("react"));

var _propTypes = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("prop-types"));

var _falconUi = /*#__PURE__*/require("@deity/falcon-ui");

var Sidebar = function Sidebar(_ref) {
  var close = _ref.close,
      isOpen = _ref.isOpen,
      side = _ref.side,
      children = _ref.children;
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_falconUi.Portal, null, /*#__PURE__*/_react.default.createElement(_falconUi.Sidebar, {
    visible: isOpen,
    side: side
  }, children), /*#__PURE__*/_react.default.createElement(_falconUi.Backdrop, {
    as: _falconUi.Portal,
    visible: isOpen,
    onClick: function onClick() {
      return close && close();
    }
  })));
};

exports.Sidebar = Sidebar;
Sidebar.propTypes = {
  isOpen: _propTypes.default.bool,
  close: _propTypes.default.func,
  side: /*#__PURE__*/_propTypes.default.oneOf(['left', 'right'])
};
Sidebar.defaultProps = {
  side: 'right'
};
