// @deity/falcon-ui-kit/dist/Newsletter/Newsletter.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.Newsletter = void 0;

var _react = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("react"));

var _falconI18n = /*#__PURE__*/require("@deity/falcon-i18n");

var _falconUi = /*#__PURE__*/require("@deity/falcon-ui");

var _NewsletterLayout = /*#__PURE__*/require("./NewsletterLayout");

var Newsletter = function Newsletter() {
  return /*#__PURE__*/_react.default.createElement(_NewsletterLayout.NewsletterLayout, null, /*#__PURE__*/_react.default.createElement(_falconUi.H3, null, /*#__PURE__*/_react.default.createElement(_falconI18n.T, {
    id: "newsletter.title"
  })), /*#__PURE__*/_react.default.createElement(_falconUi.Text, null, /*#__PURE__*/_react.default.createElement(_falconI18n.T, {
    id: "newsletter.message"
  })), /*#__PURE__*/_react.default.createElement(_falconI18n.I18n, null, function (t) {
    return /*#__PURE__*/_react.default.createElement(_falconUi.Box, {
      as: "form",
      justifySelf: "stretch"
    }, /*#__PURE__*/_react.default.createElement(_falconUi.Group, null, /*#__PURE__*/_react.default.createElement(_falconUi.Input, {
      type: "email",
      required: true,
      "aria-label": t('newsletter.emailPlaceholder'),
      placeholder: t('newsletter.emailPlaceholder')
    }), /*#__PURE__*/_react.default.createElement(_falconUi.Button, {
      as: "input",
      type: "submit",
      value: t('newsletter.subscribe'),
      flex: "none"
    })), /*#__PURE__*/_react.default.createElement(_falconUi.Label, {
      htmlFor: "subscribe",
      my: "sm",
      display: "flex",
      justifyContent: "center",
      alignItems: "center"
    }, /*#__PURE__*/_react.default.createElement(_falconUi.Checkbox, {
      id: "subscribe",
      required: true,
      mr: "xs"
    }), " ", t('newsletter.consent')));
  }));
};

exports.Newsletter = Newsletter;
