// @deity/falcon-front-kit/dist/Search/searchState.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.searchStateFromURL = searchStateFromURL;
exports.searchStateToURL = searchStateToURL;

var _qs = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("qs"));

var _falconShopData = /*#__PURE__*/require("@deity/falcon-shop-data");

function searchStateFromURL(url) {
  var parts = _qs.default.parse(url.replace('?', ''));

  var searchState = {};

  if (parts.q) {
    searchState.term = parts.q;
  }

  if (parts.p) {
    searchState.pagination = {
      page: parseInt(parts.p, 10),
      perPage: parseInt(parts.pp, 10)
    };
  }

  if (parts.sort) {
    var _split = parts.sort.split(':'),
        field = _split[0],
        direction = _split[1];

    searchState.sort = {
      field: field,
      direction: direction
    };
  }

  if (parts.filters) {
    var names = Object.keys(parts.filters);
    searchState.filters = [];

    for (var i = 0; i < names.length; i++) {
      var _names$i$split = names[i].split(':'),
          _field = _names$i$split[0],
          operator = _names$i$split[1];

      searchState.filters.push({
        field: _field,
        operator: operator || _falconShopData.FilterOperator.equals,
        value: parts.filters[names[i]].split(',')
      });
    }
  }

  return searchState;
}

function searchStateToURL(state) {
  var term = state.term,
      filters = state.filters,
      sort = state.sort,
      pagination = state.pagination;
  var parts = {};

  if (filters) {
    parts.filters = {};

    for (var i = 0; i < filters.length; i++) {
      var filter = filters[i];
      var name = filter.operator === _falconShopData.FilterOperator.equals ? filter.field : filter.field + ":" + filter.operator;
      parts.filters[name] = filter.value.join(',');
    }
  }

  if (sort && sort.field && sort.direction) {
    parts.sort = sort.field + ":" + sort.direction;
  }

  if (pagination) {
    if (pagination.perPage) {
      parts.pp = pagination.perPage.toString();
    }

    if (pagination.page && pagination.page > 1) {
      parts.p = pagination.page.toString();
    }
  }

  if (term) {
    parts.q = term; // eslint-disable-line id-length
  }

  return _qs.default.stringify(parts, {
    encode: false
  });
}
