// @deity/falcon-front-kit/dist/Forms/Field.js
"use strict";

var _interopRequireWildcard = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.Field = exports.getPlaceholderI18nId = exports.getLabelI18nId = void 0;

var _objectWithoutPropertiesLoose2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("react"));

var _formik = /*#__PURE__*/require("formik");

var _falconI18n = /*#__PURE__*/require("@deity/falcon-i18n");

var _FormContext = /*#__PURE__*/require("./FormContext");

var _IValidator = /*#__PURE__*/require("./IValidator");

var translateIfExists = function translateIfExists(t, key) {
  return key ? t(key, {
    defaultValue: ''
  }) : undefined;
};

var LABEL_SUFFIX = 'FieldLabel';
var PLACEHOLDER_SUFFIX = 'FieldPlaceholder';

var getLabelI18nId = function getLabelI18nId(formI18nId, fieldName) {
  return formI18nId + "." + fieldName + LABEL_SUFFIX;
};

exports.getLabelI18nId = getLabelI18nId;

var getPlaceholderI18nId = function getPlaceholderI18nId(formI18nId, fieldName) {
  return formI18nId + "." + fieldName + PLACEHOLDER_SUFFIX;
};

exports.getPlaceholderI18nId = getPlaceholderI18nId;

var Field = /*#__PURE__*/_react.default.memo(function (props) {
  var name = props.name,
      label = props.label,
      placeholder = props.placeholder,
      validate = props.validate,
      children = props.children,
      fast = props.fast,
      restProps = (0, _objectWithoutPropertiesLoose2.default)(props, ["name", "label", "placeholder", "validate", "children", "fast"]);

  var _useI18n = (0, _falconI18n.useI18n)(),
      t = _useI18n.t;

  var _useFormContext = (0, _FormContext.useFormContext)(),
      formId = _useFormContext.id,
      formI18nId = _useFormContext.i18nId;

  var i18nIds = formI18nId ? {
    label: getLabelI18nId(formI18nId, name),
    placeholder: getPlaceholderI18nId(formI18nId, name)
  } : {};
  var fieldLabel = label || translateIfExists(t, i18nIds.label);
  var fieldPlaceholder = placeholder || translateIfExists(t, i18nIds.placeholder);
  var validateCallback = (0, _react.useCallback)(function () {
    return (0, _IValidator.fieldValidator)(validate, {
      name: name,
      label: fieldLabel,
      formI18nId: formI18nId,
      t: t
    });
  }, [validate, name, fieldLabel, formI18nId, t]);

  if (!children) {
    return null;
  }

  var FieldComponent = fast ? _formik.FastField : _formik.Field;
  return /*#__PURE__*/_react.default.createElement(FieldComponent, {
    name: name,
    validate: validateCallback()
  }, function (_ref) {
    var form = _ref.form,
        field = _ref.field,
        _ref$meta = _ref.meta,
        touched = _ref$meta.touched,
        error = _ref$meta.error;
    return children({
      form: Object.assign({}, form, {
        id: formId
      }),
      field: Object.assign({
        id: formId + "-" + name
      }, restProps, field, {
        placeholder: fieldPlaceholder,
        invalid: !!touched && !!error
      }),
      error: error,
      label: fieldLabel,
      i18nIds: i18nIds
    });
  });
});

exports.Field = Field;
