// @deity/falcon-front-kit/dist/Forms/IValidator.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.isI18nValidationError = isI18nValidationError;
exports.fieldValidator = exports.getErrorI18nId = void 0;

var _objectWithoutPropertiesLoose2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _string = /*#__PURE__*/require("./string");

function isI18nValidationError(error) {
  return typeof error === 'object' && error.errorI18nId !== undefined;
}

var getErrorI18nId = function getErrorI18nId(error, name, formI18nId) {
  if (formI18nId) {
    return [formI18nId + "." + name + "Field" + (0, _string.capitalize)(error), "formError." + error];
  }

  return "formError." + error;
};

exports.getErrorI18nId = getErrorI18nId;

var fieldValidator = function fieldValidator(validators, _ref) {
  if (validators === void 0) {
    validators = [];
  }

  var name = _ref.name,
      label = _ref.label,
      formI18nId = _ref.formI18nId,
      t = _ref.t;
  return function (value) {
    for (var i = 0; i < validators.length; i++) {
      var result = validators[i]({
        name: name,
        label: label,
        value: value,
        formI18nId: formI18nId,
        t: t
      });

      if (isI18nValidationError(result)) {
        var _error = result.errorI18nId,
            errorProps = (0, _objectWithoutPropertiesLoose2.default)(result, ["errorI18nId"]);
        return t(getErrorI18nId(_error, name, formI18nId), Object.assign({
          name: name,
          value: value
        }, errorProps, {
          label: label || (0, _string.capitalize)(name)
        }));
      }

      if (result !== undefined) {
        return result;
      }
    }

    return undefined;
  };
};

exports.fieldValidator = fieldValidator;
