// @deity/falcon-front-kit/dist/Forms/validators.js
"use strict";

exports.__esModule = true;
exports.isCheckedValidator = isCheckedValidator;
exports.confirmPasswordValidator = confirmPasswordValidator;
exports.lengthValidator = lengthValidator;
exports.numLinesValidator = numLinesValidator;
exports.rangeValidator = rangeValidator;
exports.dateRangeValidator = dateRangeValidator;
exports.integerValidator = exports.numberValidator = exports.passwordValidator = exports.emailValidator = exports.requiredValidator = void 0;
var VALID_EMAIL_REGEX = /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
var VALID_PASSWORD_REGEX = /^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)/;

var isValueProvided = function isValueProvided(x) {
  return x !== '' && x !== undefined && x !== null && x !== false && x !== 'placeholder' || Array.isArray(x) && x.length;
};

var requiredValidator = function requiredValidator(_ref) {
  var value = _ref.value;

  if (!isValueProvided(value)) {
    return {
      errorI18nId: 'required'
    };
  }

  return undefined;
};

exports.requiredValidator = requiredValidator;

function isCheckedValidator(checkedValue) {
  if (checkedValue === void 0) {
    checkedValue = true;
  }

  return function (_ref2) {
    var value = _ref2.value;

    if (!isValueProvided(value) || value !== checkedValue) {
      return {
        errorI18nId: 'isNotChecked'
      };
    }

    return undefined;
  };
}

var emailValidator = function emailValidator(_ref3) {
  var value = _ref3.value;

  if (isValueProvided(value) && !VALID_EMAIL_REGEX.test(value.toLowerCase())) {
    return {
      errorI18nId: 'invalid'
    };
  }

  return undefined;
};

exports.emailValidator = emailValidator;

var passwordValidator = function passwordValidator(_ref4) {
  var value = _ref4.value;

  if (isValueProvided(value) && !VALID_PASSWORD_REGEX.test(value)) {
    return {
      errorI18nId: 'uncomplexPassword'
    };
  }

  return undefined;
};

exports.passwordValidator = passwordValidator;

function confirmPasswordValidator(password) {
  if (password === void 0) {
    password = '';
  }

  return function (_ref5) {
    var value = _ref5.value;

    if (password !== '' && isValueProvided(value)) {
      return password !== value ? {
        errorI18nId: 'differentPasswords'
      } : undefined;
    }

    return undefined;
  };
}

function lengthValidator(min, max) {
  if (min === void 0) {
    min = 0;
  }

  if (max && max < min) {
    throw new Error("value of 'min' can not be greater than 'max'!");
  }

  return function (_ref6) {
    var value = _ref6.value;

    if (isValueProvided(value) && typeof value === 'string') {
      // Ignore newlines when counting characters
      var _value$replace = value.replace(/\n/g, ''),
          length = _value$replace.length;

      if (length < min) {
        return Object.assign({
          errorI18nId: 'tooShort',
          min: min
        }, max ? {
          max: max
        } : {});
      }

      if (max && length > max) {
        return {
          errorI18nId: 'tooLong',
          min: min,
          max: max
        };
      }
    }

    return undefined;
  };
}

function numLinesValidator(min, max) {
  if (min === void 0) {
    min = 0;
  }

  if (max && max < min) {
    throw new Error("value of 'min' can not be greater than 'max'!");
  }

  return function (_ref7) {
    var value = _ref7.value;

    if (isValueProvided(value) && typeof value === 'string') {
      var numLines = value.split('\n').length;

      if (numLines < min) {
        return {
          errorI18nId: 'tooLittleLines',
          min: min
        };
      }

      if (numLines > max) {
        return {
          errorI18nId: 'tooManyLines',
          max: max
        };
      }
    }

    return undefined;
  };
}

var numberValidator = function numberValidator(_ref8) {
  var value = _ref8.value;

  if (isValueProvided(value)) {
    if (value.toString() !== parseFloat(value).toString()) {
      return {
        errorI18nId: 'notAnNumber'
      };
    }
  }

  return undefined;
};

exports.numberValidator = numberValidator;

var integerValidator = function integerValidator(_ref9) {
  var value = _ref9.value;

  if (isValueProvided(value)) {
    if (value.toString() !== parseInt(value, 10).toString()) {
      return {
        errorI18nId: 'notInteger'
      };
    }
  }

  return undefined;
};

exports.integerValidator = integerValidator;

function rangeValidator(min, max) {
  if (max !== undefined && max < min) {
    throw new Error("value of 'min' can not be greater than 'max'!");
  }

  return function (_ref10) {
    var value = _ref10.value;
    value = parseInt(value.toString(), 10);

    if (isValueProvided(value) && typeof value === 'number') {
      if (value < min) {
        return Object.assign({
          errorI18nId: 'belowRange',
          min: min
        }, max ? {
          max: max
        } : {});
      }

      if (max && value > max) {
        return {
          errorI18nId: 'aboveRange',
          min: min,
          max: max
        };
      }
    }

    return undefined;
  };
}

function dateRangeValidator(min, max, dateTimeFormat) {
  var minDate = min && new Date(min);
  var maxDate = max && new Date(max);
  var formattedMinDate = (dateTimeFormat == null ? void 0 : dateTimeFormat(minDate)) || minDate.toISOString().split('T')[0];
  var formattedMaxDate = (dateTimeFormat == null ? void 0 : dateTimeFormat(maxDate)) || maxDate.toISOString().split('T')[0];

  if (maxDate && maxDate < minDate) {
    throw new Error("value of 'min' can not be earlier than 'max'!");
  }

  return function (_ref11) {
    var value = _ref11.value;

    if (isValueProvided(value)) {
      var valueDate = new Date(value);

      if (minDate && valueDate < minDate) {
        return {
          errorI18nId: 'belowDateRange',
          min: formattedMinDate
        };
      }

      if (maxDate && valueDate > maxDate) {
        return {
          errorI18nId: 'aboveDateRange',
          max: formattedMaxDate
        };
      }
    }

    return undefined;
  };
}
