// @deity/falcon-front-kit/dist/Checkout/PlaceOrder.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.PlaceOrder = exports.usePlaceOrder = void 0;

var _regenerator = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/asyncToGenerator"));

var _react = /*#__PURE__*/require("react");

var _falconShopData = /*#__PURE__*/require("@deity/falcon-shop-data");

var _CheckoutAddress = /*#__PURE__*/require("./CheckoutAddress");

var _CheckoutConsumer = /*#__PURE__*/require("./CheckoutConsumer");

var _shippingMethodToCheckoutDetailsInput = /*#__PURE__*/require("./shippingMethodToCheckoutDetailsInput");

var _PaymentMethodData = /*#__PURE__*/require("./PaymentMethodData");

var usePlaceOrder = function usePlaceOrder() {
  var _useCheckout = (0, _CheckoutConsumer.useCheckout)(),
      isLoading = _useCheckout.isLoading,
      setLoading = _useCheckout.setLoading,
      setOrderData = _useCheckout.setOrderData,
      setResult = _useCheckout.setResult;

  var _usePlaceOrderMutatio = (0, _falconShopData.usePlaceOrderMutation)({
    onCompleted: function onCompleted(data) {
      if (data) {
        setResult(data.placeOrder);
        setLoading(false);
      }
    }
  }),
      mutation = _usePlaceOrderMutatio[0],
      result = _usePlaceOrderMutatio[1];

  if (result.loading !== isLoading) {
    setLoading(result.loading);
  }

  return [(0, _react.useCallback)( /*#__PURE__*/function () {
    var _ref = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee(input, options) {
      return _regenerator.default.wrap(function _callee$(_context) {
        while (1) {
          switch (_context.prev = _context.next) {
            case 0:
              setLoading(true);
              setOrderData(input);
              return _context.abrupt("return", mutation(Object.assign({}, options || {}, {
                variables: {
                  input: {
                    email: input.email,
                    billingAddress: (0, _CheckoutAddress.addressToCheckoutAddressInput)(input.billingAddress),
                    shippingAddress: (0, _CheckoutAddress.addressToCheckoutAddressInput)(input.shippingAddress),
                    shippingMethod: (0, _shippingMethodToCheckoutDetailsInput.shippingMethodToCheckoutDetailsInput)(input.shippingMethod),
                    paymentMethod: input.paymentMethod ? (0, _PaymentMethodData.paymentMethodToCheckoutDetailsInput)(input.paymentMethod) : null
                  }
                }
              })));

            case 3:
            case "end":
              return _context.stop();
          }
        }
      }, _callee);
    }));

    return function (_x, _x2) {
      return _ref.apply(this, arguments);
    };
  }(), [setLoading, setOrderData, mutation]), result];
};

exports.usePlaceOrder = usePlaceOrder;

var PlaceOrder = function PlaceOrder(_ref2) {
  var children = _ref2.children;
  return children.apply(void 0, usePlaceOrder());
};

exports.PlaceOrder = PlaceOrder;
