// @deity/falcon-shop-data/dist/Cart/ApplyCouponMutation.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.useApplyCouponMutation = exports.ApplyCouponMutation = exports.APPLY_COUPON = void 0;

var _inheritsLoose2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/inheritsLoose"));

var _client = /*#__PURE__*/require("@apollo/client");

var _falconData = /*#__PURE__*/require("@deity/falcon-data");

var APPLY_COUPON = {
  "kind": "Document",
  "definitions": [{
    "kind": "OperationDefinition",
    "operation": "mutation",
    "name": {
      "kind": "Name",
      "value": "ApplyCoupon"
    },
    "variableDefinitions": [{
      "kind": "VariableDefinition",
      "variable": {
        "kind": "Variable",
        "name": {
          "kind": "Name",
          "value": "input"
        }
      },
      "type": {
        "kind": "NonNullType",
        "type": {
          "kind": "NamedType",
          "name": {
            "kind": "Name",
            "value": "CouponInput"
          }
        }
      },
      "directives": []
    }],
    "directives": [],
    "selectionSet": {
      "kind": "SelectionSet",
      "selections": [{
        "kind": "Field",
        "name": {
          "kind": "Name",
          "value": "applyCoupon"
        },
        "arguments": [{
          "kind": "Argument",
          "name": {
            "kind": "Name",
            "value": "input"
          },
          "value": {
            "kind": "Variable",
            "name": {
              "kind": "Name",
              "value": "input"
            }
          }
        }],
        "directives": []
      }]
    }
  }],
  "loc": {
    "start": 0,
    "end": 83,
    "source": {
      "body": "\n  mutation ApplyCoupon($input: CouponInput!) {\n    applyCoupon(input: $input)\n  }\n",
      "name": "GraphQL request",
      "locationOffset": {
        "line": 1,
        "column": 1
      }
    }
  }
};
exports.APPLY_COUPON = APPLY_COUPON;
var defaultOptions = {
  mutation: APPLY_COUPON,
  refetchQueries: ['Cart', 'CheckoutState'],
  awaitRefetchQueries: true
};

var ApplyCouponMutation = /*#__PURE__*/function (_Mutation) {
  (0, _inheritsLoose2.default)(ApplyCouponMutation, _Mutation);

  function ApplyCouponMutation() {
    return _Mutation.apply(this, arguments) || this;
  }

  return ApplyCouponMutation;
}(_falconData.Mutation);

exports.ApplyCouponMutation = ApplyCouponMutation;
ApplyCouponMutation.defaultProps = defaultOptions;

var useApplyCouponMutation = function useApplyCouponMutation(options) {
  if (options === void 0) {
    options = {};
  }

  return (0, _client.useMutation)(defaultOptions.mutation, Object.assign({}, defaultOptions, options));
};

exports.useApplyCouponMutation = useApplyCouponMutation;
