// @deity/falcon-shop-data/dist/Cart/RestoreCartMutation.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.useRestoreCartMutation = exports.RestoreCartMutation = exports.RESTORE_CART = void 0;

var _inheritsLoose2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/inheritsLoose"));

var _client = /*#__PURE__*/require("@apollo/client");

var _falconData = /*#__PURE__*/require("@deity/falcon-data");

var RESTORE_CART = {
  "kind": "Document",
  "definitions": [{
    "kind": "OperationDefinition",
    "operation": "mutation",
    "name": {
      "kind": "Name",
      "value": "RestoreCart"
    },
    "variableDefinitions": [],
    "directives": [],
    "selectionSet": {
      "kind": "SelectionSet",
      "selections": [{
        "kind": "Field",
        "name": {
          "kind": "Name",
          "value": "restoreCart"
        },
        "arguments": [],
        "directives": []
      }]
    }
  }],
  "loc": {
    "start": 0,
    "end": 46,
    "source": {
      "body": "\n  mutation RestoreCart {\n    restoreCart\n  }\n",
      "name": "GraphQL request",
      "locationOffset": {
        "line": 1,
        "column": 1
      }
    }
  }
};
exports.RESTORE_CART = RESTORE_CART;

var RestoreCartMutation = /*#__PURE__*/function (_Mutation) {
  (0, _inheritsLoose2.default)(RestoreCartMutation, _Mutation);

  function RestoreCartMutation() {
    return _Mutation.apply(this, arguments) || this;
  }

  return RestoreCartMutation;
}(_falconData.Mutation);

exports.RestoreCartMutation = RestoreCartMutation;
RestoreCartMutation.defaultProps = {
  mutation: RESTORE_CART,
  refetchQueries: ['CheckoutState', 'Cart', 'MiniCart'],
  awaitRefetchQueries: true
};

var useRestoreCartMutation = function useRestoreCartMutation(options) {
  return (0, _client.useMutation)(RESTORE_CART, {
    refetchQueries: ['CheckoutState', 'Cart', 'MiniCart'],
    awaitRefetchQueries: true
  });
};

exports.useRestoreCartMutation = useRestoreCartMutation;
