// @deity/falcon-client/build-utils/webpack/config/env.js
const path = require('path');

/**
 * @typedef {object} ClientFalconEnv
 * @property {string} NODE_ENV
 * @property {string} BABEL_ENV
 * @property {string | undefined} DEV_SERVER_PORT
 * @property {'client' | 'server'} BUILD_TARGET
 * @property {string | undefined} PUBLIC_PATH
 * @property {string} WEBPACK_ASSETS
 * @property {string} OUTPUT_DIR
 * @property {string} PUBLIC_DIR
 * @property {string} SW_DIR
 */

/**
 * @typedef {object} ServerFalconEnv
 * @property {string} NODE_ENV
 * @property {string} BABEL_ENV
 * @property {string | undefined} DEV_SERVER_PORT
 * @property {'client' | 'server'} BUILD_TARGET
 * @property {string | undefined} PUBLIC_PATH
 * @property {string} WEBPACK_ASSETS
 * @property {string} OUTPUT_DIR
 * @property {string} PUBLIC_DIR
 * @property {string} SW_DIR
 * @property {string} HTTPS_ENABLED
 * @property {string} HTTPS_KEY_PATH
 * @property {string} HTTPS_CERT_PATH
 */

/**
 * @param {'web'|'node'} target
 * @param {'development'|'production'| 'test'} env
 * @param {string} publicPath
 * @param {import('../../paths')} paths
 * @param {boolean} startDevServer
 * @param {string} devServerPort
 * @param {{enabled: boolean; key?: string, cert?: string}} httpsConfig
 * @param {string[]} envToBuildIn
 * @returns {{client: ClientFalconEnv, server: ServerFalconEnv}}
 */
function buildClientEnv(
  target,
  env,
  publicPath,
  paths,
  startDevServer = false,
  devServerPort = undefined,
  httpsConfig = undefined,
  envToBuildIn = []
) {
  httpsConfig = httpsConfig || { enabled: false };

  const envVariablesToBuildIn = Object.keys(process.env)
    .filter(x => envToBuildIn.find(e => e === x))
    .reduce((result, x) => {
      result[x] = process.env[x];
      return result;
    }, {});
  const environmentEnvVariables = {
    NODE_ENV: env,
    BABEL_ENV: env,
    ...(env === 'development' ? { DEV_SERVER_PORT: devServerPort } : {}),
    BUILD_TARGET: target === 'web' ? 'client' : 'server'
  };
  const buildEnvVariables = {
    PUBLIC_PATH: env === 'production' ? publicPath : undefined,
    WEBPACK_ASSETS: paths.appWebpackAssets,
    OUTPUT_DIR: path.relative(paths.appPath, paths.appBuildPublic),
    PUBLIC_DIR: path.relative(paths.appPath, !startDevServer ? paths.appBuildPublic : paths.appPublic),
    SW_DIR: path.relative(paths.appPath, env === 'production' ? paths.appBuildPublic : paths.appBuild)
  };
  const httpsEnvVariables = httpsConfig.enabled
    ? {
        HTTPS_ENABLED: httpsConfig.enabled,
        HTTPS_KEY_PATH: startDevServer
          ? httpsConfig.key
          : path.join(paths.appRelativeBuildSSL, path.basename(httpsConfig.key)),
        HTTPS_CERT_PATH: startDevServer
          ? httpsConfig.cert
          : path.join(paths.appRelativeBuildSSL, path.basename(httpsConfig.cert))
      }
    : {};

  return {
    client: {
      ...environmentEnvVariables,
      ...envVariablesToBuildIn,
      ...buildEnvVariables
    },
    server: {
      ...environmentEnvVariables,
      ...envVariablesToBuildIn,
      ...buildEnvVariables,
      ...httpsEnvVariables
    }
  };
}

/**
 * @param {object} data env variables name/value map
 * @returns {object}
 */
function serializeEnvVariables(data) {
  return Object.keys(data).reduce((result, x) => {
    result[`process.env.${x}`] = JSON.stringify(data[x]);
    return result;
  }, {});
}

module.exports = {
  buildClientEnv,
  serializeEnvVariables
};
