// @deity/falcon-front-kit/dist/Filters/FiltersDataProvider.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.FiltersDataProvider = exports.getSelectedFilterOptionsFor = exports.getFiltersData = void 0;

var _react = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("react"));

var _Search = /*#__PURE__*/require("../Search");

var _FilterData = /*#__PURE__*/require("./FilterData");

var getFiltersData = function getFiltersData(filters, aggregations, filterData) {
  if (aggregations === void 0) {
    aggregations = [];
  }

  if (filterData === void 0) {
    filterData = [];
  }

  return [].concat(aggregations.map(function (x) {
    return (0, _FilterData.aggregationToFilterData)(x);
  }), filterData).map(function (x) {
    var selected = filters.find(function (filter) {
      return filter.field === x.field;
    });
    return Object.assign({}, x, {
      value: selected ? selected.value : []
    });
  });
};

exports.getFiltersData = getFiltersData;

var getSelectedFilterOptionsFor = function getSelectedFilterOptionsFor(data, field) {
  var filter = data.find(function (x) {
    return x.field === field;
  });

  if (filter === undefined) {
    return [];
  }

  var options = filter.options,
      value = filter.value;
  return options.filter(function (option) {
    return value.some(function (x) {
      return x === option.value;
    });
  });
};

exports.getSelectedFilterOptionsFor = getSelectedFilterOptionsFor;

var FiltersDataProvider = function FiltersDataProvider(_ref) {
  var children = _ref.children,
      aggregations = _ref.aggregations,
      data = _ref.data;
  return /*#__PURE__*/_react.default.createElement(_Search.SearchConsumer, null, function (_ref2) {
    var filters = _ref2.state.filters,
        setFilter = _ref2.setFilter,
        removeFilter = _ref2.removeFilter,
        removeFilters = _ref2.removeFilters;
    return children({
      filters: getFiltersData(filters, aggregations || [], data || []),
      anySelected: filters.length > 0,
      getSelectedFilterOptionsFor: getSelectedFilterOptionsFor,
      setFilter: setFilter,
      removeFilter: removeFilter,
      removeFilters: removeFilters
    });
  });
};

exports.FiltersDataProvider = FiltersDataProvider;
