// @deity/falcon-ui-kit/dist/Address/EditAddressForm.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.EditAddressForm = void 0;

var _extends2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("react"));

var _falconFrontKit = /*#__PURE__*/require("@deity/falcon-front-kit");

var _falconUi = /*#__PURE__*/require("@deity/falcon-ui");

var _falconI18n = /*#__PURE__*/require("@deity/falcon-i18n");

var _falconShopData = /*#__PURE__*/require("@deity/falcon-shop-data");

var _Forms = /*#__PURE__*/require("../Forms");

var _Error = /*#__PURE__*/require("../Error");

var _AddressFormFields = /*#__PURE__*/require("./AddressFormFields");

var EditAddressForm = function EditAddressForm(_ref) {
  var id = _ref.id,
      onSuccess = _ref.onSuccess,
      onCancel = _ref.onCancel,
      formProps = (0, _objectWithoutPropertiesLoose2.default)(_ref, ["id", "onSuccess", "onCancel"]);
  return /*#__PURE__*/_react.default.createElement(_falconShopData.AddressQuery, {
    variables: {
      id: id
    }
  }, function (_ref2) {
    var address = _ref2.data.address;
    return /*#__PURE__*/_react.default.createElement(_falconFrontKit.EditAddressFormProvider, {
      address: address,
      onSuccess: onSuccess
    }, function (_ref3) {
      var isSubmitting = _ref3.isSubmitting,
          status = _ref3.status;
      return /*#__PURE__*/_react.default.createElement(_Forms.Form, (0, _extends2.default)({
        id: "edit-address",
        i18nId: "addressForm"
      }, formProps), /*#__PURE__*/_react.default.createElement(_AddressFormFields.AddressFormFields, {
        twoColumns: true,
        askDefault: true
      }), /*#__PURE__*/_react.default.createElement(_falconUi.FlexLayout, {
        justifyContent: "flex-end",
        alignItems: "center",
        mt: "md"
      }, /*#__PURE__*/_react.default.createElement(_falconUi.Button, {
        onClick: onCancel,
        mr: "md"
      }, /*#__PURE__*/_react.default.createElement(_falconI18n.T, {
        id: "editAddress.cancelButton"
      })), /*#__PURE__*/_react.default.createElement(_falconUi.Button, {
        type: "submit",
        variant: isSubmitting ? 'loader' : undefined
      }, /*#__PURE__*/_react.default.createElement(_falconI18n.T, {
        id: "editAddress.submitButton"
      }))), /*#__PURE__*/_react.default.createElement(_Error.ErrorSummary, {
        errors: status.error
      }));
    });
  });
};

exports.EditAddressForm = EditAddressForm;
