// @deity/falcon-ui-kit/dist/Menu/MenuNavbar.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.MenuNavbar = void 0;

var _react = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("react"));

var _falconUi = /*#__PURE__*/require("@deity/falcon-ui");

var _Router = /*#__PURE__*/require("../Router");

var MenuNavbar = function MenuNavbar(_ref) {
  var items = _ref.items;
  return /*#__PURE__*/_react.default.createElement(_falconUi.Navbar, null, items.map(function (item) {
    return /*#__PURE__*/_react.default.createElement(_falconUi.NavbarItem, {
      key: item.urlPath
    }, /*#__PURE__*/_react.default.createElement(_Router.RouterLink, {
      to: item.urlPath,
      p: "sm"
    }, item.name), item.children && item.children.length > 0 && /*#__PURE__*/_react.default.createElement(_falconUi.NavbarItemMenu, null, /*#__PURE__*/_react.default.createElement(_falconUi.List, {
      width: "full",
      display: "grid",
      gridTemplateColumns: {
        xs: '1fr',
        sm: '1fr 1fr 1fr',
        md: '1fr 1fr 1fr 1fr',
        lg: '1fr 1fr 1fr 1fr 1fr'
      }
    }, item.children.map(function (subItem) {
      return /*#__PURE__*/_react.default.createElement(_falconUi.ListItem, {
        key: subItem.urlPath
      }, /*#__PURE__*/_react.default.createElement(_Router.RouterLink, {
        to: subItem.urlPath,
        display: "block",
        textDecoration: "underline",
        py: "xs"
      }, subItem.name), subItem.children && subItem.children.length > 0 && /*#__PURE__*/_react.default.createElement(_falconUi.List, null, subItem.children.map(function (nextSubItem) {
        return /*#__PURE__*/_react.default.createElement(_falconUi.ListItem, {
          key: nextSubItem.urlPath
        }, /*#__PURE__*/_react.default.createElement(_Router.RouterLink, {
          to: nextSubItem.urlPath,
          display: "block",
          py: "xs"
        }, nextSubItem.name));
      })));
    }))));
  }));
};

exports.MenuNavbar = MenuNavbar;
