// @deity/falcon-logger/dist/index.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.default = void 0;

var _pino = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("pino"));

var _symbols = /*#__PURE__*/require("pino/lib/symbols");

var levelToSeverity = function levelToSeverity(level) {
  switch (level) {
    case 'trace':
    case 'debug':
      return 'DEBUG';

    case 'info':
      return 'INFO';

    case 'warn':
      return 'WARNING';

    case 'error':
      return 'ERROR';

    case 'fatal':
      return 'CRITICAL';

    default:
      return 'DEFAULT';
  }
};

var isAppSet = false;
var logger = /*#__PURE__*/(0, _pino.default)({
  name: 'falcon-logger',
  formatters: {
    level: /*#__PURE__*/function (_level) {
      function level(_x, _x2) {
        return _level.apply(this, arguments);
      }

      level.toString = function () {
        return _level.toString();
      };

      return level;
    }(function (level, number) {
      return {
        level: number,
        severity: levelToSeverity(level)
      };
    })
  }
});
/**
 * Sets the log level for the logger instance
 * @param level Required log level
 * @returns {void}
 */

logger.setLogLevel = function (level) {
  logger.level = level;
};
/**
 * Sets the "app" key to every log message via the root logger instance.
 * It can be called only once, preferably at the very beginning of your application.
 * @param name Application name
 * @returns {void}
 */


logger.setApp = function (name) {
  if (isAppSet) {
    logger.warn('logger.setApp() can be called only once.');
    return;
  } // @ts-ignore: Tweaking internal Pino settings


  logger[_symbols.chindingsSym] += ",\"app\":\"" + name + "\"";
  isAppSet = true;
};
/**
 * Initializes an extra sub-logger instance for the provided module name
 * and adds "module" key to every log message automatically.
 * Handy for defining sub-loggers for your nested modules.
 * In conjunction with `falcon-pretty` formatter - it will render an additional "[my-module]" section in the log message output.
 * @example
 * logger.getFor("my-module").info("Calling my-module...");
 * @param moduleName Module name
 * @param extra map with values that will be added to each logger call automatically
 * @returns {Logger} Module-specific Logger instance
 */


logger.getFor = function (moduleName, extra) {
  if (extra === void 0) {
    extra = {};
  }

  return logger.child(Object.assign({
    module: moduleName
  }, extra));
};
/**
 * Measures the timing for the provided callback.
 * If your log level is set to trace - Logger will produce the following log message:
 * "TRACE: My time (XX ms)"
 * @param label Log label
 * @param fn Function to trace the execution time of
 * @returns {Promise<T>} `fn` result
 */


logger.traceTime = function traceTime(label, fn) {
  var _this = this;

  // using `function()` statement to preserve the context in case of "getFor" call
  if (!this.isLevelEnabled('trace')) {
    return fn();
  }

  var startTime = Date.now();
  return fn().finally(function () {
    var duration = Date.now() - startTime;

    _this.trace(label + " (" + duration + "ms)");
  });
};

var _default = logger;
exports.default = _default;
