// @deity/falcon-ui-kit/dist/Loader/Loader.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.Loader = void 0;

var _react = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("react"));

var _falconUi = /*#__PURE__*/require("@deity/falcon-ui");

var LoaderInnerDOM = function LoaderInnerDOM(props) {
  return /*#__PURE__*/_react.default.createElement(_falconUi.Box, props, /*#__PURE__*/_react.default.createElement(_falconUi.Icon, {
    src: "loader"
  }));
};

var Loader = /*#__PURE__*/(0, _falconUi.themed)({
  tag: LoaderInnerDOM,
  defaultTheme: {
    loader: {
      display: 'flex',
      justifyContent: 'center',
      alignItems: 'center',
      css: {
        height: '50vh',
        width: '100%'
      },
      variants: {
        overlay: {
          position: 'absolute',
          css: {
            background: 'rgba(255, 255, 255, 0.7)',
            height: '100%',
            width: '100%',
            zIndex: 1
          }
        }
      }
    }
  }
});
exports.Loader = Loader;
