// @deity/falcon-service-worker/dist/ServiceWorker.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.ServiceWorker = void 0;

var _assertThisInitialized2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/assertThisInitialized"));

var _inheritsLoose2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/inheritsLoose"));

var _react = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("react"));

var _ServiceWorkerContext = /*#__PURE__*/require("./ServiceWorkerContext");

var ServiceWorker = function ServiceWorker(_ref) {
  var children = _ref.children;
  return /*#__PURE__*/_react.default.createElement(_ServiceWorkerContext.ServiceWorkerContext.Consumer, null, function (_ref2) {
    var isSupported = _ref2.isSupported,
        registration = _ref2.registration;
    return isSupported && registration ? /*#__PURE__*/_react.default.createElement(ServiceWorkerInner, {
      registration: registration
    }, function (_ref3) {
      var isWaiting = _ref3.isWaiting,
          skipWaiting = _ref3.skipWaiting;
      return children({
        isSupported: isSupported,
        registration: registration,
        isWaiting: isWaiting,
        skipWaiting: skipWaiting
      });
    }) : children({
      isSupported: isSupported,
      isWaiting: false,
      skipWaiting: function skipWaiting() {}
    });
  });
};

exports.ServiceWorker = ServiceWorker;

var ServiceWorkerInner = /*#__PURE__*/function (_React$Component) {
  (0, _inheritsLoose2.default)(ServiceWorkerInner, _React$Component);

  function ServiceWorkerInner(props) {
    var _this;

    _this = _React$Component.call(this, props) || this;
    _this.skipWaiting = _this.skipWaiting.bind((0, _assertThisInitialized2.default)(_this));
    _this.onUpdateFound = _this.onUpdateFound.bind((0, _assertThisInitialized2.default)(_this));
    _this.state = {
      isWaiting: false
    };
    return _this;
  }

  var _proto = ServiceWorkerInner.prototype;

  _proto.componentDidMount = function componentDidMount() {
    var registration = this.props.registration;

    if (!navigator.serviceWorker.controller) {
      return;
    }

    if (registration.active && registration.waiting) {
      this.setState(function (state) {
        return Object.assign({}, state, {
          isWaiting: true
        });
      });
      return;
    }

    if (registration.installing) {
      return this.onUpdateFound();
    }

    registration.addEventListener('updatefound', this.onUpdateFound);
  };

  _proto.componentWillUnmount = function componentWillUnmount() {
    var registration = this.props.registration;
    registration.removeEventListener('updatefound', this.onUpdateFound);
  };

  _proto.onUpdateFound = function onUpdateFound() {
    var _this2 = this;

    var registration = this.props.registration;

    if (registration.installing) {
      registration.installing.addEventListener('statechange', function (event) {
        if (event.target.state === 'installed') {
          _this2.setState(function (state) {
            return Object.assign({}, state, {
              isWaiting: true
            });
          });
        }
      });
    }
  };

  _proto.skipWaiting = function skipWaiting() {
    var registration = this.props.registration;
    return registration.waiting ? registration.waiting.postMessage({
      type: 'SKIP_WAITING',
      payload: undefined
    }) : function () {};
  };

  _proto.render = function render() {
    var children = this.props.children;
    var isWaiting = this.state.isWaiting;
    return children({
      isWaiting: isWaiting,
      skipWaiting: this.skipWaiting
    });
  };

  return ServiceWorkerInner;
}(_react.default.Component);
