// @deity/falcon-service-worker/dist/ServiceWorkerRegistrar.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.ServiceWorkerRegistrar = void 0;

var _createClass2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/assertThisInitialized"));

var _inheritsLoose2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/inheritsLoose"));

var _react = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("react"));

var _propTypes = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("prop-types"));

var _ServiceWorkerContext = /*#__PURE__*/require("./ServiceWorkerContext");

var ServiceWorkerRegistrar = /*#__PURE__*/function (_React$Component) {
  (0, _inheritsLoose2.default)(ServiceWorkerRegistrar, _React$Component);

  function ServiceWorkerRegistrar(props) {
    var _this;

    _this = _React$Component.call(this, props) || this;
    _this.onControllerChange = _this.onControllerChange.bind((0, _assertThisInitialized2.default)(_this));
    _this.state = {
      isSupported: false,
      registration: undefined
    };
    return _this;
  }

  var _proto = ServiceWorkerRegistrar.prototype;

  _proto.componentDidMount = function componentDidMount() {
    var _this2 = this;

    if (!this.isSupported) {
      return;
    }

    this.whenReady(function () {
      var _this2$props = _this2.props,
          scriptUrl = _this2$props.scriptUrl,
          options = _this2$props.options;
      navigator.serviceWorker.register(scriptUrl, options).then(function (registration) {
        _this2.setState(function (state) {
          return Object.assign({}, state, {
            isSupported: true,
            registration: registration
          });
        });

        navigator.serviceWorker.addEventListener('controllerchange', _this2.onControllerChange);
        console.log("Service Worker '" + scriptUrl + "' registered in '" + registration.scope + "' scope.");
      }).catch(function (error) {
        console.warn("Service Worker '" + scriptUrl + "' registration failed.", error);
      });
    });
  };

  _proto.componentWillUnmount = function componentWillUnmount() {
    if (!this.isSupported) {
      return;
    }

    navigator.serviceWorker.removeEventListener('controllerchange', this.onControllerChange);
  };

  _proto.onControllerChange = function onControllerChange(event) {
    if (event.currentTarget !== navigator.serviceWorker.controller) {
      return window.location.reload();
    }
  };

  /**
   * if `document.readyState` is `complete` invoke `callback` immediately, otherwise delay it until `load` event is fired
   * @param {Function} callback function
   */
  _proto.whenReady = function whenReady(callback) {
    if (document.readyState === 'complete') {
      callback();
    } else {
      window.addEventListener('load', callback);
    }
  };

  _proto.render = function render() {
    var _this$props = this.props,
        children = _this$props.children,
        scriptUrl = _this$props.scriptUrl,
        options = _this$props.options;
    var _this$state = this.state,
        isSupported = _this$state.isSupported,
        registration = _this$state.registration;
    return /*#__PURE__*/_react.default.createElement(_ServiceWorkerContext.ServiceWorkerContext.Provider, {
      value: {
        isSupported: isSupported,
        registration: registration,
        scriptUrl: scriptUrl,
        options: options
      }
    }, children);
  };

  (0, _createClass2.default)(ServiceWorkerRegistrar, [{
    key: "isSupported",
    get: function get() {
      var _window = window,
          navigator = _window.navigator,
          location = _window.location;
      var isApiAvailable = navigator && 'serviceWorker' in navigator;
      var isHttps = location.protocol === 'https:';
      var isLocalHost = location.host.match(/(localhost|127.0.0.1)/);
      return isApiAvailable && (isHttps || isLocalHost);
    }
  }]);
  return ServiceWorkerRegistrar;
}(_react.default.Component);

exports.ServiceWorkerRegistrar = ServiceWorkerRegistrar;
ServiceWorkerRegistrar.propTypes = {
  scriptUrl: _propTypes.default.string,
  options: /*#__PURE__*/_propTypes.default.shape({
    scope: _propTypes.default.string,
    type: /*#__PURE__*/_propTypes.default.oneOf(['classic', 'module']),
    updateViaCache: /*#__PURE__*/_propTypes.default.oneOf(['imports', 'all', 'none'])
  })
};
ServiceWorkerRegistrar.defaultProps = {
  scriptUrl: '/sw.js'
};
