// @deity/falcon-ui/src/theme/responsiveprops.ts
import CSS from 'csstype';
import { Theme, CSSObject } from '.';

export type ResponsivePropMapping = {
  cssProp?: keyof CSS.Properties;
  themeProp?: keyof Theme;
  transformToCss?: (value: number | string) => CSSObject;
};

type PropsMapping = {
  [name: string]: ResponsivePropMapping;
};

function propsMapping<T extends PropsMapping>(param: T) {
  return param;
}

export const mappings = propsMapping({
  m: {
    cssProp: 'margin',
    themeProp: 'spacing'
  },
  mt: {
    cssProp: 'marginTop',
    themeProp: 'spacing'
  },
  ml: {
    cssProp: 'marginLeft',
    themeProp: 'spacing'
  },
  mr: {
    cssProp: 'marginRight',
    themeProp: 'spacing'
  },
  mb: {
    cssProp: 'marginBottom',
    themeProp: 'spacing'
  },
  mx: {
    themeProp: 'spacing',
    transformToCss: value => ({
      marginLeft: value,
      marginRight: value
    })
  },
  my: {
    themeProp: 'spacing',
    transformToCss: value => ({
      marginTop: value,
      marginBottom: value
    })
  },

  p: {
    cssProp: 'padding',
    themeProp: 'spacing'
  },
  pt: {
    cssProp: 'paddingTop',
    themeProp: 'spacing'
  },
  pl: {
    cssProp: 'paddingLeft',
    themeProp: 'spacing'
  },
  pr: {
    cssProp: 'paddingRight',
    themeProp: 'spacing'
  },
  pb: {
    cssProp: 'paddingBottom',
    themeProp: 'spacing'
  },
  px: {
    themeProp: 'spacing',
    transformToCss: value => ({
      paddingLeft: value,
      paddingRight: value
    })
  },
  py: {
    themeProp: 'spacing',
    transformToCss: value => ({
      paddingTop: value,
      paddingBottom: value
    })
  },

  height: {
    themeProp: 'sizes'
  },
  minHeight: {
    themeProp: 'sizes'
  },
  maxHeight: {
    themeProp: 'sizes'
  },

  width: {
    themeProp: 'sizes'
  },
  minWidth: {
    themeProp: 'sizes'
  },
  maxWidth: {
    themeProp: 'sizes'
  },

  size: {
    themeProp: 'sizes',
    transformToCss: value => ({
      height: value,
      width: value
    })
  },
  minSize: {
    themeProp: 'sizes',
    transformToCss: value => ({
      minHeight: value,
      minWidth: value
    })
  },
  maxSize: {
    themeProp: 'sizes',
    transformToCss: value => ({
      maxHeight: value,
      maxWidth: value
    })
  },

  gridGap: {
    themeProp: 'spacing'
  },
  gridRowGap: {
    themeProp: 'spacing'
  },
  gridColumnGap: {
    themeProp: 'spacing'
  },

  color: {
    cssProp: 'color',
    themeProp: 'colors'
  },
  bg: {
    cssProp: 'backgroundColor',
    themeProp: 'colors'
  },
  fill: {
    themeProp: 'colors'
  },
  stroke: {
    themeProp: 'colors'
  },

  borderColor: {
    themeProp: 'colors'
  },

  bgFullWidth: {
    themeProp: 'colors',
    transformToCss: value => ({
      position: 'relative',
      zIndex: 1,
      ':before': {
        content: '""',
        width: '200vw',
        height: '100%',
        background: value,
        position: 'absolute',
        left: '-50vw',
        right: '50vw',
        top: 0,
        zIndex: -1
      }
    })
  },

  fontSize: {
    themeProp: 'fontSizes'
  },
  fontFamily: {
    themeProp: 'fonts'
  },

  lineHeight: {
    themeProp: 'lineHeights'
  },
  fontWeight: {
    themeProp: 'fontWeights'
  },
  letterSpacing: {
    themeProp: 'letterSpacings'
  },

  border: {
    themeProp: 'borders'
  },
  borderTop: {
    themeProp: 'borders'
  },
  borderRight: {
    themeProp: 'borders'
  },
  borderBottom: {
    themeProp: 'borders'
  },
  borderLeft: {
    themeProp: 'borders'
  },

  borderRadius: {
    themeProp: 'borderRadius'
  },

  boxShadow: {
    themeProp: 'boxShadows'
  },

  position: {},

  transitionTimingFunction: {
    themeProp: 'easingFunctions'
  },

  transitionDuration: {
    themeProp: 'transitionDurations'
  },

  opacity: {
    themeProp: 'opacities'
  },

  textDecoration: {
    themeProp: 'textDecorations'
  },

  top: {},
  right: {},
  bottom: {},
  left: {},

  display: {},

  visibility: {},

  alignItems: {},
  justifyContent: {},
  flexWrap: {},
  flexDirection: {},
  flex: {},
  alignContent: {},
  justifySelf: {},
  justifyItems: {},
  textAlign: {},
  alignSelf: {},
  order: {},
  flexBasis: {},

  gridColumn: {},
  gridRow: {},
  gridAutoFlow: {},
  gridAutoRows: {},
  gridAutoColumns: {},
  gridTemplateRows: {},
  gridTemplateColumns: {},
  gridTemplateAreas: {},
  gridArea: {},
  gridTemplate: {},

  overflow: {},
  overflowX: {},
  overflowY: {},

  cursor: {}
});

export const themablePropsKeys = [...Object.keys(mappings), 'css'];

export type PropsMappings = typeof mappings;
export const propsMappingKeys = Object.keys(mappings) as (keyof PropsMappings)[];
