// @deity/falcon-ui/dist/components/Details.js
import { themed } from '../theme';
export var Details = /*#__PURE__*/themed({
  tag: 'details',
  defaultProps: {
    open: false
  },
  defaultTheme: {
    details: {
      display: 'flex',
      flexDirection: 'column',
      css: function css(_ref) {
        var theme = _ref.theme,
            open = _ref.open;
        return {
          '> :not(summary, style)': {
            display: open ? 'block' : 'none',
            flex: open ? '1' : 0
          },
          '> summary::-webkit-details-marker': {
            display: 'none'
          },
          '> summary:after': {
            display: 'block',
            content: open ? '"-"' : '"+"',
            marginLeft: theme.spacing.sm,
            fontSize: theme.fontSizes.md,
            lineHeight: 0.6,
            fontWeight: theme.fontWeights.bold,
            color: theme.colors.secondaryText
          }
        };
      }
    }
  }
});
export var Summary = /*#__PURE__*/themed({
  tag: 'summary',
  defaultTheme: {
    summary: {
      fontSize: 'sm',
      py: 'xs',
      px: 'sm',
      mb: 'xs',
      bg: 'secondaryLight',
      lineHeight: 'small',
      borderRadius: 'medium',
      css: {
        userSelect: 'none',
        cursor: 'pointer',
        display: 'flex',
        justifyContent: 'space-between',
        position: 'relative'
      }
    }
  }
});
export var DetailsContent = /*#__PURE__*/themed({
  tag: 'article',
  defaultTheme: {
    detailsContent: {
      py: 'xs',
      pl: 'xs'
    }
  }
});
