// @deity/falcon-ui/dist/components/Navbar.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/esm/objectWithoutPropertiesLoose";
import React, { useState } from 'react';
import { themed } from '../theme';
import { Box } from './Box';
import { ListItem } from './List';
export var Navbar = /*#__PURE__*/themed({
  tag: 'ul',
  defaultTheme: {
    navbar: {
      p: 'none',
      m: 'none',
      bgFullWidth: 'primary',
      color: 'primaryText',
      css: {
        display: 'flex',
        listStyle: 'none',
        position: 'relative'
      }
    }
  }
});
var MenuItemContext = /*#__PURE__*/React.createContext({});

var MenuItemInnerDOM = function MenuItemInnerDOM(_ref) {
  var _onMouseEnter = _ref.onMouseEnter,
      _onMouseLeave = _ref.onMouseLeave,
      _onClick = _ref.onClick,
      props = _objectWithoutPropertiesLoose(_ref, ["onMouseEnter", "onMouseLeave", "onClick"]);

  var _useState = useState(false),
      isOpen = _useState[0],
      setIsOpen = _useState[1];

  return /*#__PURE__*/React.createElement(MenuItemContext.Provider, {
    value: {
      isOpen: isOpen
    }
  }, /*#__PURE__*/React.createElement(ListItem, _extends({}, props, {
    onMouseEnter: function onMouseEnter(e) {
      if (_onMouseEnter) _onMouseEnter(e);
      setIsOpen(true);
    },
    onMouseLeave: function onMouseLeave(e) {
      setIsOpen(false);
      if (_onMouseLeave) _onMouseLeave(e);
    },
    onClick: function onClick(e) {
      setIsOpen(false);
      if (_onClick) _onClick(e);
    }
  })));
};

export var NavbarItem = /*#__PURE__*/themed({
  tag: MenuItemInnerDOM,
  defaultTheme: {
    navbarItem: {
      fontSize: 'sm',
      display: 'flex',
      color: 'primaryText',
      css: function css(_ref2) {
        var theme = _ref2.theme;
        return {
          cursor: 'pointer',
          userSelect: 'none',
          listStyle: 'none',
          ':hover': {
            background: theme.colors.secondary,
            color: theme.colors.secondaryText
          }
        };
      }
    }
  }
});

var NavbarItemMenuInnerDOM = function NavbarItemMenuInnerDOM(props) {
  return /*#__PURE__*/React.createElement(MenuItemContext.Consumer, null, function (_ref3) {
    var isOpen = _ref3.isOpen;
    return /*#__PURE__*/React.createElement(Box, _extends({}, props, {
      display: isOpen ? 'block' : 'none'
    }));
  });
};

export var NavbarItemMenu = /*#__PURE__*/themed({
  tag: NavbarItemMenuInnerDOM,
  defaultProps: {
    role: 'menu'
  },
  defaultTheme: {
    navbarItemMenu: {
      p: 'sm',
      bg: 'secondary',
      color: 'secondaryText',
      boxShadow: 'subtle',
      css: {
        position: 'absolute',
        top: '100%',
        left: 0,
        right: 0
      }
    }
  }
});
