// @deity/falcon-ui/dist/components/Radio.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/esm/objectWithoutPropertiesLoose";
import React from 'react';
import { themed, extractThemableProps } from '../theme';
import { Box } from './Box';
import { Icon } from './Icon';

var RadioInnerDOM = function RadioInnerDOM(props) {
  var className = props.className,
      icon = props.icon,
      invalid = props.invalid,
      remaining = _objectWithoutPropertiesLoose(props, ["className", "icon", "invalid"]);

  var _extractThemableProps = extractThemableProps(remaining),
      themableProps = _extractThemableProps.themableProps,
      rest = _extractThemableProps.rest;

  return /*#__PURE__*/React.createElement(Box, _extends({}, themableProps, {
    className: className
  }), /*#__PURE__*/React.createElement("input", _extends({}, rest, {
    type: "radio"
  })), /*#__PURE__*/React.createElement("div", {
    className: "-inner-radio-frame"
  }, icon));
};

var iconFallback = /*#__PURE__*/React.createElement("svg", {
  viewBox: "0 0 24 24",
  className: "-inner-radio-icon",
  focusable: "false"
}, /*#__PURE__*/React.createElement("circle", {
  cx: "12",
  cy: "12",
  r: "10"
}));
export var Radio = /*#__PURE__*/themed({
  tag: RadioInnerDOM,
  defaultProps: {
    invalid: false,
    icon: /*#__PURE__*/React.createElement(Icon, {
      className: "-inner-radio-icon",
      src: "radioCheckedIcon",
      fallback: iconFallback
    })
  },
  defaultTheme: {
    radio: {
      size: 'md',
      css: function css(_ref) {
        var theme = _ref.theme;
        return {
          display: 'inline-flex',
          position: 'relative',
          cursor: 'default',
          ':focus-within': {
            outline: "1px solid " + theme.colors.focus
          },
          // radio input is not visible but interactive
          input: {
            position: 'absolute',
            top: 0,
            left: 0,
            width: '100%',
            height: '100%',
            margin: 0,
            opacity: 0,
            zIndex: 1,
            cursor: 'inherit',
            ':checked + .-inner-radio-frame': {
              borderColor: theme.colors.primary,
              '.-inner-radio-icon': {
                opacity: 1,
                fill: theme.colors.primary
              }
            },
            ':hover + .-inner-radio-frame': {
              borderColor: theme.colors.primaryLight,
              '.-inner-radio-icon': {
                fill: theme.colors.primaryLight
              }
            }
          },
          '.-inner-radio-icon': {
            height: 'calc(100% - 4px)',
            width: 'calc(100% - 4px)',
            display: 'block',
            opacity: 0,
            stroke: 'none',
            fill: theme.colors.white,
            transitionProperty: 'opacity, fill',
            transitionTimingFunction: theme.easingFunctions.easeIn,
            transitionDuration: theme.transitionDurations.short
          },
          '.-inner-radio-frame': {
            height: '100%',
            width: '100%',
            position: 'relative',
            display: 'flex',
            cursor: 'pointer',
            borderRadius: theme.borderRadius.round,
            border: theme.borders.bold,
            borderColor: theme.colors.secondaryDark,
            transitionProperty: 'border, fill',
            transitionTimingFunction: theme.easingFunctions.easeIn,
            transitionDuration: theme.transitionDurations.short,
            justifyContent: 'center',
            alignItems: 'center'
          }
        };
      }
    }
  }
});
