// @deity/falcon-ui/dist/components/Sidebar.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/esm/objectWithoutPropertiesLoose";
import React, { useEffect } from 'react';
import { themed } from '../theme';
import { withTheme } from './ThemeProvider';
var SidebarInnerDOM = /*#__PURE__*/withTheme(function (_ref) {
  var theme = _ref.theme,
      visible = _ref.visible,
      side = _ref.side,
      props = _objectWithoutPropertiesLoose(_ref, ["theme", "visible", "side"]);

  var sidebarRef = React.useRef();
  useEffect(function () {
    var sidebar = sidebarRef.current || undefined;

    if (sidebar) {
      var transitionDuration = theme.components.sidebar.transitionDuration || theme.transitionDurations.short; // Remove Animation && Place Sidebar Offscreen

      sidebar.style.transitionDuration = '0s';
      sidebar.style.transform = "translateX(" + (side === 'left' ? '-100%' : '100%') + ")"; // Set Animation -> change the position which will trigger the animation

      requestAnimationFrame(function () {
        sidebar.style.transitionDuration = transitionDuration;
        sidebar.style.transform = 'translateX(0)';
      }); // Reset properties to the provided theme by setting them to empty strings

      requestAnimationFrame(function () {
        sidebar.style.transitionDuration = '';
        sidebar.style.transform = '';
      });
    }
  }, [side, theme.components.sidebar.transitionDuration, theme.transitionDurations.short]);
  return /*#__PURE__*/React.createElement("div", _extends({
    ref: sidebarRef
  }, props));
});
export var Sidebar = /*#__PURE__*/themed({
  tag: SidebarInnerDOM,
  defaultProps: {
    visible: false,
    side: 'left'
  },
  defaultTheme: {
    sidebar: {
      overflowX: 'hidden',
      display: 'flex',
      bg: 'white',
      position: 'fixed',
      css: function css(_ref2) {
        var _ref3;

        var visible = _ref2.visible,
            theme = _ref2.theme,
            side = _ref2.side;
        return _ref3 = {
          top: 0,
          bottom: 0
        }, _ref3[side] = 0, _ref3.height = '100%', _ref3.zIndex = theme.zIndex.sidebar, _ref3.transform = visible ? 'translateX(0)' : side === 'left' ? 'translateX(-100%)' : 'translateX(100%)', _ref3.transitionProperty = 'transform', _ref3.transitionDuration = visible ? theme.transitionDurations.short : theme.transitionDurations.standard, _ref3.transitionTimingFunction = visible ? theme.easingFunctions.easeIn : theme.easingFunctions.easeOut, _ref3;
      }
    }
  }
});
