// @deity/falcon-ui/dist/components/Table.js
import { themed } from '../theme';
export var Table = /*#__PURE__*/themed({
  tag: 'table',
  defaultTheme: {
    table: {
      borderRadius: 'medium',
      fontSize: 'sm',
      display: {
        xs: 'block',
        sm: 'table'
      },
      css: {
        width: '100%',
        overflowY: {
          xs: 'hidden',
          md: 'initial'
        },
        tableLayout: 'auto',
        borderSpacing: 0,
        borderCollapse: 'collapse',
        borderStyle: 'hidden'
      }
    }
  }
});
export var Thead = /*#__PURE__*/themed({
  tag: 'thead',
  defaultTheme: {
    thead: {
      bg: 'secondary'
    }
  }
});
export var Th = /*#__PURE__*/themed({
  tag: 'th',
  defaultTheme: {
    th: {
      py: 'xs',
      px: 'sm',
      fontWeight: 'regular',
      css: {
        textAlign: 'left'
      }
    }
  }
});
export var Td = /*#__PURE__*/themed({
  tag: 'td',
  defaultTheme: {
    td: {
      p: 'sm',
      fontWeight: 'light',
      lineHeight: 'large',
      css: {
        textAlign: 'left'
      }
    }
  }
});
export var Tr = /*#__PURE__*/themed({
  tag: 'tr',
  defaultTheme: {
    tr: {
      display: 'table-row',
      borderTop: 'regular',
      borderColor: 'secondary'
    }
  }
});
export var Tbody = /*#__PURE__*/themed({
  tag: 'tbody',
  defaultTheme: {
    tbody: {
      fontSize: 'sm'
    }
  }
});
