// @deity/falcon-ui/dist/components/Tabs.js
import { themed } from '../theme';
export var Tabs = /*#__PURE__*/themed({
  tag: 'div',
  defaultProps: {
    role: 'tablist'
  },
  defaultTheme: {
    tabs: {
      display: 'flex',
      p: 'xs',
      m: 'none',
      bg: 'primary',
      color: 'primaryText',
      boxShadow: 'subtle',
      css: {
        listStyle: 'none'
      },
      variants: {
        secondary: {
          boxShadow: 'none',
          color: 'black',
          bg: 'transparent'
        }
      }
    }
  }
});
export var Tab = /*#__PURE__*/themed({
  tag: 'button',
  defaultProps: {
    role: 'tab',
    type: 'button',
    active: false
  },
  defaultTheme: {
    tab: {
      fontSize: 'sm',
      flex: 0,
      mr: 'md',
      px: 'md',
      py: 'xs',
      borderBottom: 'regular',
      css: function css(props) {
        return {
          // Reset button styling
          background: 'none',
          color: 'inherit',
          borderTop: 'none',
          borderLeft: 'none',
          borderRight: 'none',
          font: 'inherit',
          outline: 'inherit',
          // Additional styling
          transitionProperty: 'border-bottom-color',
          transitionTimingFunction: props.theme.easingFunctions.easeIn,
          transitionDuration: props.theme.transitionDurations.short,
          borderBottomColor: props.active ? props.theme.colors.secondary : props.theme.colors.primary,
          cursor: props.active ? 'default' : 'pointer'
        };
      },
      variants: {
        secondary: {
          borderBottom: 'bold',
          css: function css(props) {
            return {
              borderBottomColor: props.active ? props.theme.colors.primary : props.theme.colors.secondary
            };
          }
        }
      }
    }
  }
});
