// @deity/falcon-ui/dist/components/ThemeProvider.js
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/esm/objectWithoutPropertiesLoose";
import React from 'react';
import { ThemeProvider as Provider } from 'emotion-theming';
import { Global, CacheProvider } from '@emotion/core';
import createCache from '@emotion/cache';
import { createTheme } from '../theme';
import { Root } from './Root'; // re-export withTheme from emotion so theme can be accessed from code

export { withTheme } from 'emotion-theming'; // IMPORTANT: those styles get injected as global styles
// every other reset style can be applied on Root component but not body margin

var tinyNormalizeStyles = {
  body: {
    margin: 0
  }
}; // IMPORTANT: temporary hack/workaround for not rendering styles
// when using emotion together with react-apollo after error on the server occured
// falcon-ui ThemeProvider now renders CacheProvider with always new cache provided when running on the server
// analogous to https://github.com/emotion-js/emotion/blob/master/packages/core/src/context.js#L54

var isServer = typeof document === 'undefined';
var cache = /*#__PURE__*/createCache();
export var ThemeProvider = function ThemeProvider(_ref) {
  var _ref$theme = _ref.theme,
      theme = _ref$theme === void 0 ? createTheme() : _ref$theme,
      _ref$globalCss = _ref.globalCss,
      globalCss = _ref$globalCss === void 0 ? tinyNormalizeStyles : _ref$globalCss,
      _ref$withoutRoot = _ref.withoutRoot,
      withoutRoot = _ref$withoutRoot === void 0 ? false : _ref$withoutRoot,
      rest = _objectWithoutPropertiesLoose(_ref, ["theme", "globalCss", "withoutRoot"]);

  return /*#__PURE__*/React.createElement(CacheProvider, {
    value: isServer ? createCache() : cache
  }, /*#__PURE__*/React.createElement(Provider, {
    theme: theme
  }, /*#__PURE__*/React.createElement(Global, {
    styles: globalCss
  }), withoutRoot ? rest.children : /*#__PURE__*/React.createElement(Root, rest)));
};
