// @deity/falcon-front-kit/dist/DynamicRoute/DynamicRoute.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.DynamicRoute = void 0;

var _react = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("react"));

var _propTypes = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("prop-types"));

var _reactRouterDom = /*#__PURE__*/require("react-router-dom");

var _falconData = /*#__PURE__*/require("@deity/falcon-data");

var _Router = /*#__PURE__*/require("../Router");

var DynamicRoute = function DynamicRoute(props) {
  var components = props.components,
      NotFound = props.notFound;
  return /*#__PURE__*/_react.default.createElement(_Router.Router, null, function (router) {
    var location = props.location || router.location;
    var path = location.pathname;
    return /*#__PURE__*/_react.default.createElement(_falconData.UrlQuery, {
      variables: {
        path: path
      }
    }, function (_ref) {
      var url = _ref.data.url;

      if (!url) {
        return /*#__PURE__*/_react.default.createElement(NotFound, {
          location: location
        });
      }

      if (!('type' in url)) {
        return /*#__PURE__*/_react.default.createElement(_reactRouterDom.Redirect, {
          to: url.path
        });
      }

      var Component = components[url.type];

      if (!Component) {
        throw new Error("[DynamicRoute]: Please register component for '" + url.type + "' content type!");
      }

      return /*#__PURE__*/_react.default.createElement(Component, url);
    });
  });
};

exports.DynamicRoute = DynamicRoute;
DynamicRoute.propTypes = {
  location: _propTypes.default.any,
  components: /*#__PURE__*/_propTypes.default.objectOf( /*#__PURE__*/_propTypes.default.oneOfType([_propTypes.default.object, _propTypes.default.func]).isRequired).isRequired,
  notFound: _propTypes.default.func.isRequired
};
