// @deity/falcon-front-kit/dist/FalconScope/FalconScopeProvider.js
"use strict";

var _interopRequireWildcard = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireWildcard");

exports.__esModule = true;
exports.FalconScope = exports.useFalconScope = exports.FalconScopeProvider = void 0;

var _url = /*#__PURE__*/require("url");

var _react = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("react"));

var _reactCookie = /*#__PURE__*/require("react-cookie");

var _React = /*#__PURE__*/require("../React");

var _GeoLocation = /*#__PURE__*/require("../GeoLocation");

var _FalconScopeContext = /*#__PURE__*/require("./FalconScopeContext");

var FalconScopeProvider = function FalconScopeProvider(props) {
  var available = props.available,
      currentId = props.currentId,
      defaultId = props.defaultId,
      basename = props.pathBasename,
      set = props.set,
      children = props.children;
  var isMounted = (0, _React.useMounted)();

  var _useCookies = (0, _reactCookie.useCookies)(),
      allCookies = _useCookies[0],
      setCookie = _useCookies[1],
      removeCookie = _useCookies[2];

  var getById = function getById(id) {
    return available.find(function (x) {
      return x.id === id;
    });
  };

  var _useState = (0, _react.useState)(getById(currentId)),
      current = _useState[0],
      setCurrent = _useState[1];

  var _useState2 = (0, _react.useState)(basename),
      pathBasename = _useState2[0],
      setPathBasename = _useState2[1];

  var geoLocation = (0, _GeoLocation.useGeoLocation)();

  var setNewScope = function setNewScope(id) {
    var newFalconScope = available.find(function (x) {
      return x.id === id;
    });

    if (newFalconScope === undefined || current.id === id) {
      return;
    }

    var location = (0, _url.parse)(window.location.href);
    set(id, {
      scopes: available,
      defaultId: defaultId
    }, {
      cookies: {
        all: allCookies,
        remove: removeCookie,
        set: setCookie
      },
      url: location,
      geoLocation: geoLocation
    });

    if (isMounted.current) {
      setCurrent(getById(id));
    }
  };

  return /*#__PURE__*/_react.default.createElement(_FalconScopeContext.FalconScopeContext.Provider, {
    value: {
      available: available,
      current: current,
      default: getById(defaultId),
      pathBasename: pathBasename,
      set: setNewScope
    }
  }, children);
};

exports.FalconScopeProvider = FalconScopeProvider;

var useFalconScope = function useFalconScope() {
  return (0, _react.useContext)(_FalconScopeContext.FalconScopeContext);
};

exports.useFalconScope = useFalconScope;

var FalconScope = function FalconScope(_ref) {
  var children = _ref.children;
  return children(Object.assign({}, useFalconScope()));
};

exports.FalconScope = FalconScope;
