// @deity/falcon-ui-kit/dist/Order/OrderListItemLayout.js
"use strict";

exports.__esModule = true;
exports.OrderListItemLayout = exports.OrderListItemArea = void 0;

var _falconUi = /*#__PURE__*/require("@deity/falcon-ui");

var _helpers = /*#__PURE__*/require("../helpers");

var OrderListItemArea = {
  id: 'id',
  createdAt: 'createdAt',
  shipTo: 'shipTo',
  grandTotal: 'grandTotal',
  status: 'status',
  actions: 'actions'
};
exports.OrderListItemArea = OrderListItemArea;
var OrderListItemLayout = /*#__PURE__*/(0, _falconUi.themed)({
  tag: _falconUi.Box,
  defaultTheme: {
    orderListItemLayout: {
      display: 'grid',
      gridGap: {
        sm: 'none',
        md: 'sm'
      },
      py: 'sm',
      borderBottom: 'regular',
      gridTemplate: {
        xs: /*#__PURE__*/(0, _helpers.toGridTemplate)([['1fr'], [OrderListItemArea.id], [OrderListItemArea.createdAt], [OrderListItemArea.shipTo], [OrderListItemArea.grandTotal], [OrderListItemArea.status], [OrderListItemArea.actions]]),
        md: /*#__PURE__*/(0, _helpers.toGridTemplate)([['1fr', '1fr', '2fr', '1fr', '1fr', '1fr'], [OrderListItemArea.id, OrderListItemArea.createdAt, OrderListItemArea.shipTo, OrderListItemArea.grandTotal, OrderListItemArea.status, OrderListItemArea.actions]])
      }
    }
  }
});
exports.OrderListItemLayout = OrderListItemLayout;
