// @deity/falcon-ui-kit/dist/Order/OrderLayout.js
"use strict";

exports.__esModule = true;
exports.OrderLayout = exports.OrderLayoutArea = void 0;

var _falconUi = /*#__PURE__*/require("@deity/falcon-ui");

var _helpers = /*#__PURE__*/require("../helpers");

var OrderLayoutArea = {
  status: 'status',
  items: 'items',
  summary: 'summary',
  divider: 'divider'
};
exports.OrderLayoutArea = OrderLayoutArea;
var OrderLayout = /*#__PURE__*/(0, _falconUi.themed)({
  tag: 'article',
  defaultTheme: {
    orderLayout: {
      display: 'grid',
      gridGap: 'md',
      // prettier-ignore
      gridTemplate: {
        xs: /*#__PURE__*/(0, _helpers.toGridTemplate)([['1fr'], [OrderLayoutArea.status], [OrderLayoutArea.items], [OrderLayoutArea.divider], [OrderLayoutArea.summary]]),
        md: /*#__PURE__*/(0, _helpers.toGridTemplate)([['2fr', '1px', '1fr'], [OrderLayoutArea.status, OrderLayoutArea.status, OrderLayoutArea.status], [OrderLayoutArea.items, OrderLayoutArea.divider, OrderLayoutArea.summary]])
      }
    }
  }
});
exports.OrderLayout = OrderLayout;
