// @deity/falcon-ui-kit/dist/Currency/CurrencyPicker.js
"use strict";

var _interopRequireWildcard = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.CurrencyPicker = exports.CurrencyPickerLayout = void 0;

var _extends2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("react"));

var _falconI18n = /*#__PURE__*/require("@deity/falcon-i18n");

var _falconFrontKit = /*#__PURE__*/require("@deity/falcon-front-kit");

var _falconUi = /*#__PURE__*/require("@deity/falcon-ui");

var _Forms = /*#__PURE__*/require("../Forms");

var CurrencyPickerLayout = /*#__PURE__*/(0, _falconUi.themed)({
  tag: _falconUi.Box,
  defaultTheme: {
    currencyPickerLayout: {
      css: {
        minWidth: 96
      }
    }
  }
});
exports.CurrencyPickerLayout = CurrencyPickerLayout;

var CurrencyPicker = function CurrencyPicker(_ref) {
  var value = _ref.value,
      options = _ref.options,
      _onChange = _ref.onChange,
      showCartWarning = _ref.showCartWarning,
      id = _ref.id,
      label = _ref.label,
      rest = (0, _objectWithoutPropertiesLoose2.default)(_ref, ["value", "options", "onChange", "showCartWarning", "id", "label"]);

  var _useLocale = (0, _falconFrontKit.useLocale)(),
      locale = _useLocale.locale,
      localeFallback = _useLocale.localeFallback;

  var formattedOptions = options.map(function (option) {
    return {
      optionValue: option,
      optionLabel: (0, _falconFrontKit.getCurrencyLabel)(option, locale || localeFallback)
    };
  });
  var selectOptions = (0, _react.useMemo)(function () {
    return formattedOptions.map(function (_ref2) {
      var optionValue = _ref2.optionValue,
          optionLabel = _ref2.optionLabel;
      return {
        value: optionValue,
        label: optionLabel
      };
    });
  }, [formattedOptions]);

  var _useI18n = (0, _falconI18n.useI18n)(),
      t = _useI18n.t;

  return /*#__PURE__*/_react.default.createElement(CurrencyPickerLayout, null, label && /*#__PURE__*/_react.default.createElement(_falconUi.Label, {
    htmlFor: id ? "currency-switcher-" + id : 'currency-switcher'
  }, label), /*#__PURE__*/_react.default.createElement(_Forms.SelectInput, (0, _extends2.default)({}, rest, {
    id: id ? "currency-switcher-" + id : 'currency-switcher',
    value: value,
    options: selectOptions,
    "aria-label": label || t('currency.label'),
    onChange: function onChange(e) {
      var confirmation = showCartWarning ? confirm(t('currency.changeNote')) : true;

      if (confirmation === true) {
        _onChange(e.target.value);
      }
    }
  })));
};

exports.CurrencyPicker = CurrencyPicker;
