// @deity/falcon-ui/src/theme/baseTheme.ts
export const baseTheme = {
  colors: {
    primary: '#A9CF38',
    primaryLight: '#CBDE6E',
    primaryDark: '#95B82C',
    primaryText: '#FFFFFF',

    secondary: '#EEF0F2',
    secondaryLight: '#F8F8F8',
    secondaryDark: '#E8E8E8',
    secondaryText: '#5F6367',

    error: '#E74C3C',
    errorText: '#000000',

    black: '#5f6367',
    white: '#FFFFFF',
    transparent: 'transparent',

    focus: '#005fcc'
  },

  breakpoints: {
    xs: 0,
    sm: 640,
    md: 860,
    lg: 1280,
    xl: 1920
  },

  spacing: {
    auto: 'auto',

    none: 0,
    xs: 8,
    sm: 16,
    md: 24,
    lg: 32,
    xl: 40,
    xxl: 48,
    xxxl: 64,

    full: '100%'
  },

  sizes: {
    auto: 'auto',

    none: 0,
    xs: 8,
    sm: 16,
    md: 24,
    lg: 32,
    xl: 40,
    xxl: 48,
    xxxl: 64,

    /** 12.5% */
    XXS: `${100 / 8}%`,
    /** 25.0% */
    XS: `${200 / 8}%`,
    /** 37.5% */
    SM: `${300 / 8}%`,
    /** 50.0% */
    MD: `${400 / 8}%`,
    /** 62.5% */
    LG: `${500 / 8}%`,
    /** 75.0% */
    XL: `${600 / 8}%`,
    /** 87.5% */
    XXL: `${700 / 8}%`,

    full: '100%'
  },

  fonts: {
    sans: '"Segoe UI", system-ui, sans-serif',
    mono: '"SF Mono", monospace'
  },

  fontSizes: {
    xxs: 12,
    xs: 14,
    sm: 16,
    md: 20,
    lg: 26,
    xl: 39,
    xxl: 48,
    xxxl: 96
  },

  fontWeights: {
    light: 300,
    regular: 400,
    bold: 700
  },

  lineHeights: {
    small: 1,
    default: 1.5,
    large: 2
  },

  letterSpacings: {
    normal: 'normal',
    caps: '0.025em'
  },

  borders: {
    none: 'none',
    regular: '1px solid',
    bold: '2px solid'
  },

  borderRadius: {
    none: 0,
    small: 2,
    medium: 4,
    large: 8,
    round: 333
  },

  boxShadows: {
    none: 'none',
    subtle: '0 5px 5px rgba(0,0,0,.1)',
    pronounced: '0 0 2px 0 rgba(0,0,0,.08),0 2px 8px 0 rgba(0,0,0,.16)',
    strong: '0 0 2px 0 rgba(0,0,0,.08),0 4px 16px 0 rgba(0,0,0,.16)',
    distant: '0 0 2px 0 rgba(0,0,0,.08),0 8px 32px 0 rgba(0,0,0,.16)'
  },

  easingFunctions: {
    easeInOut: 'cubic-bezier(0.4, 0, 0.2, 1)',
    easeOut: 'cubic-bezier(0.0, 0, 0.2, 1)',
    easeIn: 'cubic-bezier(0.4, 0, 1, 1)',
    sharp: 'cubic-bezier(0.4, 0, 0.6, 1)'
  },

  transitionDurations: {
    short: '150ms',
    standard: '250ms',
    long: '375ms'
  },

  keyframes: {},

  zIndex: {
    dropDownMenu: 600,
    backdrop: 800,
    sidebar: 1000,
    auto: 'auto'
  },

  opacities: {
    none: 0,
    sm: 0.25,
    md: 0.5,
    lg: 0.75,
    full: 1
  },

  textDecorations: {
    none: 'none',
    underline: 'underline',
    old: 'solid line-through'
  },

  components: {},

  icons: {}
};
