// @deity/falcon-ui-kit/dist/Error/Error.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.Error = void 0;

var _extends2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("react"));

var _falconUi = /*#__PURE__*/require("@deity/falcon-ui");

var errorInsights = function errorInsights(_ref) {
  var message = _ref.message,
      rest = (0, _objectWithoutPropertiesLoose2.default)(_ref, ["message"]);

  if (process.env.NODE_ENV !== 'production') {
    return JSON.stringify(Object.assign({}, rest), null, 2);
  }

  return '';
};

var Error = function Error(_ref2) {
  var children = _ref2.children,
      insights = _ref2.insights,
      rest = (0, _objectWithoutPropertiesLoose2.default)(_ref2, ["children", "insights"]);
  return /*#__PURE__*/_react.default.createElement(_falconUi.Text, (0, _extends2.default)({
    defaultTheme: {
      error: {
        color: 'error'
      }
    },
    title: errorInsights(insights)
  }, rest), children);
};

exports.Error = Error;
