// @deity/falcon-ui-kit/dist/FormattedDate/FormattedDate.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.FormattedDate = void 0;

var _objectWithoutPropertiesLoose2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("react"));

var _propTypes = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("prop-types"));

var _falconUi = /*#__PURE__*/require("@deity/falcon-ui");

var _falconFrontKit = /*#__PURE__*/require("@deity/falcon-front-kit");

var FormattedDateInnerDOM = function FormattedDateInnerDOM(_ref) {
  var value = _ref.value,
      formatOptions = _ref.formatOptions,
      rest = (0, _objectWithoutPropertiesLoose2.default)(_ref, ["value", "formatOptions"]);

  var _useLocale = (0, _falconFrontKit.useLocale)(),
      dateTimeFormat = _useLocale.dateTimeFormat;

  return /*#__PURE__*/_react.default.createElement(_falconUi.Text, rest, dateTimeFormat(value, formatOptions));
};

FormattedDateInnerDOM.propTypes = {
  value: /*#__PURE__*/_propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.string, /*#__PURE__*/_propTypes.default.instanceOf(Date)]).isRequired,
  formatOptions: _propTypes.default.any
};
var FormattedDate = /*#__PURE__*/(0, _falconUi.themed)({
  tag: FormattedDateInnerDOM,
  defaultProps: {
    formatOptions: undefined
  },
  defaultTheme: {
    formattedDate: {
      display: 'block',
      m: 'none'
    }
  }
});
exports.FormattedDate = FormattedDate;
