// @deity/falcon-ui-kit/dist/SEO/ProductSchema.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.ProductListSchema = exports.ProductSchema = void 0;

var _react = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("react"));

var _falconFrontKit = /*#__PURE__*/require("@deity/falcon-front-kit");

var _JsonLd = /*#__PURE__*/require("./JsonLd");

/**
 * Get the final price of a product
 * @param {*} price
 */
var getProductFinalPrice = function getProductFinalPrice(price) {
  var regular = price.regular,
      special = price.special;

  if (special > 0 && special < regular) {
    return special;
  }

  return regular;
};
/**
 * Get the image for a product
 * @param {*} product
 */


var getProductImage = function getProductImage(product) {
  var _product$gallery = product.gallery,
      gallery = _product$gallery === void 0 ? [] : _product$gallery,
      _product$thumbnail = product.thumbnail,
      thumbnail = _product$thumbnail === void 0 ? null : _product$thumbnail;

  if (gallery.length) {
    return gallery[0].thumbnail;
  }

  if (thumbnail) {
    return thumbnail;
  }

  return null;
};

/**
 * Get the schema data for a product.
 * @param product
 * @param fallbackCurrency currency to be used when product does not have one
 */
var getProductSchemaData = function getProductSchemaData(product, fallbackCurrency) {
  if (!product) {
    return null;
  }

  var schema = {
    '@context': 'http://schema.org',
    '@type': 'Product',
    name: product.name,
    image: getProductImage(product),
    sku: product.sku,
    description: product.description,
    offers: {
      '@type': 'Offer',
      priceCurrency: product.currency ? product.currency : fallbackCurrency,
      price: getProductFinalPrice(product.price),
      availability: 'http://schema.org/InStock'
    }
  };

  if (product.urlPath) {
    schema.url = product.urlPath;
  }

  return schema;
};

var ProductSchema = function ProductSchema(_ref) {
  var product = _ref.product;

  var _useCurrency = (0, _falconFrontKit.useCurrency)(),
      currency = _useCurrency.currency;

  var schemaData = getProductSchemaData(product, currency);
  return /*#__PURE__*/_react.default.createElement(_JsonLd.JsonLd, {
    item: schemaData
  });
};

exports.ProductSchema = ProductSchema;

var ProductListSchema = function ProductListSchema(_ref2) {
  var items = _ref2.items;

  var _useCurrency2 = (0, _falconFrontKit.useCurrency)(),
      currency = _useCurrency2.currency;

  if (!items.length) {
    return null;
  }

  var products = items.map(function (item) {
    return getProductSchemaData(item, currency);
  });
  var schemaData = {
    '@context': 'http://schema.org',
    '@graph': products
  };
  return /*#__PURE__*/_react.default.createElement(_JsonLd.JsonLd, {
    item: schemaData
  });
};

exports.ProductListSchema = ProductListSchema;
