// @deity/falcon-ui/dist/theme/themed.js
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/esm/objectWithoutPropertiesLoose";
import React from 'react';
import styled from '@emotion/styled-base';
import isPropValid from '@emotion/is-prop-valid';
import { extractThemableProps } from './utils';
import { baseTheme } from './baseTheme';
import { propsMappingKeys, mappings } from './responsiveprops';
import { getThemeBreakpointNames } from './getThemeBreakpoints';
import { extractBreakpointBlockThemingProps, filterOutNonBreakpointBlockThemingProps, mapBreakpointBlockThemingPropsToBaseThemmingProps } from './formatBreakpointCssBlocks';
var NESTED_CSS_OBJECT_SELECTORS = [':', '&', '*', '>', '@'];

var convertPropToCss = function convertPropToCss(mappingKey, propMapping, matchingProp, theme) {
  var _ref;

  // if mapping does not have cssProp specified fallback to it's key as css property name
  var cssPropName = propMapping.cssProp || mappingKey; // if matching props is themable prop then get it's actual value from theme props otherwise
  // then just pass it as css prop value
  // TODO: typescript: is there a way to improve those typings?

  var cssPropValue = !propMapping.themeProp ? matchingProp : theme[propMapping.themeProp][matchingProp];

  if (propMapping.transformToCss) {
    return propMapping.transformToCss(cssPropValue);
  }

  return _ref = {}, _ref[cssPropName] = cssPropValue, _ref;
};

function convertThemedPropsToCss(props, theme) {
  //  if theme is not provided via theme provider do not map anything
  if (!theme) {
    return {};
  } // TODO: typescript: can typings be improved for that object?


  var targetCss = {}; // eslint-disable-next-line

  for (var mappingKey in props) {
    var propMapping = mappings[mappingKey];
    var matchingProp = props[mappingKey]; // move along if there is no matching prop in mappings for given key found

    if (!propMapping) {
      continue;
    } // if matching prop is typeof string it means it's not responsive


    if (typeof matchingProp === 'string' || typeof matchingProp === 'number') {
      var cssObject = convertPropToCss(mappingKey, propMapping, matchingProp, theme); // eslint-disable-next-line

      for (var _mappingKey in cssObject) {
        targetCss[_mappingKey] = cssObject[_mappingKey];
      }
    } else {
      // if it's not string it needs to be object that has responsive breakpoints keys
      // here we only translate all themed values to css values, we don't create media queries
      // eslint-disable-next-line
      for (var breakpointKey in matchingProp) {
        var breakpointValue = theme.breakpoints[breakpointKey];
        var matchingResponsiveProp = matchingProp[breakpointKey];

        if (breakpointValue === undefined) {
          continue;
        }

        var _cssObject = convertPropToCss(mappingKey, propMapping, matchingResponsiveProp, theme); // eslint-disable-next-line


        for (var _mappingKey2 in _cssObject) {
          if (!targetCss[_mappingKey2]) {
            targetCss[_mappingKey2] = {};
          }

          targetCss[_mappingKey2][breakpointKey] = _cssObject[_mappingKey2];
        }
      }
    }
  }

  return targetCss;
}

function convertResponsivePropsToMediaQueries(css, theme) {
  var target = {};
  var mediaQueries = {}; // eslint-disable-next-line

  for (var cssProp in css) {
    var cssValue = css[cssProp];

    if (!cssValue || typeof cssValue !== 'object' || Array.isArray(cssValue)) {
      target[cssProp] = cssValue;
    } // we need to look for responsive props in nested css as well
    // for example in :hover object
    else if (NESTED_CSS_OBJECT_SELECTORS.indexOf(cssProp[0]) !== -1) {
        target[cssProp] = convertResponsivePropsToMediaQueries(cssValue, theme);
      } else {
        // eslint-disable-next-line
        for (var potentialBreakpointKey in cssValue) {
          var breakpointValue = theme.breakpoints[potentialBreakpointKey];
          var valueForBreakpoint = cssValue[potentialBreakpointKey];

          if (breakpointValue) {
            // add media query key to mediaQueries object if it hasn't already got one
            if (!mediaQueries[potentialBreakpointKey]) {
              mediaQueries[potentialBreakpointKey] = {};
            }

            mediaQueries[potentialBreakpointKey][cssProp] = valueForBreakpoint;
          } else if (breakpointValue === 0) {
            target[cssProp] = valueForBreakpoint;
          } else {
            if (!target[cssProp]) {
              target[cssProp] = {};
            }

            target[cssProp][potentialBreakpointKey] = valueForBreakpoint;
          }
        }
      }
  } // media queries need to be handled in very careful way as order matters
  // so media min-width with smaller px value always apper before media min-width with larger px value
  // in resulting style


  Object.keys(mediaQueries).sort(function (first, second) {
    return theme.breakpoints[first] > theme.breakpoints[second] ? 1 : -1;
  }).forEach(function (sortedMediaQueryKey) {
    var mediaQueryPxValue = theme.breakpoints[sortedMediaQueryKey];
    target["@media screen and (min-width: " + mediaQueryPxValue + "px)"] = mediaQueries[sortedMediaQueryKey];
  });
  return target;
}

function getCss(css, props) {
  return typeof css === 'function' ? css(props) : css;
}

/**
 * `<themedComponent>.defaultTheme` specifies it's `props` in nested object which key is used as `themeKey`
 *
 * TODO: write unit tests!
 * @param defaultTheme
 */
function extractDefaultThemeFromThemedComponent(defaultTheme) {
  if (!defaultTheme) {
    return undefined;
  } // default theme should have only one key, so we try to take first one, we also expect an `object` under it


  var potentialKey = Object.keys(defaultTheme)[0];
  var themeKey = typeof defaultTheme[potentialKey] === 'object' ? potentialKey : undefined;

  if (!themeKey) {
    return undefined;
  }

  return {
    key: themeKey,
    props: defaultTheme[themeKey]
  };
}
/**
 * @private
 * exported only because of unit test !!!
 *
 * following diagram shows theme props order (values in brackets)
 *
 * | theme                                            | themed component                                     | inline                          |
 * | ------------------------------------------------ | ---------------------------------------------------- | ------------------------------- |
 * | `[theme].components[component-theme-name]`       | `themed({defaultTheme:{[component-theme-name]:{}}})` | `<Box bg="primary"`             |
 * | **(2)** `component.breakpointBlockProps`         | **(1)** `defaultTheme.breakpointBlockProps`          | **(13)** `breakpointBlockProps` |
 * | **(6)** `component.baseThemingProps`             | **(5)** `defaultTheme.baseThemingProps`              | **(14)** `baseThemingProps`     |
 * | **(10)** `component.cssProp`                     | **(9)** `defaultTheme.cssProp`                       | **(15)** `cssProp`              |
 * | **(4)** `component.variant.breakpointBlockProps` | **(3)** `defaultTheme.variant.breakpointBlockProps`  |                                 |
 * | **(8)** `component.variant.baseThemingProps`     | **(7)** `defaultTheme.variant.baseThemingProps`      |                                 |
 * | **(12)** `component.variant.cssProp`             | **(11)** `defaultTheme.variant.cssProp`              |                                 |
 * @param props
 */


export function mergeThemingProps(props) {
  var defaultThemeProp = props.defaultTheme,
      theme = props.theme,
      variant = props.variant,
      inlineCss = props.css,
      restProps = _objectWithoutPropertiesLoose(props, ["defaultTheme", "theme", "variant", "css"]);

  var defaultTheme = extractDefaultThemeFromThemedComponent(defaultThemeProp);
  var breakpointNames = getThemeBreakpointNames(theme);
  var breakpointBlockThemingProps = extractBreakpointBlockThemingProps(breakpointNames, restProps);
  var remainingProps = filterOutNonBreakpointBlockThemingProps(breakpointNames, restProps);

  var _extractThemableProps = extractThemableProps(remainingProps),
      themableProps = _extractThemableProps.themableProps; // first we need to check where themed props and css props are defined and merge them
  // // css props need to merged separately as those do not need to be processed to extract css
  // Merging order
  // 1 -  props defined in defaultTheme props  as those are defaults
  // 2 -  props defined in theme.components for given `defaultTheme.key` as those are defaults
  // 3 -  props defined in defaultTheme variant prop if props.variant is defined
  // 4 -  props defined in theme.components[]variants if props.variant is defined
  // 5 -  props defined directly on component


  var breakpointBlockThemingPropsToMerge = [];
  var baseThemingPropsToMerge = [];
  var cssPropsToMerge = [];

  var addPropsToMerge = function addPropsToMerge(propsToMerge) {
    var css = propsToMerge.css,
        rest = _objectWithoutPropertiesLoose(propsToMerge, ["css"]);

    if (css) {
      cssPropsToMerge.push(getCss(css, props));
    } // Format breakpoint CSS blocks


    breakpointBlockThemingPropsToMerge.push(extractBreakpointBlockThemingProps(breakpointNames, rest));
    baseThemingPropsToMerge.push(filterOutNonBreakpointBlockThemingProps(breakpointNames, rest));
  }; //  start with props defined in `defaultTheme` prop as base


  if (defaultTheme !== undefined) {
    addPropsToMerge(defaultTheme.props); // if props are defined in `theme.components` object for `defaultTheme.key` merge them with default ones

    var areComponentPropsDefinedInTheme = theme.components[defaultTheme.key] !== undefined;

    if (areComponentPropsDefinedInTheme) {
      addPropsToMerge(theme.components[defaultTheme.key]);
    } // themed props can also be defined for component variant


    if (variant) {
      // check for `variant` prop defined in themed component `defaultTheme`
      var defaultThemeVariants = defaultTheme.props && defaultTheme.props.variants;

      if (defaultThemeVariants && defaultThemeVariants[variant]) {
        addPropsToMerge(defaultThemeVariants[variant]);
      } // check for `variant` prop defined in `theme.components` object


      var themeVariants = areComponentPropsDefinedInTheme && theme.components[defaultTheme.key].variants;

      if (themeVariants && themeVariants[variant]) {
        addPropsToMerge(themeVariants[variant]);
      }
    }
  }

  var mergedBreakpointBlockThemingProps = Object.assign.apply(Object, [{}].concat(breakpointBlockThemingPropsToMerge));
  var baseThemingPropsFromBreakpointBlockThemingProps = mapBreakpointBlockThemingPropsToBaseThemmingProps(breakpointNames, mergedBreakpointBlockThemingProps);
  var mergedBaseThemingProps = Object.assign.apply(Object, [{}].concat([baseThemingPropsFromBreakpointBlockThemingProps].concat(baseThemingPropsToMerge))); // merged themable props need to be converted to css before returning

  var cssFromThemedProps = convertThemedPropsToCss(mergedBaseThemingProps, theme);
  cssPropsToMerge.push(convertThemedPropsToCss(mapBreakpointBlockThemingPropsToBaseThemmingProps(breakpointNames, breakpointBlockThemingProps), theme)); // out of all component props extract themable ones, convert to css and add to merge

  cssPropsToMerge.push(convertThemedPropsToCss(themableProps, theme)); // as last step add for merging those `css` prop which defined directly on component

  if (inlineCss) {
    cssPropsToMerge.push(getCss(inlineCss, props));
  }

  var cssProps = Object.assign.apply(Object, [{}].concat(cssPropsToMerge)); // finally merge css from themed props with css from css props

  var mergedCss = Object.assign({}, cssFromThemedProps, cssProps);
  return mergedCss;
}
/**
 * this function responsibility is to extract css object from
 * both themed props (that use props values from theme) and css object/function props
 * TODO: perhaps this function could be written in prettier way?
 *  @param props
 */

function getThemedCss(props) {
  //  if theme is not provided via theme provider or inline theme prop
  // fall back to default theme
  if (!props.theme || !props.theme.components) {
    props = Object.assign({}, props, {
      theme: baseTheme
    });
  }

  var mergedCss = mergeThemingProps(props); // as a last step we need to check each css prop if it's value is responsive

  return convertResponsivePropsToMediaQueries(mergedCss, props.theme);
} // filtering which props to forward to next component is tricky
// and behaves differently if next component is html element, custom component
// or custom component whihch is themed component


var customPropsBlacklist = ['as', 'tag', 'variant', 'defaultTheme', 'css'];

var filterPropsToForward = function filterPropsToForward(baseComponent, props, ref) {
  var filteredProps = {};
  var isHtmlTag = typeof baseComponent === 'string'; // eslint-disable-next-line

  for (var key in props) {
    // when html tag is provided forward only valid html props to it
    if (isHtmlTag && !isPropValid(key)) continue; // if custom component is provided via `as` prop do not forward themable props to it (bg, color, m, p etc)
    // neighter forward any of the blacklisted props

    var themableProp = propsMappingKeys.indexOf(key) !== -1 || customPropsBlacklist.indexOf(key) !== -1;
    if (themableProp) continue;
    filteredProps[key] = props[key];
  }

  filteredProps.ref = ref;
  return filteredProps;
};

/**
 * Handles dynamic html tag rendering via and `as` prop as well as forwards `ref` and only allowed html props to DOM element
 */
var DynamicHtmlTag = /*#__PURE__*/React.forwardRef(function (props, ref) {
  var Base = props.as || props.tag;
  var nextProps = filterPropsToForward(Base, props, ref);
  return /*#__PURE__*/React.createElement(Base, nextProps);
});
export function themed(options) {
  var label = '';

  if (options.defaultTheme) {
    var componentKey = Object.keys(options.defaultTheme)[0];

    if (typeof options.defaultTheme[componentKey] === 'object') {
      label = "" + componentKey;
    }
  }

  var styledComponentWithThemeProps = styled(DynamicHtmlTag, {
    label: label,
    // label is transformed for displayName of styled component,
    // target inserted as css class in resulting element so this could potentially be used as a fallback
    // to style components via traditional css
    target: "themed" + (label ? "-" + label.replace(/([a-z])([A-Z])/g, '$1-$2').replace(/\s+/g, '-').toLowerCase() : '')
  })(getThemedCss); // default theme is also passed as part of default props

  styledComponentWithThemeProps.defaultProps = Object.assign({}, options.defaultProps, {
    defaultTheme: options.defaultTheme,
    tag: options.tag
  });
  return /*#__PURE__*/React.memo(styledComponentWithThemeProps);
}
