// @deity/falcon-ui/dist/theme/utils.js
import merge from 'deepmerge';
import isPlainObject from 'is-plain-object';
import { keyframes } from '@emotion/core';
import { themablePropsKeys } from './responsiveprops';
export function extractThemableProps(props) {
  var themableProps = {};
  var rest = {}; // eslint-disable-next-line

  for (var _key in props) {
    if (themablePropsKeys.indexOf(_key) !== -1) {
      themableProps[_key] = props[_key];
    } else {
      rest[_key] = props[_key];
    }
  }

  return {
    themableProps: themableProps,
    rest: rest
  };
}
/**
 * Determines if object is `@emotion`'s keyframe definition,
 * they sets `anim=1` to keyframes processed by `@emotion/core/keyframes` function
 * @param {*} object
 */

function isEmotionKeyframe(object) {
  return object && object.anim === 1;
}

export function mergeThemes(theme, themeOverride) {
  var newTheme = merge(theme, themeOverride, {
    isMergeableObject: function isMergeableObject(x) {
      return isPlainObject(x) && !isEmotionKeyframe(x);
    }
  }); // keyframes merging needs to be handled in very special way
  // each not yet processed keyframes definitions needs to be passed
  // to emotion's keyframes function in order to be consumed in styles

  if (newTheme.keyframes) {
    // eslint-disable-next-line
    for (var keyframeKey in newTheme.keyframes) {
      // only process keyframes not yet processed by emotion (anim property not et)
      if (!newTheme.keyframes[keyframeKey].anim) {
        // TODO remove casting to `any` when https://github.com/deity-io/falcon/issues/545 fixed
        newTheme.keyframes[keyframeKey] = keyframes(newTheme.keyframes[keyframeKey]);
      }
    }
  }

  return newTheme;
}
export function gridSizes(columnAmount, theme) {
  function roundToMaxTwoDecimals(num) {
    return Math.round((num + Number.EPSILON) * 100) / 100;
  }

  var generatedSteps = {};

  for (var step = 1; step < columnAmount; step++) {
    var _key2 = step + "/" + columnAmount;

    var unformattedValue = step / columnAmount * 100;
    var value = roundToMaxTwoDecimals(unformattedValue) + "%";
    generatedSteps[_key2] = value;
  }

  theme.sizes = Object.assign({}, theme.sizes, generatedSteps);
}
export function mapNegativeSpacings(theme) {
  var spacing = theme.spacing;
  if (!spacing || typeof spacing !== 'object' || Object.keys(spacing).length <= 0) return theme;
  var negativeSpacings = {}; // Foreach value which is a number create a negative one with a key of '-key'

  Object.keys(spacing).forEach(function (key) {
    var value = spacing[key];
    if (typeof value !== 'number' || value <= 0) return;
    negativeSpacings["-" + key] = value * -1;
  });
  return Object.assign({}, theme, {
    spacing: Object.assign({}, spacing, negativeSpacings)
  });
}
