// @deity/falcon-front-kit/dist/SEO/SEO.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.SEO = void 0;

var _react = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("react"));

var _propTypes = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("prop-types"));

var _reactHelmetAsync = /*#__PURE__*/require("react-helmet-async");

var SEO = function SEO(_ref) {
  var title = _ref.title,
      meta = _ref.meta;
  var metaTitle = meta.title,
      description = meta.description,
      keywords = meta.keywords;
  return /*#__PURE__*/_react.default.createElement(_reactHelmetAsync.Helmet, null, /*#__PURE__*/_react.default.createElement("title", null, metaTitle || title), description && /*#__PURE__*/_react.default.createElement("meta", {
    name: "description",
    content: description
  }), keywords && /*#__PURE__*/_react.default.createElement("meta", {
    name: "keywords",
    content: keywords
  }));
};

exports.SEO = SEO;
SEO.propTypes = {
  title: _propTypes.default.string.isRequired,
  meta: /*#__PURE__*/_propTypes.default.shape({
    title: _propTypes.default.string,
    description: _propTypes.default.string,
    keywords: _propTypes.default.string
  })
};
SEO.defaultProps = {
  meta: {}
};
