// @deity/falcon-front-kit/dist/GeoLocation/GeoLocationProvider.js
"use strict";

var _interopRequireWildcard = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireWildcard");

exports.__esModule = true;
exports.GeoLocation = exports.useGeoLocation = exports.GeoLocationProvider = void 0;

var _react = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("react"));

var _GeoLocationContext = /*#__PURE__*/require("./GeoLocationContext");

var GeoLocationProvider = function GeoLocationProvider(props) {
  var country = props.country,
      region = props.region,
      city = props.city,
      postCode = props.postCode,
      children = props.children;
  return /*#__PURE__*/_react.default.createElement(_GeoLocationContext.GeoLocationContext.Provider, {
    value: {
      country: country,
      region: region,
      city: city,
      postCode: postCode
    }
  }, children);
};

exports.GeoLocationProvider = GeoLocationProvider;

var useGeoLocation = function useGeoLocation() {
  return (0, _react.useContext)(_GeoLocationContext.GeoLocationContext);
};

exports.useGeoLocation = useGeoLocation;

var GeoLocation = function GeoLocation(_ref) {
  var children = _ref.children;
  return children(Object.assign({}, useGeoLocation()));
};

exports.GeoLocation = GeoLocation;
