// @deity/falcon-front-kit/dist/Checkout/SetPaymentMethod.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.SetPaymentMethod = exports.useSetPaymentMethod = void 0;

var _regenerator = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/asyncToGenerator"));

var _react = /*#__PURE__*/require("react");

var _falconShopData = /*#__PURE__*/require("@deity/falcon-shop-data");

var _CheckoutConsumer = /*#__PURE__*/require("./CheckoutConsumer");

var useSetPaymentMethod = function useSetPaymentMethod() {
  var _useCheckout = (0, _CheckoutConsumer.useCheckout)(),
      isLoading = _useCheckout.isLoading,
      setLoading = _useCheckout.setLoading,
      setPaymentMethod = _useCheckout.setPaymentMethod;

  var _useSetPaymentMethodM = (0, _falconShopData.useSetPaymentMethodMutation)(),
      mutation = _useSetPaymentMethodM[0],
      result = _useSetPaymentMethodM[1];

  if (result.loading !== isLoading) {
    setLoading(result.loading);
  }

  return [(0, _react.useCallback)( /*#__PURE__*/function () {
    var _ref = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee(input, options) {
      var response, data;
      return _regenerator.default.wrap(function _callee$(_context) {
        while (1) {
          switch (_context.prev = _context.next) {
            case 0:
              setLoading(true);
              _context.next = 3;
              return mutation(Object.assign({}, options, {
                variables: {
                  input: {
                    provider: input.provider,
                    method: input.method,
                    data: input.data
                  }
                }
              }));

            case 3:
              response = _context.sent;
              data = response.data;
              setLoading(false);
              if (data && data.setPaymentMethod) setPaymentMethod(data.setPaymentMethod);
              return _context.abrupt("return", response);

            case 8:
            case "end":
              return _context.stop();
          }
        }
      }, _callee);
    }));

    return function (_x, _x2) {
      return _ref.apply(this, arguments);
    };
  }(), [mutation, setLoading, setPaymentMethod]), result];
};

exports.useSetPaymentMethod = useSetPaymentMethod;

var SetPaymentMethod = function SetPaymentMethod(_ref2) {
  var children = _ref2.children;
  return children.apply(void 0, useSetPaymentMethod());
};

exports.SetPaymentMethod = SetPaymentMethod;
