// @deity/falcon-front-kit/dist/Checkout/SetShippingAddress.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.SetShippingAddress = exports.useSetShippingAddress = void 0;

var _regenerator = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/asyncToGenerator"));

var _react = /*#__PURE__*/require("react");

var _falconShopData = /*#__PURE__*/require("@deity/falcon-shop-data");

var _CheckoutConsumer = /*#__PURE__*/require("./CheckoutConsumer");

var _CheckoutAddress = /*#__PURE__*/require("./CheckoutAddress");

var useSetShippingAddress = function useSetShippingAddress() {
  var _useState = (0, _react.useState)(),
      address = _useState[0],
      setAddress = _useState[1];

  var _useCheckout = (0, _CheckoutConsumer.useCheckout)(),
      isLoading = _useCheckout.isLoading,
      setLoading = _useCheckout.setLoading,
      setShippingAddress = _useCheckout.setShippingAddress,
      isBillingSameAsShipping = _useCheckout.isBillingSameAsShipping;

  var _useSetShippingAddres = (0, _falconShopData.useSetShippingAddressMutation)({
    onCompleted: function onCompleted(data) {
      if (data) {
        setShippingAddress(address);
        setLoading(false);
      }
    }
  }),
      mutation = _useSetShippingAddres[0],
      result = _useSetShippingAddres[1];

  if (result.loading !== isLoading) {
    setLoading(result.loading);
  }

  return [(0, _react.useCallback)( /*#__PURE__*/function () {
    var _ref = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee(input, options) {
      return _regenerator.default.wrap(function _callee$(_context) {
        while (1) {
          switch (_context.prev = _context.next) {
            case 0:
              if (options === void 0) {
                options = {};
              }

              setLoading(true);
              setAddress(input);
              return _context.abrupt("return", mutation(Object.assign({}, options, {
                variables: {
                  input: {
                    address: (0, _CheckoutAddress.addressToCheckoutAddressInput)(input),
                    billingSameAsShipping: isBillingSameAsShipping
                  }
                }
              })));

            case 4:
            case "end":
              return _context.stop();
          }
        }
      }, _callee);
    }));

    return function (_x, _x2) {
      return _ref.apply(this, arguments);
    };
  }(), [isBillingSameAsShipping, mutation, setLoading]), result];
};

exports.useSetShippingAddress = useSetShippingAddress;

var SetShippingAddress = function SetShippingAddress(_ref2) {
  var children = _ref2.children;
  return children.apply(void 0, useSetShippingAddress());
};

exports.SetShippingAddress = SetShippingAddress;
