// @deity/falcon-errors/dist/errors.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.UserInputError = exports.EntityNotFoundError = exports.ForbiddenError = exports.AuthenticationError = exports.FalconError = void 0;

var _assertThisInitialized2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/assertThisInitialized"));

var _inheritsLoose2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/inheritsLoose"));

var _wrapNativeSuper2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/wrapNativeSuper"));

var _codes = /*#__PURE__*/require("./codes");

// FalconError is simplified version of ApolloError
// that does not import whole `graphql` package resulting in much smaller bundle size
// https://github.com/apollographql/apollo-server/blob/master/packages/apollo-server-errors/src/index.ts
var FalconError = /*#__PURE__*/function (_Error) {
  (0, _inheritsLoose2.default)(FalconError, _Error);

  function FalconError(message, code) {
    var _this;

    _this = _Error.call(this, message) || this; // if no name provided, use the default. defineProperty ensures that it stays non-enumerable

    if (!_this.name) {
      Object.defineProperty((0, _assertThisInitialized2.default)(_this), 'name', {
        value: 'FalconError'
      });
    }

    if (code) {
      Object.defineProperty((0, _assertThisInitialized2.default)(_this), 'extensions', {
        value: {
          code: code
        }
      });
    }

    return _this;
  }

  return FalconError;
}( /*#__PURE__*/(0, _wrapNativeSuper2.default)(Error));

exports.FalconError = FalconError;

var AuthenticationError = /*#__PURE__*/function (_FalconError) {
  (0, _inheritsLoose2.default)(AuthenticationError, _FalconError);

  /**
   * @param {string} message
   * @param {string | undefined} code optional error code, default is `UNAUTHENTICATED`
   */
  function AuthenticationError(message, code) {
    var _this2;

    _this2 = _FalconError.call(this, message, code || _codes.codes.UNAUTHENTICATED) || this;
    Object.defineProperty((0, _assertThisInitialized2.default)(_this2), 'name', {
      value: 'AuthenticationError'
    });
    return _this2;
  }

  return AuthenticationError;
}(FalconError);

exports.AuthenticationError = AuthenticationError;

var ForbiddenError = /*#__PURE__*/function (_FalconError2) {
  (0, _inheritsLoose2.default)(ForbiddenError, _FalconError2);

  /**
   * @param {string} [message]
   */
  function ForbiddenError(message) {
    var _this3;

    if (message === void 0) {
      message = 'Access is forbidden';
    }

    _this3 = _FalconError2.call(this, message, _codes.codes.FORBIDDEN) || this;
    Object.defineProperty((0, _assertThisInitialized2.default)(_this3), 'name', {
      value: 'ForbiddenError'
    });
    return _this3;
  }

  return ForbiddenError;
}(FalconError);

exports.ForbiddenError = ForbiddenError;

var EntityNotFoundError = /*#__PURE__*/function (_FalconError3) {
  (0, _inheritsLoose2.default)(EntityNotFoundError, _FalconError3);

  /**
   * @param {string} [message]
   */
  function EntityNotFoundError(message) {
    var _this4;

    if (message === void 0) {
      message = 'Entity not found';
    }

    _this4 = _FalconError3.call(this, message, _codes.codes.NOT_FOUND) || this;
    Object.defineProperty((0, _assertThisInitialized2.default)(_this4), 'name', {
      value: 'EntityNotFoundError'
    });
    return _this4;
  }

  return EntityNotFoundError;
}(FalconError);
/**
 * @example
 * ```javascript
 *  throw new UserInputError(e.message, {
 *    errors: {
 *      email: { code: 'email_already_taken', message: e.message }
 *    }
 *  });
 * ```
 */


exports.EntityNotFoundError = EntityNotFoundError;

var UserInputError = /*#__PURE__*/function (_FalconError4) {
  (0, _inheritsLoose2.default)(UserInputError, _FalconError4);

  /**
   * @param {string} message
   * @param {object} [properties]
   */
  function UserInputError(message, properties) {
    var _this5;

    _this5 = _FalconError4.call(this, message, _codes.codes.BAD_USER_INPUT, properties) || this;
    Object.defineProperty((0, _assertThisInitialized2.default)(_this5), 'name', {
      value: 'UserInputError'
    });
    return _this5;
  }

  return UserInputError;
}(FalconError);

exports.UserInputError = UserInputError;
