// @deity/falcon-ui-kit/dist/Account/ForgotPasswordForm.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.ForgotPasswordForm = void 0;

var _extends2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("react"));

var _falconI18n = /*#__PURE__*/require("@deity/falcon-i18n");

var _falconUi = /*#__PURE__*/require("@deity/falcon-ui");

var _falconFrontKit = /*#__PURE__*/require("@deity/falcon-front-kit");

var _Error = /*#__PURE__*/require("../Error");

var _Forms = /*#__PURE__*/require("../Forms");

var ForgotPasswordForm = function ForgotPasswordForm(_ref) {
  var onSuccess = _ref.onSuccess,
      formProps = (0, _objectWithoutPropertiesLoose2.default)(_ref, ["onSuccess"]);
  return /*#__PURE__*/_react.default.createElement(_falconFrontKit.ForgotPasswordFormProvider, {
    onSuccess: onSuccess
  }, function (_ref2) {
    var isSubmitting = _ref2.isSubmitting,
        status = _ref2.status,
        values = _ref2.values,
        submitCount = _ref2.submitCount,
        isValid = _ref2.isValid;
    var submitSucceed = isValid && !isSubmitting && submitCount > 0 && !status.error;
    return /*#__PURE__*/_react.default.createElement(_Forms.Form, (0, _extends2.default)({
      id: "forgot-password",
      i18nId: "forgotPassword"
    }, formProps), /*#__PURE__*/_react.default.createElement(_Forms.FormField, {
      name: "email",
      required: true,
      type: "email",
      autoComplete: "email"
    }), /*#__PURE__*/_react.default.createElement(_Forms.FormSubmit, null), /*#__PURE__*/_react.default.createElement(_Error.ErrorSummary, {
      errors: status.error
    }), submitSucceed && /*#__PURE__*/_react.default.createElement(_falconUi.Text, {
      mt: "md",
      fontSize: "md"
    }, /*#__PURE__*/_react.default.createElement(_falconI18n.T, {
      id: "forgotPassword.successMessage",
      email: values.email
    })));
  });
};

exports.ForgotPasswordForm = ForgotPasswordForm;
