// @deity/falcon-ui-kit/dist/Account/InvalidResetPasswordToken.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.InvalidResetPasswordToken = void 0;

var _react = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("react"));

var _falconI18n = /*#__PURE__*/require("@deity/falcon-i18n");

var _falconUi = /*#__PURE__*/require("@deity/falcon-ui");

var InvalidResetPasswordToken = function InvalidResetPasswordToken(_ref) {
  var onRequestAnotherToken = _ref.onRequestAnotherToken;
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_falconUi.Text, {
    fontSize: "md",
    color: "error",
    style: {
      textAlign: 'center'
    }
  }, /*#__PURE__*/_react.default.createElement(_falconI18n.T, {
    id: "resetPassword.failureMessage"
  })), /*#__PURE__*/_react.default.createElement(_falconUi.Box, {
    justifySelf: "center"
  }, /*#__PURE__*/_react.default.createElement(_falconUi.Button, {
    onClick: function onClick() {
      return onRequestAnotherToken();
    }
  }, /*#__PURE__*/_react.default.createElement(_falconI18n.T, {
    id: "resetPassword.requestAnotherToken"
  }))));
};

exports.InvalidResetPasswordToken = InvalidResetPasswordToken;
