// @deity/falcon-ui-kit/dist/Account/ResetPasswordForm.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.ResetPasswordForm = void 0;

var _extends2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("react"));

var _falconFrontKit = /*#__PURE__*/require("@deity/falcon-front-kit");

var _Error = /*#__PURE__*/require("../Error");

var _Forms = /*#__PURE__*/require("../Forms");

var ResetPasswordForm = function ResetPasswordForm(_ref) {
  var onSuccess = _ref.onSuccess,
      resetToken = _ref.resetToken,
      formProps = (0, _objectWithoutPropertiesLoose2.default)(_ref, ["onSuccess", "resetToken"]);
  return /*#__PURE__*/_react.default.createElement(_falconFrontKit.ResetPasswordFormProvider, {
    initialValues: {
      resetToken: resetToken,
      password: ''
    },
    onSuccess: onSuccess
  }, function (_ref2) {
    var status = _ref2.status;
    return /*#__PURE__*/_react.default.createElement(_Forms.Form, (0, _extends2.default)({
      id: "reset-password",
      i18nId: "resetPassword"
    }, formProps), /*#__PURE__*/_react.default.createElement(_Forms.FormField, {
      name: "password",
      required: true,
      type: "password",
      autoComplete: "new-password"
    }, function (_ref3) {
      var field = _ref3.field;
      return /*#__PURE__*/_react.default.createElement(_Forms.PasswordRevealInput, field);
    }), /*#__PURE__*/_react.default.createElement(_Forms.FormSubmit, {
      justifySelf: "center"
    }), /*#__PURE__*/_react.default.createElement(_Error.ErrorSummary, {
      errors: status.error
    }));
  });
};

exports.ResetPasswordForm = ResetPasswordForm;
