// @deity/falcon-ui-kit/dist/Account/ResetPasswordSuccess.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.ResetPasswordSuccess = void 0;

var _react = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("react"));

var _falconI18n = /*#__PURE__*/require("@deity/falcon-i18n");

var _falconUi = /*#__PURE__*/require("@deity/falcon-ui");

var ResetPasswordSuccess = function ResetPasswordSuccess(_ref) {
  var onSignIn = _ref.onSignIn;
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_falconUi.Text, {
    fontSize: "md",
    style: {
      textAlign: 'center'
    }
  }, /*#__PURE__*/_react.default.createElement(_falconI18n.T, {
    id: "resetPassword.successMessage"
  })), /*#__PURE__*/_react.default.createElement(_falconUi.Box, {
    justifySelf: "center"
  }, /*#__PURE__*/_react.default.createElement(_falconUi.Button, {
    onClick: function onClick() {
      return onSignIn();
    }
  }, /*#__PURE__*/_react.default.createElement(_falconI18n.T, {
    id: "signIn.submitButton"
  }))));
};

exports.ResetPasswordSuccess = ResetPasswordSuccess;
