// @deity/falcon-ui-kit/dist/Account/SignInForm.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.SignInForm = void 0;

var _extends2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("react"));

var _falconUi = /*#__PURE__*/require("@deity/falcon-ui");

var _falconI18n = /*#__PURE__*/require("@deity/falcon-i18n");

var _falconFrontKit = /*#__PURE__*/require("@deity/falcon-front-kit");

var _Error = /*#__PURE__*/require("../Error");

var _Forms = /*#__PURE__*/require("../Forms");

var SignInForm = function SignInForm(_ref) {
  var onSuccess = _ref.onSuccess,
      onForgotPassword = _ref.onForgotPassword,
      mutationOptions = _ref.mutationOptions,
      formProps = (0, _objectWithoutPropertiesLoose2.default)(_ref, ["onSuccess", "onForgotPassword", "mutationOptions"]);
  return /*#__PURE__*/_react.default.createElement(_falconFrontKit.SignInFormProvider, {
    onSuccess: onSuccess,
    mutationOptions: mutationOptions
  }, function (_ref2) {
    var status = _ref2.status;
    return /*#__PURE__*/_react.default.createElement(_Forms.Form, (0, _extends2.default)({
      i18nId: "signIn"
    }, formProps), /*#__PURE__*/_react.default.createElement(_Forms.FormField, {
      name: "email",
      type: "email",
      required: true,
      autoComplete: "email"
    }), /*#__PURE__*/_react.default.createElement(_Forms.FormField, {
      name: "password",
      type: "password" // pass empty array, so default password strength validator does not get triggered
      ,
      validate: [],
      required: true,
      autoComplete: "current-password"
    }, function (_ref3) {
      var field = _ref3.field;
      return /*#__PURE__*/_react.default.createElement(_Forms.PasswordRevealInput, field);
    }), /*#__PURE__*/_react.default.createElement(_falconUi.FlexLayout, {
      justifyContent: "space-between",
      alignItems: "center",
      mt: "md"
    }, /*#__PURE__*/_react.default.createElement(_falconUi.Link, {
      fontSize: "xs",
      onClick: function onClick() {
        return onForgotPassword();
      }
    }, /*#__PURE__*/_react.default.createElement(_falconI18n.T, {
      id: "signIn.forgotPasswordLink"
    })), /*#__PURE__*/_react.default.createElement(_Forms.FormSubmit, null)), /*#__PURE__*/_react.default.createElement(_Error.ErrorSummary, {
      errors: status.error
    }));
  });
};

exports.SignInForm = SignInForm;
