// @deity/falcon-ui-kit/dist/Blog/BlogPostExcerpt.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.BlogPostExcerpt = void 0;

var _extends2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("react"));

var _propTypes = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("prop-types"));

var _reactRouterDom = /*#__PURE__*/require("react-router-dom");

var _falconI18n = /*#__PURE__*/require("@deity/falcon-i18n");

var _falconUi = /*#__PURE__*/require("@deity/falcon-ui");

var _FormattedDate = /*#__PURE__*/require("../FormattedDate");

var _BlogPostExcerptLayout = /*#__PURE__*/require("./BlogPostExcerptLayout");

var BlogPostExcerpt = function BlogPostExcerpt(props) {
  var title = props.title,
      slug = props.slug,
      date = props.date,
      image = props.image,
      excerpt = props.excerpt,
      rest = (0, _objectWithoutPropertiesLoose2.default)(props, ["title", "slug", "date", "image", "excerpt"]);
  return /*#__PURE__*/_react.default.createElement(_BlogPostExcerptLayout.BlogPostExcerptLayout, (0, _extends2.default)({
    as: _reactRouterDom.Link,
    to: slug
  }, rest), image && /*#__PURE__*/_react.default.createElement(_falconUi.Image, {
    css: {
      height: 300
    },
    gridArea: _BlogPostExcerptLayout.BlogPostExcerptArea.image,
    src: image.url,
    alt: image.description
  }), /*#__PURE__*/_react.default.createElement(_falconUi.H3, {
    gridArea: _BlogPostExcerptLayout.BlogPostExcerptArea.title
  }, title), /*#__PURE__*/_react.default.createElement(_FormattedDate.FormattedDate, {
    gridArea: _BlogPostExcerptLayout.BlogPostExcerptArea.date,
    value: date
  }), /*#__PURE__*/_react.default.createElement(_falconUi.Text, {
    gridArea: _BlogPostExcerptLayout.BlogPostExcerptArea.excerpt
  }, excerpt), /*#__PURE__*/_react.default.createElement(_falconUi.Text, {
    gridArea: _BlogPostExcerptLayout.BlogPostExcerptArea.readMore,
    css: {
      textDecoration: 'underline'
    }
  }, /*#__PURE__*/_react.default.createElement(_falconI18n.T, {
    id: "blog.readMore"
  })));
};

exports.BlogPostExcerpt = BlogPostExcerpt;
BlogPostExcerpt.propTypes = {
  title: _propTypes.default.string.isRequired,
  slug: _propTypes.default.string.isRequired,
  date: /*#__PURE__*/_propTypes.default.oneOfType([_propTypes.default.string, /*#__PURE__*/_propTypes.default.instanceOf(Date)]).isRequired,
  excerpt: _propTypes.default.string.isRequired,
  // @ts-ignore https://github.com/DefinitelyTyped/DefinitelyTyped/issues/34692
  image: /*#__PURE__*/_propTypes.default.shape({
    url: _propTypes.default.string.isRequired,
    description: _propTypes.default.string.isRequired
  })
};
