// @deity/falcon-ui-kit/dist/Layouts/SidebarLayout.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.SidebarContentLayout = exports.SidebarLayout = void 0;

var _objectWithoutPropertiesLoose2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("react"));

var _falconUi = /*#__PURE__*/require("@deity/falcon-ui");

var SidebarLayoutInnerDOM = function SidebarLayoutInnerDOM(_ref) {
  var title = _ref.title,
      onClose = _ref.onClose,
      children = _ref.children,
      rest = (0, _objectWithoutPropertiesLoose2.default)(_ref, ["title", "onClose", "children"]);
  return /*#__PURE__*/_react.default.createElement(_falconUi.Box, rest, /*#__PURE__*/_react.default.createElement(_falconUi.FlexLayout, null, /*#__PURE__*/_react.default.createElement(_falconUi.Box, {
    flex: 1
  }, !!title && /*#__PURE__*/_react.default.createElement(_falconUi.H3, null, title)), /*#__PURE__*/_react.default.createElement(_falconUi.Icon, {
    src: "close",
    stroke: "black",
    onClick: onClose,
    "aria-controls": rest.id
  })), /*#__PURE__*/_react.default.createElement(SidebarContentLayout, null, children));
};

var SidebarLayout = /*#__PURE__*/(0, _falconUi.themed)({
  tag: SidebarLayoutInnerDOM,
  defaultTheme: {
    sidebarLayout: {
      display: 'grid',
      gridRowGap: 'md',
      gridTemplate: 'auto 1fr / 100%',
      css: {
        width: '100%',
        height: '100%'
      }
    }
  }
});
exports.SidebarLayout = SidebarLayout;
var SidebarContentLayout = /*#__PURE__*/(0, _falconUi.themed)({
  tag: _falconUi.Box,
  defaultProps: {
    maxWidth: 'XL'
  },
  defaultTheme: {
    sidebarContentLayout: {
      css: function css(_ref2) {
        var maxWidth = _ref2.maxWidth;
        return {
          maxWidth: maxWidth,
          width: '100%',
          margin: '0 auto'
        };
      }
    }
  }
});
exports.SidebarContentLayout = SidebarContentLayout;
