// @deity/falcon-front-kit/dist/Address/EditAddressFormProvider.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.EditAddressFormProvider = void 0;

var _extends2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("react"));

var _formik = /*#__PURE__*/require("formik");

var _falconData = /*#__PURE__*/require("@deity/falcon-data");

var _falconShopData = /*#__PURE__*/require("@deity/falcon-shop-data");

var _React = /*#__PURE__*/require("../React");

var EditAddressFormProvider = function EditAddressFormProvider(props) {
  var address = props.address,
      onSuccess = props.onSuccess,
      initialValues = props.initialValues,
      mutationOptions = props.mutationOptions,
      formikProps = (0, _objectWithoutPropertiesLoose2.default)(props, ["address", "onSuccess", "initialValues", "mutationOptions"]);
  var __typename = address.__typename,
      street = address.street,
      rest = (0, _objectWithoutPropertiesLoose2.default)(address, ["__typename", "street"]);
  var defaultInitialValues = Object.assign({
    street1: street.length > 0 ? street[0] : undefined,
    street2: street.length > 1 ? street[1] : undefined
  }, rest);

  var _useEditAddressMutati = (0, _falconShopData.useEditAddressMutation)(),
      editAddress = _useEditAddressMutati[0];

  var getUserError = (0, _falconData.useGetUserError)();
  var isMounted = (0, _React.useMounted)();
  return /*#__PURE__*/_react.default.createElement(_formik.Formik, (0, _extends2.default)({
    initialStatus: {},
    initialValues: initialValues || defaultInitialValues,
    onSubmit: function onSubmit(_ref, _ref2) {
      var setSubmitting = _ref2.setSubmitting,
          setStatus = _ref2.setStatus;
      var street1 = _ref.street1,
          street2 = _ref.street2,
          country = _ref.country,
          region = _ref.region,
          values = (0, _objectWithoutPropertiesLoose2.default)(_ref, ["street1", "street2", "country", "region"]);
      return editAddress(Object.assign({
        variables: {
          input: Object.assign({}, values, {
            id: address.id,
            street: [street1, street2].filter(Boolean),
            countryId: country.id,
            regionId: region ? region.id : undefined
          })
        }
      }, mutationOptions || {})).then(function (_ref3) {
        var data = _ref3.data;

        if (isMounted.current) {
          setSubmitting(false);
          setStatus({
            data: data
          });
          return onSuccess && onSuccess(data);
        }
      }).catch(function (e) {
        var error = getUserError(e);

        if (error.length && isMounted.current) {
          setStatus({
            error: error
          });
          setSubmitting(false);
        }
      });
    }
  }, formikProps));
};

exports.EditAddressFormProvider = EditAddressFormProvider;
