// @deity/falcon-ui-kit/dist/SortOrder/SortOrderPicker.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.SortOrderPicker = void 0;

var _react = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("react"));

var _falconUi = /*#__PURE__*/require("@deity/falcon-ui");

var _falconI18n = /*#__PURE__*/require("@deity/falcon-i18n");

var _SortOrderPickerLayout = /*#__PURE__*/require("./SortOrderPickerLayout");

var SortOrderPicker = function SortOrderPicker(_ref) {
  var items = _ref.items,
      value = _ref.value,
      onChange = _ref.onChange;
  return /*#__PURE__*/_react.default.createElement(_SortOrderPickerLayout.SortOrderPickerLayout, null, /*#__PURE__*/_react.default.createElement(_falconUi.Text, {
    mr: "sm"
  }, /*#__PURE__*/_react.default.createElement(_falconI18n.T, {
    id: "productList.sort.title"
  })), /*#__PURE__*/_react.default.createElement(_falconUi.Box, {
    display: "flex"
  }, /*#__PURE__*/_react.default.createElement(_falconUi.Dropdown, {
    css: {
      width: '100%',
      minWidth: 200
    },
    onChange: onChange
  }, /*#__PURE__*/_react.default.createElement(_falconUi.DropdownLabel, null, value.name), /*#__PURE__*/_react.default.createElement(_falconUi.DropdownMenu, null, items.map(function (x) {
    return /*#__PURE__*/_react.default.createElement(_falconUi.DropdownMenuItem, {
      key: x.name,
      value: x
    }, x.name);
  })))));
};

exports.SortOrderPicker = SortOrderPicker;
