// @deity/falcon-front-kit/dist/Account/SignUpFormProvider.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.SignUpFormProvider = void 0;

var _extends2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("react"));

var _formik = /*#__PURE__*/require("formik");

var _falconData = /*#__PURE__*/require("@deity/falcon-data");

var _falconShopData = /*#__PURE__*/require("@deity/falcon-shop-data");

var _React = /*#__PURE__*/require("../React");

var SignUpFormProvider = function SignUpFormProvider(props) {
  var onSuccess = props.onSuccess,
      initialValues = props.initialValues,
      mutationOptions = props.mutationOptions,
      formikProps = (0, _objectWithoutPropertiesLoose2.default)(props, ["onSuccess", "initialValues", "mutationOptions"]);
  var defaultInitialValues = {
    firstname: '',
    lastname: '',
    email: '',
    password: '',
    confirmPassword: '',
    autoSignIn: true
  };

  var _useSignUpMutation = (0, _falconShopData.useSignUpMutation)(),
      signUp = _useSignUpMutation[0];

  var getUserError = (0, _falconData.useGetUserError)();
  var isMounted = (0, _React.useMounted)();
  return /*#__PURE__*/_react.default.createElement(_formik.Formik, (0, _extends2.default)({
    initialStatus: {},
    initialValues: initialValues || defaultInitialValues,
    onSubmit: function onSubmit(_ref, _ref2) {
      var setSubmitting = _ref2.setSubmitting,
          setStatus = _ref2.setStatus,
          setErrors = _ref2.setErrors;
      var confirmPassword = _ref.confirmPassword,
          signUpValues = (0, _objectWithoutPropertiesLoose2.default)(_ref, ["confirmPassword"]);
      return signUp(Object.assign({
        variables: {
          input: signUpValues
        }
      }, mutationOptions || {})).then(function (_ref3) {
        var data = _ref3.data;

        if (isMounted.current) {
          setSubmitting(false);
          setStatus({
            data: data
          });
          return onSuccess && onSuccess(data);
        }
      }).catch(function (e) {
        var error = getUserError(e);

        if (error.length && isMounted.current) {
          setErrors(error.getFieldErrors());
          setStatus({
            error: error.getFormErrors()
          });
          setSubmitting(false);
        }
      });
    }
  }, formikProps));
};

exports.SignUpFormProvider = SignUpFormProvider;
