// @deity/falcon-i18n/dist/I18nProvider.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.I18nProvider = void 0;

var _assertThisInitialized2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/assertThisInitialized"));

var _inheritsLoose2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/inheritsLoose"));

var _react = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("react"));

var _context = /*#__PURE__*/require("./context");

var I18nProvider = /*#__PURE__*/function (_React$Component) {
  (0, _inheritsLoose2.default)(I18nProvider, _React$Component);

  function I18nProvider(props) {
    var _this;

    _this = _React$Component.call(this, props) || this;
    _this.state = {
      updatedAt: new Date()
    };
    _this.onI18nChanged = _this.onI18nChanged.bind((0, _assertThisInitialized2.default)(_this));
    return _this;
  }

  var _proto = I18nProvider.prototype;

  _proto.componentDidMount = function componentDidMount() {
    var _this2 = this;

    var i18n = this.props.i18n;

    _context.defaultOptions.rerenderOn.forEach(function (x) {
      return i18n.on(x, _this2.onI18nChanged);
    });
  };

  _proto.componentWillUnmount = function componentWillUnmount() {
    var _this3 = this;

    var i18n = this.props.i18n;

    _context.defaultOptions.rerenderOn.forEach(function (x) {
      return i18n.off(x, _this3.onI18nChanged);
    });
  };

  _proto.onI18nChanged = function onI18nChanged() {
    this.setState({
      updatedAt: new Date()
    });
  };

  _proto.render = function render() {
    var _this$props = this.props,
        i18n = _this$props.i18n,
        children = _this$props.children;
    var updatedAt = this.state.updatedAt;
    (0, _context.setI18n)(i18n);
    return /*#__PURE__*/_react.default.createElement(_context.I18nContext.Provider, {
      value: {
        updatedAt: updatedAt,
        i18n: i18n,
        language: i18n.language,
        t: i18n.t.bind(i18n),
        options: _context.defaultOptions
      }
    }, children);
  };

  return I18nProvider;
}(_react.default.Component);

exports.I18nProvider = I18nProvider;
