// @deity/falcon-data/dist/Error/userError.js
"use strict";

exports.__esModule = true;
exports.tryGetUserError = exports.isUserError = exports.isNetworkError = void 0;

var _client = /*#__PURE__*/require("@apollo/client");

var _falconErrors = /*#__PURE__*/require("@deity/falcon-errors");

/**
 * Determines if `error` is Apollo network error
 * @param error
 */
var isNetworkError = function isNetworkError(error) {
  return (0, _client.isApolloError)(error) && !!error.networkError;
};
/**
 * Determines if `error` is Apollo graphQl error, which is considered as user error
 * @param error
 */


exports.isNetworkError = isNetworkError;

var isUserError = function isUserError(error) {
  if ((0, _client.isApolloError)(error)) {
    var graphQLErrors = error.graphQLErrors,
        networkError = error.networkError;

    if (!networkError && Array.isArray(graphQLErrors) && graphQLErrors.length > 0) {
      return true;
    }
  }

  return false;
};
/**
 * Extract all user errors if any
 * @param error
 */


exports.isUserError = isUserError;

var tryGetUserError = function tryGetUserError(error) {
  if (isUserError(error)) {
    var _ref = error,
        graphQLErrors = _ref.graphQLErrors;
    return graphQLErrors.reduce(function (result, _ref2) {
      var message = _ref2.message,
          _ref2$extensions = _ref2.extensions,
          extensions = _ref2$extensions === void 0 ? {} : _ref2$extensions;

      if (extensions.code === _falconErrors.codes.BAD_USER_INPUT && extensions.errors) {
        var errors = extensions.errors;
        var userInputErrors = Object.keys(errors).map(function (x) {
          return {
            message: errors[x].message || errors[x],
            code: errors[x].code || 'UNKNOWN',
            path: x
          };
        });
        return [].concat(result, userInputErrors);
      }

      return [].concat(result, [{
        message: message,
        code: extensions.code
      }]);
    }, []);
  }

  return [];
};

exports.tryGetUserError = tryGetUserError;
