// @deity/falcon-ui-kit/dist/Order/EmptyOrderList.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.EmptyOrderList = void 0;

var _react = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("react"));

var _reactRouterDom = /*#__PURE__*/require("react-router-dom");

var _falconUi = /*#__PURE__*/require("@deity/falcon-ui");

var _falconI18n = /*#__PURE__*/require("@deity/falcon-i18n");

var EmptyOrderList = function EmptyOrderList() {
  return /*#__PURE__*/_react.default.createElement(_falconUi.FlexLayout, {
    flexDirection: "column",
    alignItems: "center",
    p: "sm"
  }, /*#__PURE__*/_react.default.createElement(_falconUi.Text, {
    fontSize: "md",
    mb: "xs"
  }, /*#__PURE__*/_react.default.createElement(_falconI18n.T, {
    id: "orderList.empty"
  })), /*#__PURE__*/_react.default.createElement(_falconUi.Button, {
    as: _reactRouterDom.Link,
    to: "/"
  }, /*#__PURE__*/_react.default.createElement(_falconI18n.T, {
    id: "cart.goShoppingButton"
  })));
};

exports.EmptyOrderList = EmptyOrderList;
