// @deity/falcon-ui-kit/dist/Order/OrderItemSummary.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.OrderItemSummary = void 0;

var _react = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("react"));

var _falconI18n = /*#__PURE__*/require("@deity/falcon-i18n");

var _falconUi = /*#__PURE__*/require("@deity/falcon-ui");

var _Price = /*#__PURE__*/require("../Price");

var _Layouts = /*#__PURE__*/require("../Layouts");

var _OrderItemSummaryLayout = /*#__PURE__*/require("./OrderItemSummaryLayout");

var OrderItemSummary = function OrderItemSummary(props) {
  var thumbMaxSize = 80;
  return /*#__PURE__*/_react.default.createElement(_OrderItemSummaryLayout.OrderItemSummaryLayout, {
    maxThumbSize: thumbMaxSize
  }, /*#__PURE__*/_react.default.createElement(_falconUi.Image, {
    gridArea: _OrderItemSummaryLayout.OrderItemSummaryArea.thumb,
    src: props.thumbnailUrl,
    css: {
      maxWidth: thumbMaxSize + "px",
      maxHeight: thumbMaxSize + "px"
    }
  }), /*#__PURE__*/_react.default.createElement(_falconUi.Text, {
    gridArea: _OrderItemSummaryLayout.OrderItemSummaryArea.name,
    fontWeight: "bold"
  }, props.name), /*#__PURE__*/_react.default.createElement(_Price.Price, {
    gridArea: _OrderItemSummaryLayout.OrderItemSummaryArea.price,
    value: props.rowTotalInclTax,
    fontWeight: "bold",
    css: {
      textAlign: 'right'
    }
  }), /*#__PURE__*/_react.default.createElement(_falconUi.Box, {
    gridArea: _OrderItemSummaryLayout.OrderItemSummaryArea.details
  }, /*#__PURE__*/_react.default.createElement(_Layouts.PropertyRowLayout, null, /*#__PURE__*/_react.default.createElement(_falconI18n.T, {
    id: "order.quantityLabel"
  }), /*#__PURE__*/_react.default.createElement(_falconUi.Text, null, props.qty))));
};

exports.OrderItemSummary = OrderItemSummary;
