// @deity/falcon-ui-kit/dist/Order/OrderItemSummaryLayout.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.OrderItemSummaryLayout = exports.OrderItemSummaryArea = void 0;

var _propTypes = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("prop-types"));

var _falconUi = /*#__PURE__*/require("@deity/falcon-ui");

var _helpers = /*#__PURE__*/require("../helpers");

var OrderItemSummaryArea = {
  thumb: 'thumb',
  name: 'name',
  price: 'price',
  details: 'details'
};
exports.OrderItemSummaryArea = OrderItemSummaryArea;
var OrderItemSummaryLayout = /*#__PURE__*/(0, _falconUi.themed)({
  tag: _falconUi.Box,
  defaultTheme: {
    orderedItemSummaryLayout: {
      display: 'grid',
      gridGap: 'sm',
      fontSize: 'xs',
      css: function css(_ref) {
        var maxThumbSize = _ref.maxThumbSize;
        return {
          gridTemplate: (0, _helpers.toGridTemplate)([[maxThumbSize + "px", '1fr', 'auto'], [OrderItemSummaryArea.thumb, OrderItemSummaryArea.name, OrderItemSummaryArea.price], [OrderItemSummaryArea.thumb, OrderItemSummaryArea.details, OrderItemSummaryArea.price]])
        };
      }
    }
  }
});
exports.OrderItemSummaryLayout = OrderItemSummaryLayout;
OrderItemSummaryLayout.propTypes = {
  maxThumbSize: _propTypes.default.number.isRequired
};
