// @deity/falcon-theme-editor/dist/editortheme.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.editorTheme = void 0;

var _extends2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/extends"));

var _react = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("react"));

var _falconUi = /*#__PURE__*/require("@deity/falcon-ui");

var editorTheme = /*#__PURE__*/(0, _falconUi.createTheme)({
  components: {
    summary: {
      bg: 'secondaryLight',
      borderRadius: 'round'
    },
    button: {
      borderRadius: 'round'
    }
  },
  icons: {
    presets: {
      icon: function icon(props) {
        return /*#__PURE__*/_react.default.createElement("svg", (0, _extends2.default)({
          viewBox: "0 0 50 48"
        }, props), /*#__PURE__*/_react.default.createElement("path", {
          d: "M45.188,19.75c0.138,0,0.25-0.112,0.25-0.25V4.875c0-0.138-0.112-0.25-0.25-0.25H30.563c-0.138,0-0.25,0.112-0.25,0.25  v7.969H19.688V4.875c0-0.138-0.112-0.25-0.25-0.25H4.813c-0.138,0-0.25,0.112-0.25,0.25V19.5c0,0.138,0.112,0.25,0.25,0.25h8.031  v10.5H4.813c-0.138,0-0.25,0.112-0.25,0.25v14.625c0,0.138,0.112,0.25,0.25,0.25h14.625c0.138,0,0.25-0.112,0.25-0.25v-7.969h10.625  v7.969c0,0.138,0.112,0.25,0.25,0.25h14.625c0.138,0,0.25-0.112,0.25-0.25V30.5c0-0.138-0.112-0.25-0.25-0.25h-8.031v-10.5H45.188z   M30.813,5.125h14.125V19.25h-7.781v-6.156c0-0.138-0.112-0.25-0.25-0.25h-6.094V5.125z M13.344,36.656V13.344h23.313v6.154  c0,0.001-0.001,0.001-0.001,0.002s0.001,0.001,0.001,0.002v10.996c0,0-0.001,0.001-0.001,0.002s0.001,0.002,0.001,0.002v6.154  H13.344z M5.063,19.25V5.125h14.125v7.719h-6.094c-0.138,0-0.25,0.112-0.25,0.25v6.156H5.063z M19.188,44.875H5.063V30.75h7.781  v6.156c0,0.138,0.112,0.25,0.25,0.25h6.094V44.875z M44.938,30.75v14.125H30.813v-7.719h6.094c0.138,0,0.25-0.112,0.25-0.25V30.75  H44.938z"
        }));
      }
    },
    component: {
      icon: function icon(props) {
        return /*#__PURE__*/_react.default.createElement("svg", (0, _extends2.default)({
          viewBox: "5 0 90 90"
        }, props), /*#__PURE__*/_react.default.createElement("path", {
          d: "M17.6,63.9c0,1.5,0.7,2.8,1.8,3.7L47.1,89c0.9,0.7,1.9,1,2.9,1s2-0.3,2.9-1l27.7-21.4c1.2-0.9,1.8-2.3,1.8-3.7    c0-1.5-0.7-2.8-1.8-3.7L76.4,57l4.2-3.2c1.2-0.9,1.8-2.3,1.8-3.7c0-1.5-0.7-2.8-1.8-3.7L76.4,43l4.2-3.2c1.2-0.9,1.8-2.3,1.8-3.7    s-0.7-2.8-1.8-3.7L52.9,11c-1.6-1.3-4.1-1.3-5.8,0L19.4,32.3c-1.2,0.9-1.8,2.3-1.8,3.7s0.7,2.8,1.8,3.7l4.2,3.2l-4.2,3.2    c-1.2,0.9-1.8,2.3-1.8,3.7c0,1.5,0.7,2.8,1.8,3.7l4.2,3.2l-4.2,3.2C18.3,61.1,17.6,62.4,17.6,63.9z M21.6,36.1    c0-0.1,0-0.4,0.3-0.6l27.7-21.4C49.7,14,49.9,14,50,14s0.3,0,0.4,0.2l27.7,21.4c0.2,0.2,0.3,0.4,0.3,0.6s0,0.4-0.3,0.6L50.4,58    c-0.3,0.3-0.6,0.3-0.9,0L21.9,36.7C21.6,36.5,21.6,36.2,21.6,36.1z M21.6,50c0-0.1,0-0.4,0.3-0.6l5-3.9l20.2,15.6    c0.8,0.6,1.9,1,2.9,1s2.1-0.4,2.9-1l20.2-15.6l5,3.9c0.2,0.2,0.3,0.4,0.3,0.6s0,0.4-0.3,0.6l-6.2,4.8c0,0-0.1,0-0.1,0.1L50.4,71.9    c-0.3,0.2-0.6,0.2-0.9,0L28.2,55.5c0,0-0.1,0-0.1-0.1l-6.2-4.8C21.6,50.4,21.6,50.1,21.6,50z M21.9,63.3l5-3.9l20.2,15.6    c0.9,0.7,1.9,1,2.9,1s2-0.3,2.9-1l20.2-15.6l5,3.9c0.2,0.2,0.3,0.4,0.3,0.6c0,0.1,0,0.4-0.3,0.6L50.4,85.9c-0.3,0.2-0.6,0.2-0.9,0    L21.9,64.5c-0.2-0.2-0.3-0.4-0.3-0.6C21.6,63.8,21.6,63.5,21.9,63.3z"
        }));
      }
    },
    finder: {
      icon: function icon(props) {
        return /*#__PURE__*/_react.default.createElement("svg", (0, _extends2.default)({
          xmlns: "http://www.w3.org/2000/svg",
          viewBox: "0 0 24 24"
        }, props), /*#__PURE__*/_react.default.createElement("path", {
          d: "M8 3H5a2 2 0 0 0-2 2v3m18 0V5a2 2 0 0 0-2-2h-3m0 18h3a2 2 0 0 0 2-2v-3M3 16v3a2 2 0 0 0 2 2h3"
        }));
      },
      css: {
        cursor: 'pointer',
        fill: 'none'
      }
    },
    download: {
      icon: function icon(props) {
        return /*#__PURE__*/_react.default.createElement("svg", (0, _extends2.default)({
          viewBox: "10 0 85 85"
        }, props), /*#__PURE__*/_react.default.createElement("path", {
          d: "M22.25,76.813v4h55.5v-4H22.25z M72.98,45.41l-2.828-2.828L52,60.734V21.188h-4v39.547L29.848,42.582L27.02,45.41L50,68.391  L72.98,45.41z"
        }));
      },
      css: {
        cursor: 'pointer',
        fill: 'none'
      }
    },
    copy: {
      icon: function icon(props) {
        return /*#__PURE__*/_react.default.createElement("svg", (0, _extends2.default)({
          viewBox: "0 0 24 24",
          strokeWidth: 2
        }, props), /*#__PURE__*/_react.default.createElement("rect", {
          x: "9",
          y: "9",
          width: "13",
          height: "13",
          rx: "2",
          ry: "2"
        }), /*#__PURE__*/_react.default.createElement("path", {
          d: "M5 15H4a2 2 0 0 1-2-2V4a2 2 0 0 1 2-2h9a2 2 0 0 1 2 2v1"
        }));
      },
      fill: 'transparent'
    },
    theme: {
      icon: function icon(props) {
        return /*#__PURE__*/_react.default.createElement("svg", (0, _extends2.default)({
          viewBox: "0 0 100 100"
        }, props), /*#__PURE__*/_react.default.createElement("path", {
          d: "M12,52H30v7H48V52H88V48H48V41H30v7H12Zm22-7H44V55H34Z"
        }), /*#__PURE__*/_react.default.createElement("path", {
          d: "M12,22H49v7H67V22H88V18H67V11H49v7H12Zm41-7H63V25H53Z"
        }), /*#__PURE__*/_react.default.createElement("path", {
          d: "M12,82H49v7H67V82H88V78H67V71H49v7H12Zm41-7H63V85H53Z"
        }));
      }
    },
    dropdownArrowDown: {
      icon: function icon(props) {
        return /*#__PURE__*/_react.default.createElement("svg", (0, _extends2.default)({
          viewBox: "0 0 24 24"
        }, props), /*#__PURE__*/_react.default.createElement("polyline", {
          points: "6 9 12 15 18 9"
        }));
      },
      stroke: 'primary',
      fill: 'transparent',
      size: 'md',
      ml: 'xs'
    },
    dropdownArrowUp: {
      icon: function icon(props) {
        return /*#__PURE__*/_react.default.createElement("svg", (0, _extends2.default)({
          viewBox: "0 0 24 24"
        }, props), /*#__PURE__*/_react.default.createElement("polyline", {
          points: "18 15 12 9 6 15"
        }));
      },
      stroke: 'primary',
      fill: 'transparent',
      size: 'md',
      ml: 'xs'
    }
  }
});
exports.editorTheme = editorTheme;
