// @deity/falcon-theme-editor/dist/ThemePresets.js
"use strict";

var _interopRequireWildcard = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireWildcard");

exports.__esModule = true;
exports.ThemePresets = void 0;

var _react = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("react"));

var _falconUi = /*#__PURE__*/require("@deity/falcon-ui");

var _ThemeEditorState = /*#__PURE__*/require("./ThemeEditorState");

var ThemePresets = function ThemePresets(_ref) {
  var activePreset = _ref.activePreset,
      onPresetChange = _ref.onPresetChange;

  var _useContext = (0, _react.useContext)(_ThemeEditorState.ThemeStateContext),
      presets = _useContext.presets;

  return /*#__PURE__*/_react.default.createElement(_falconUi.Group, {
    my: "md",
    mx: "md",
    display: "flex"
  }, presets.map(function (preset) {
    return /*#__PURE__*/_react.default.createElement(_falconUi.Button, {
      key: preset.name,
      variant: preset.name === activePreset ? '' : 'secondary',
      flex: "1",
      height: "xxl",
      onClick: onPresetChange(preset)
    }, preset.name);
  }));
};

exports.ThemePresets = ThemePresets;
