// @deity/falcon-theme-editor/dist/ThemePropsEditor.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.ThemePropsEditor = void 0;

var _react = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("react"));

var _falconUi = /*#__PURE__*/require("@deity/falcon-ui");

var _thememeta = /*#__PURE__*/require("./thememeta");

var _fonts = /*#__PURE__*/require("./fonts");

var EditorSection = function EditorSection(_ref) {
  var theme = _ref.theme,
      section = _ref.section,
      onChange = _ref.onChange,
      onFontChange = _ref.onFontChange;
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, Object.keys(theme[section]).map(function (themeProp) {
    var meta = _thememeta.themeMeta[section];
    if (!meta) return null;
    return /*#__PURE__*/_react.default.createElement(_falconUi.GridLayout, {
      alignItems: "center",
      gridGap: "xs",
      mb: "xs",
      gridTemplateColumns: meta.input.type === 'dropdown' ? '50px auto 1.8fr 20px' : '1.1fr auto 1.8fr 35px',
      key: themeProp
    }, /*#__PURE__*/_react.default.createElement(_falconUi.Text, {
      p: "xs"
    }, "\"" + themeProp + "\""), meta.input.type === 'number' && /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_falconUi.NumberInput, {
      value: theme[section][themeProp],
      min: meta.input.min,
      max: meta.input.max,
      step: meta.input.step,
      onChange: onChange(section, themeProp, true)
    }), /*#__PURE__*/_react.default.createElement(_falconUi.RangeInput, {
      value: theme[section][themeProp],
      min: meta.input.min,
      max: meta.input.max,
      step: meta.input.step,
      onChange: onChange(section, themeProp, true)
    }), /*#__PURE__*/_react.default.createElement(_falconUi.Text, {
      p: "none",
      ml: "xs",
      fontSize: "sm"
    }, "px")), meta.input.type === 'dropdown' && /*#__PURE__*/_react.default.createElement(_falconUi.Dropdown, {
      onChange: onFontChange(themeProp),
      gridColumn: "span 3"
    }, /*#__PURE__*/_react.default.createElement(_falconUi.DropdownLabel, null, theme[section][themeProp]), /*#__PURE__*/_react.default.createElement(_falconUi.DropdownMenu, null, _fonts.fonts.map(function (font) {
      return /*#__PURE__*/_react.default.createElement(_falconUi.DropdownMenuItem, {
        key: font.value,
        value: font
      }, font.value + " " + (font.google ? ' - (Google Font)' : ''));
    }))), meta.input.type === 'color' && /*#__PURE__*/_react.default.createElement(_falconUi.Input, {
      onChange: onChange(section, themeProp),
      type: "color",
      gridColumn: "span 2",
      value: theme[section][themeProp],
      css: {
        padding: 0,
        width: 60,
        borderRadius: 0,
        border: 'none'
      }
    }), meta.input.type === 'text' && /*#__PURE__*/_react.default.createElement(_falconUi.Input, {
      onChange: onChange(section, themeProp),
      gridColumn: "span 3",
      type: "text",
      value: theme[section][themeProp]
    }));
  }));
};

var ThemePropsEditor = function ThemePropsEditor(_ref2) {
  var openPanels = _ref2.openPanels,
      toggleOpenPanel = _ref2.toggleOpenPanel,
      theme = _ref2.theme,
      onChange = _ref2.onChange,
      onFontChange = _ref2.onFontChange;
  return /*#__PURE__*/_react.default.createElement(_falconUi.GridLayout, {
    gridGap: "xs"
  }, Object.keys(_thememeta.themeCategories).map(function (categoryKey) {
    var category = _thememeta.themeCategories[categoryKey];
    return /*#__PURE__*/_react.default.createElement(_falconUi.Details, {
      key: category.name,
      open: openPanels[categoryKey]
    }, /*#__PURE__*/_react.default.createElement(_falconUi.Summary, {
      onClick: function onClick(e) {
        e.preventDefault();
        toggleOpenPanel(categoryKey);
      }
    }, category.name), openPanels[categoryKey] && /*#__PURE__*/_react.default.createElement(_falconUi.DetailsContent, {
      pr: "xs"
    }, category.themeSections.length === 1 && /*#__PURE__*/_react.default.createElement(EditorSection, {
      section: category.themeSections[0],
      theme: theme,
      onChange: onChange,
      onFontChange: onFontChange
    }), category.themeSections.length > 1 && category.themeSections.map(function (section) {
      var key = categoryKey + section;
      return /*#__PURE__*/_react.default.createElement(_falconUi.Details, {
        mb: "sm",
        key: key,
        open: openPanels[key]
      }, /*#__PURE__*/_react.default.createElement(_falconUi.Summary, {
        onClick: function onClick(e) {
          e.preventDefault();
          toggleOpenPanel(key);
        }
      }, section), /*#__PURE__*/_react.default.createElement(_falconUi.DetailsContent, {
        mr: "xs"
      }, /*#__PURE__*/_react.default.createElement(EditorSection, {
        section: section,
        theme: theme,
        onChange: onChange,
        onFontChange: onFontChange
      })));
    })));
  }));
};

exports.ThemePropsEditor = ThemePropsEditor;
