// @deity/falcon-theme-editor/dist/thememeta.js
"use strict";

exports.__esModule = true;
exports.themeCategories = exports.themeMeta = void 0;

var _fonts = /*#__PURE__*/require("./fonts");

var themeMeta = {
  colors: {
    input: {
      type: 'color'
    },
    previewCss: function previewCss(value) {
      return {
        backgroundColor: value
      };
    }
  },
  spacing: {
    input: {
      type: 'number',
      min: 0,
      step: 1,
      max: 100
    },
    unit: 'px',
    previewCss: function previewCss(value) {
      return {
        width: value
      };
    }
  },
  fonts: {
    input: {
      type: 'dropdown',
      options: _fonts.fonts
    }
  },
  fontSizes: {
    input: {
      type: 'number',
      min: 0,
      step: 1,
      max: 80
    },
    previewCss: function previewCss(value) {
      return {
        fontSize: value
      };
    },
    unit: 'px'
  },
  fontWeights: {
    input: {
      type: 'text'
    }
  },
  lineHeights: {
    input: {
      type: 'number',
      min: 0,
      step: 0.1,
      max: 3
    }
  },
  letterSpacings: {
    input: {
      type: 'text'
    },
    previewCss: function previewCss(value) {
      return {
        letterSpacing: value
      };
    }
  },
  breakpoints: {
    input: {
      type: 'number',
      step: 1,
      min: 0,
      max: 2048
    },
    unit: 'px'
  },
  borders: {
    input: {
      type: 'text'
    },
    previewCss: function previewCss(value) {
      return {
        border: value
      };
    }
  },
  borderRadius: {
    input: {
      type: 'number',
      min: 0,
      step: 1,
      max: 100
    },
    unit: 'px',
    previewCss: function previewCss(value) {
      return {
        borderRadius: value
      };
    }
  },
  boxShadows: {
    input: {
      type: 'text'
    },
    previewCss: function previewCss(value) {
      return {
        boxShadow: value
      };
    }
  },
  easingFunctions: {
    input: {
      type: 'text'
    }
  },
  transitionDurations: {
    input: {
      type: 'text'
    }
  }
};
exports.themeMeta = themeMeta;
var themeCategories = {
  colors: {
    name: 'Colors',
    themeSections: ['colors']
  },
  spacing: {
    name: 'Spacing',
    themeSections: ['spacing']
  },
  typography: {
    name: 'Typography',
    themeSections: ['fonts', 'fontSizes', 'fontWeights', 'letterSpacings', 'lineHeights']
  },
  breakpoints: {
    name: 'Breakpoints',
    themeSections: ['breakpoints']
  },
  borders: {
    name: 'Borders',
    themeSections: ['borders', 'borderRadius']
  },
  misc: {
    name: 'Miscellaneous',
    themeSections: ['boxShadows', 'easingFunctions', 'transitionDurations']
  }
};
exports.themeCategories = themeCategories;
