// @deity/falcon-ui-kit/dist/Address/AddressDetails.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.AddressDetails = void 0;

var _react = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("react"));

var _propTypes = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("prop-types"));

var _falconUi = /*#__PURE__*/require("@deity/falcon-ui");

var _AddressDetailsLayout = /*#__PURE__*/require("./AddressDetailsLayout");

var AddressDetails = function AddressDetails(_ref) {
  var company = _ref.company,
      firstname = _ref.firstname,
      lastname = _ref.lastname,
      street = _ref.street,
      postcode = _ref.postcode,
      city = _ref.city,
      country = _ref.country,
      telephone = _ref.telephone;
  return /*#__PURE__*/_react.default.createElement(_AddressDetailsLayout.AddressDetailsLayout, null, company && /*#__PURE__*/_react.default.createElement(_falconUi.Text, {
    fontWeight: "bold",
    color: "secondaryText"
  }, "" + company), /*#__PURE__*/_react.default.createElement(_falconUi.Text, {
    fontWeight: "bold",
    color: "secondaryText",
    mb: "xs"
  }, firstname + " " + lastname), street.map(function (x, i) {
    return (
      /*#__PURE__*/
      // because `street` can contain non unique values
      // eslint-disable-next-line react/no-array-index-key
      _react.default.createElement(_falconUi.Text, {
        key: i
      }, x)
    );
  }), /*#__PURE__*/_react.default.createElement(_falconUi.Text, null, postcode + " " + city + ", " + (country.localName || country.code)), telephone && /*#__PURE__*/_react.default.createElement(_falconUi.Text, null, telephone));
};

exports.AddressDetails = AddressDetails;
AddressDetails.propTypes = {
  company: _propTypes.default.string,
  firstname: _propTypes.default.string.isRequired,
  lastname: _propTypes.default.string.isRequired,
  street: /*#__PURE__*/_propTypes.default.arrayOf(_propTypes.default.string.isRequired).isRequired,
  postcode: _propTypes.default.string,
  city: _propTypes.default.string.isRequired,
  // @ts-ignore
  country: /*#__PURE__*/_propTypes.default.shape({
    code: _propTypes.default.string.isRequired,
    localName: _propTypes.default.string
  }).isRequired,
  telephone: _propTypes.default.string
};
