// @deity/falcon-mollie-plugin/dist/withMollie.js
"use strict";

var _interopRequireWildcard = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.withMollie = void 0;

var _extends2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/extends"));

var _react = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("react"));

var _reactHelmetAsync = /*#__PURE__*/require("react-helmet-async");

/**
 * Mollie Docs
 * https://docs.mollie.com/guides/mollie-components/overview#add-the-mollie-components-javascript-library-to-your-checkout
 */
var locales = {
  default: 'en_US',
  en: 'en_US',
  nl: 'nl_NL' // TODO: possible remaining values to be mapped: nl_BE fr_FR fr_BE de_DE de_AT de_CH es_ES ca_ES pt_PT it_IT nb_NO sv_SE fi_FI da_DK is_IS hu_HU pl_PL lv_LV lt_LT

};

var withMollie = function withMollie(WrappedComponent) {
  return function (props) {
    var _useState = (0, _react.useState)(false),
        scriptLoaded = _useState[0],
        setScriptLoaded = _useState[1];

    var _useState2 = (0, _react.useState)(null),
        mollie = _useState2[0],
        setMollie = _useState2[1];

    var errMessage = 'Mollie ProfileID was not returned from the server!';
    var _props$locale = props.locale,
        appLocale = _props$locale === void 0 ? 'en_US' : _props$locale,
        LoaderComp = props.LoaderComp,
        ErrorComponent = props.ErrorComponent,
        config = props.config;
    var profileId = config.profileId,
        _config$testMode = config.testMode,
        testMode = _config$testMode === void 0 ? false : _config$testMode;
    (0, _react.useEffect)(function () {
      if (scriptLoaded && profileId) {
        setMollie(window.Mollie(profileId, {
          locale: locales[appLocale] || locales.default,
          testmode: testMode
        }));
      }
    }, [scriptLoaded, appLocale, profileId, testMode]);
    (0, _react.useEffect)(function () {
      return function () {
        var script = document.querySelector('#mollieScript');

        if (typeof script !== 'undefined' && script != null) {
          script.parentNode.removeChild(script);
          delete window.Mollie;
        }
      };
    }, []);

    if (!profileId) {
      return ErrorComponent ? /*#__PURE__*/_react.default.createElement(ErrorComponent, null, errMessage) : /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, errMessage);
    }

    return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_reactHelmetAsync.Helmet, {
      onChangeClientState: function onChangeClientState() {
        document.querySelector('#mollieScript').addEventListener('load', function () {
          return setScriptLoaded(true);
        });
      }
    }, /*#__PURE__*/_react.default.createElement("script", {
      id: "mollieScript",
      src: "https://js.mollie.com/v1/mollie.js",
      async: true
    })), mollie ? /*#__PURE__*/_react.default.createElement(WrappedComponent, (0, _extends2.default)({}, props, {
      config: config,
      mollie: mollie
    })) : LoaderComp && /*#__PURE__*/_react.default.createElement(LoaderComp, null));
  };
};

exports.withMollie = withMollie;
