// @deity/falcon-shop-data/dist/Customer/EditCustomerMutation.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.useEditCustomerMutation = exports.EditCustomerMutation = exports.EDIT_CUSTOMER = void 0;

var _inheritsLoose2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/inheritsLoose"));

var _client = /*#__PURE__*/require("@apollo/client");

var _falconData = /*#__PURE__*/require("@deity/falcon-data");

var EDIT_CUSTOMER = {
  "kind": "Document",
  "definitions": [{
    "kind": "OperationDefinition",
    "operation": "mutation",
    "name": {
      "kind": "Name",
      "value": "EditCustomer"
    },
    "variableDefinitions": [{
      "kind": "VariableDefinition",
      "variable": {
        "kind": "Variable",
        "name": {
          "kind": "Name",
          "value": "input"
        }
      },
      "type": {
        "kind": "NonNullType",
        "type": {
          "kind": "NamedType",
          "name": {
            "kind": "Name",
            "value": "EditCustomerInput"
          }
        }
      },
      "directives": []
    }],
    "directives": [],
    "selectionSet": {
      "kind": "SelectionSet",
      "selections": [{
        "kind": "Field",
        "name": {
          "kind": "Name",
          "value": "editCustomer"
        },
        "arguments": [{
          "kind": "Argument",
          "name": {
            "kind": "Name",
            "value": "input"
          },
          "value": {
            "kind": "Variable",
            "name": {
              "kind": "Name",
              "value": "input"
            }
          }
        }],
        "directives": [],
        "selectionSet": {
          "kind": "SelectionSet",
          "selections": [{
            "kind": "Field",
            "name": {
              "kind": "Name",
              "value": "id"
            },
            "arguments": [],
            "directives": []
          }]
        }
      }]
    }
  }],
  "loc": {
    "start": 0,
    "end": 108,
    "source": {
      "body": "\n  mutation EditCustomer($input: EditCustomerInput!) {\n    editCustomer(input: $input) {\n      id\n    }\n  }\n",
      "name": "GraphQL request",
      "locationOffset": {
        "line": 1,
        "column": 1
      }
    }
  }
};
exports.EDIT_CUSTOMER = EDIT_CUSTOMER;

var EditCustomerMutation = /*#__PURE__*/function (_Mutation) {
  (0, _inheritsLoose2.default)(EditCustomerMutation, _Mutation);

  function EditCustomerMutation() {
    return _Mutation.apply(this, arguments) || this;
  }

  return EditCustomerMutation;
}(_falconData.Mutation);

exports.EditCustomerMutation = EditCustomerMutation;
EditCustomerMutation.defaultProps = {
  mutation: EDIT_CUSTOMER,
  refetchQueries: ['Customer', 'CustomerWithAddresses'],
  awaitRefetchQueries: true
};

var useEditCustomerMutation = function useEditCustomerMutation(options) {
  if (options === void 0) {
    options = {};
  }

  return (0, _client.useMutation)(EDIT_CUSTOMER, Object.assign({
    refetchQueries: ['Customer', 'CustomerWithAddresses'],
    awaitRefetchQueries: true
  }, options));
};

exports.useEditCustomerMutation = useEditCustomerMutation;
