// @deity/falcon-front-kit/dist/NetworkStatus/NetworkStatus.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.NetworkStatus = void 0;

var _assertThisInitialized2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/assertThisInitialized"));

var _inheritsLoose2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/inheritsLoose"));

var _react = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("react"));

var NetworkStatus = /*#__PURE__*/function (_React$Component) {
  (0, _inheritsLoose2.default)(NetworkStatus, _React$Component);

  function NetworkStatus(props) {
    var _this;

    _this = _React$Component.call(this, props) || this;
    _this.updateOnLineStatus = _this.updateOnLineStatus.bind((0, _assertThisInitialized2.default)(_this));
    _this.state = {
      isOnline: true
    };
    return _this;
  }

  var _proto = NetworkStatus.prototype;

  _proto.componentDidMount = function componentDidMount() {
    window.addEventListener('online', this.updateOnLineStatus);
    window.addEventListener('offline', this.updateOnLineStatus); // run detection logic when component mounts as React 16 optimizes too much and doesn't
    // re-render even when SSR html doesn't match client-side
    // @see https://github.com/facebook/react/issues/10591

    this.updateOnLineStatus();
  };

  _proto.componentWillUnmount = function componentWillUnmount() {
    window.removeEventListener('online', this.updateOnLineStatus);
    window.removeEventListener('offline', this.updateOnLineStatus);
  };

  _proto.updateOnLineStatus = function updateOnLineStatus() {
    var _window = window,
        navigator = _window.navigator;
    var online = navigator && navigator.onLine;

    if (this.state.isOnline === online) {
      return;
    }

    this.setState({
      isOnline: online
    });
  };

  _proto.render = function render() {
    var children = this.props.children;
    var isOnline = this.state.isOnline;
    return children({
      isOnline: isOnline
    });
  };

  return NetworkStatus;
}(_react.default.Component);

exports.NetworkStatus = NetworkStatus;
