// @deity/falcon-front-kit/dist/Routes/OnlyUnauthenticatedRoute.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.OnlyUnauthenticatedRoute = void 0;

var _extends2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _inheritsLoose2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/inheritsLoose"));

var _react = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("react"));

var _propTypes = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("prop-types"));

var _reactRouterDom = /*#__PURE__*/require("react-router-dom");

var _falconData = /*#__PURE__*/require("@deity/falcon-data");

var _falconShopData = /*#__PURE__*/require("@deity/falcon-shop-data");

var OnlyUnauthenticatedRoute = /*#__PURE__*/function (_React$Component) {
  (0, _inheritsLoose2.default)(OnlyUnauthenticatedRoute, _React$Component);

  function OnlyUnauthenticatedRoute() {
    return _React$Component.apply(this, arguments) || this;
  }

  var _proto = OnlyUnauthenticatedRoute.prototype;

  _proto.render = function render() {
    var _this$props = this.props,
        component = _this$props.component,
        redirectTo = _this$props.redirectTo,
        rest = (0, _objectWithoutPropertiesLoose2.default)(_this$props, ["component", "redirectTo"]);
    var Component = component;
    return /*#__PURE__*/_react.default.createElement(_reactRouterDom.Route, (0, _extends2.default)({}, rest, {
      render: function render(props) {
        return /*#__PURE__*/_react.default.createElement(_falconShopData.IsAuthenticatedQuery, {
          passLoading: true
        }, function (_ref) {
          var data = _ref.data,
              loading = _ref.loading;

          if (loading) {
            // we can not render anything until we get know if the customer is authenticated or not
            return /*#__PURE__*/_react.default.createElement(_falconData.Loader, null);
          }

          if (data.customer) {
            return /*#__PURE__*/_react.default.createElement(_reactRouterDom.Redirect, {
              to: {
                pathname: redirectTo
              }
            });
          }

          return /*#__PURE__*/_react.default.createElement(Component, props);
        });
      }
    }));
  };

  return OnlyUnauthenticatedRoute;
}(_react.default.Component);

exports.OnlyUnauthenticatedRoute = OnlyUnauthenticatedRoute;
OnlyUnauthenticatedRoute.propTypes = /*#__PURE__*/Object.assign({}, _reactRouterDom.Route.propTypes, {
  redirectTo: _propTypes.default.string
});
OnlyUnauthenticatedRoute.defaultProps = {
  redirectTo: '/'
};
