// @deity/falcon-i18n/dist/T.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.T = void 0;

var _objectWithoutPropertiesLoose2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _inheritsLoose2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/inheritsLoose"));

var _react = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("react"));

var _propTypes = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("prop-types"));

var _context = /*#__PURE__*/require("./context");

// eslint-disable-next-line id-length
var T = /*#__PURE__*/function (_React$Component) {
  (0, _inheritsLoose2.default)(T, _React$Component);

  function T() {
    return _React$Component.apply(this, arguments) || this;
  }

  var _proto = T.prototype;

  _proto.render = function render() {
    var _this$props = this.props,
        id = _this$props.id,
        translationOptions = (0, _objectWithoutPropertiesLoose2.default)(_this$props, ["id"]);
    return /*#__PURE__*/_react.default.createElement(_context.I18nContext.Consumer, null, function (_ref) {
      var t = _ref.t;
      return t(id, translationOptions);
    });
  };

  return T;
}(_react.default.Component);

exports.T = T;
T.propTypes = {
  id: _propTypes.default.string.isRequired,
  defaultValue: /*#__PURE__*/_propTypes.default.oneOfType([_propTypes.default.string]),
  count: _propTypes.default.number,
  // eslint-disable-next-line react/forbid-prop-types
  context: _propTypes.default.any,
  // used for contexts (eg. male\female)
  replace: /*#__PURE__*/_propTypes.default.shape({}),
  // object with vars for interpolation - or put them directly in options
  lng: _propTypes.default.string,
  lngs: /*#__PURE__*/_propTypes.default.arrayOf(_propTypes.default.string),
  fallbackLng: /*#__PURE__*/_propTypes.default.oneOfType([_propTypes.default.string, /*#__PURE__*/_propTypes.default.arrayOf(_propTypes.default.string)]),
  ns: /*#__PURE__*/_propTypes.default.oneOfType([_propTypes.default.string, /*#__PURE__*/_propTypes.default.arrayOf(_propTypes.default.string)]),
  keySeparator: _propTypes.default.string,
  nsSeparator: _propTypes.default.string,
  returnObjects: _propTypes.default.bool,
  // accessing an object not a translation string
  joinArrays: _propTypes.default.string,
  // char, eg. '\n' that arrays will be joined by (can be set globally too)
  postProcess: /*#__PURE__*/_propTypes.default.oneOfType([_propTypes.default.string, /*#__PURE__*/_propTypes.default.arrayOf(_propTypes.default.string)]),
  // postProcessors to apply see interval plurals as a sample
  interpolation: /*#__PURE__*/_propTypes.default.shape({})
};
