// @deity/falcon-ui-kit/dist/Category/CategoryLayout.js
"use strict";

exports.__esModule = true;
exports.CategoryLayout = exports.CategoryArea = void 0;

var _falconUi = /*#__PURE__*/require("@deity/falcon-ui");

var _helpers = /*#__PURE__*/require("../helpers");

var CategoryArea = {
  heading: 'heading',
  filters: 'filters',
  content: 'content',
  footer: 'footer'
};
exports.CategoryArea = CategoryArea;
var CategoryLayout = /*#__PURE__*/(0, _falconUi.themed)({
  tag: _falconUi.Box,
  defaultTheme: {
    categoryLayout: {
      display: 'grid',
      gridGap: 'md',
      my: 'lg',
      // prettier-ignore
      gridTemplate: {
        xs: /*#__PURE__*/(0, _helpers.toGridTemplate)([['1fr'], [CategoryArea.heading], [CategoryArea.filters], [CategoryArea.content], [CategoryArea.footer]]),
        md: /*#__PURE__*/(0, _helpers.toGridTemplate)([['1fr', '3fr'], [CategoryArea.heading, CategoryArea.heading], [CategoryArea.filters, CategoryArea.content], [CategoryArea.footer, CategoryArea.footer]]),
        lg: /*#__PURE__*/(0, _helpers.toGridTemplate)([['1fr', '4fr'], [CategoryArea.heading, CategoryArea.heading], [CategoryArea.filters, CategoryArea.content], [CategoryArea.footer, CategoryArea.footer]])
      },
      css: {
        position: 'relative'
      },
      variants: {
        noFilters: {
          // prettier-ignore
          gridTemplate: /*#__PURE__*/(0, _helpers.toGridTemplate)([['1fr'], [CategoryArea.heading], [CategoryArea.content], [CategoryArea.footer]])
        }
      }
    }
  }
});
exports.CategoryLayout = CategoryLayout;
