// @deity/falcon-front-kit/dist/EnsureTTI/EnsureTTI.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.EnsureTTI = void 0;

var _inheritsLoose2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/inheritsLoose"));

var _react = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("react"));

var _propTypes = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("prop-types"));

/**
 * TTI measures the time from Navigation Start until the page's resources are loaded and the main thread is idle for at least 5 seconds
 */
var EnsureTTI = /*#__PURE__*/function (_React$Component) {
  (0, _inheritsLoose2.default)(EnsureTTI, _React$Component);

  function EnsureTTI(props) {
    var _this;

    _this = _React$Component.call(this, props) || this;

    _this.forceReady = function () {
      if (!_this.state.isReady) {
        _this.setState(function (x) {
          return Object.assign({}, x, {
            isReady: true
          });
        });
      }
    };

    _this.state = {
      isReady: false,
      setTimeoutHandlerId: undefined,
      requestIdleHandlerId: undefined
    };
    return _this;
  }

  var _proto = EnsureTTI.prototype;

  _proto.componentDidMount = function componentDidMount() {
    var _this2 = this;

    var timeout = this.props.timeout; // set isReady flag after timeout

    var setTimeoutHandlerId = window.setTimeout(function () {
      if (_this2.state.isReady) {
        return;
      }

      if ('requestIdleCallback' in window) {
        // @ts-ignore
        var requestIdleHandlerId = window.requestIdleCallback(_this2.forceReady);
        return _this2.setState(function (x) {
          return Object.assign({}, x, {
            requestIdleHandlerId: requestIdleHandlerId
          });
        });
      }

      return _this2.forceReady();
    }, timeout); // eslint-disable-next-line

    this.setState(function (x) {
      return Object.assign({}, x, {
        setTimeoutHandlerId: setTimeoutHandlerId
      });
    });
  };

  _proto.componentDidUpdate = function componentDidUpdate() {
    if (this.props.forceReady) {
      this.forceReady();
    }
  };

  _proto.componentWillUnmount = function componentWillUnmount() {
    var _this$state = this.state,
        setTimeoutHandlerId = _this$state.setTimeoutHandlerId,
        requestIdleHandlerId = _this$state.requestIdleHandlerId;
    window.clearTimeout(setTimeoutHandlerId);

    if ('cancelIdleCallback' in window && requestIdleHandlerId !== undefined) {
      // @ts-ignore
      window.cancelIdleCallback(requestIdleHandlerId);
    }
  }
  /** Sets isReady even before `timeout` */
  ;

  _proto.render = function render() {
    return this.props.children({
      isReady: this.state.isReady,
      forceReady: this.forceReady
    });
  };

  return EnsureTTI;
}(_react.default.Component);

exports.EnsureTTI = EnsureTTI;
EnsureTTI.propTypes = {
  timeout: _propTypes.default.number,
  children: _propTypes.default.func.isRequired
};
EnsureTTI.defaultProps = {
  timeout: 6000
};
