// @deity/falcon-logger/dist/format/minimal.js
"use strict";

var _interopRequireWildcard = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.default = void 0;

var _colors = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("./colors"));

var CONSTANTS = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./constants"));

var _utils = /*#__PURE__*/require("./utils");

var colorizer = /*#__PURE__*/(0, _colors.default)(true);
var messageKey = CONSTANTS.MESSAGE_KEY;
var errorLikeObjectKeys = CONSTANTS.ERROR_LIKE_KEYS;

var formatMessage = function formatMessage(_ref, inputData) {
  var IDENT = _ref.IDENT,
      EOL = _ref.EOL;
  var log;
  var line = '';

  if (!(0, _utils.isObject)(inputData)) {
    var parsed = (0, _utils.jsonParser)(inputData);
    log = parsed.value;

    if (parsed.err || !(0, _utils.isObject)(log)) {
      // pass through
      return inputData + EOL;
    }
  } else {
    log = inputData;
  } // Short-circuit for spec allowed primitive values.


  if ([null, true, false].includes(log)) {
    return log + EOL;
  }

  if (messageKey in log === false) return undefined;
  if (typeof log[messageKey] !== 'string') return undefined;
  line += log[messageKey] + EOL;

  if (log.type === 'Error' && (log.stack || log.extensions)) {
    var prettifiedErrorLog = '';

    if (log.stack) {
      prettifiedErrorLog = (0, _utils.prettifyErrorLog)({
        log: log,
        messageKey: messageKey,
        colorizer: colorizer,
        errorLikeKeys: errorLikeObjectKeys,
        errorProperties: [],
        ident: IDENT,
        eol: EOL
      });
    } else if (log.extensions) {
      prettifiedErrorLog = (0, _utils.prettifyGraphQLErrorLog)({
        log: log,
        colorizer: colorizer,
        ident: IDENT,
        eol: EOL
      });
    }

    line += prettifiedErrorLog;
  } else {
    var skipKeys = typeof log[messageKey] === 'string' ? [messageKey] : undefined;
    var prettifiedObject = (0, _utils.prettifyObject)({
      log: log,
      skipKeys: skipKeys,
      errorLikeKeys: errorLikeObjectKeys,
      eol: EOL,
      ident: IDENT
    });
    line += prettifiedObject;
  }

  return line;
};

var _default = function _default(_ref2) {
  var IDENT = _ref2.IDENT,
      EOL = _ref2.EOL;
  return function (inputData) {
    try {
      return formatMessage({
        IDENT: IDENT,
        EOL: EOL
      }, inputData);
    } catch (_unused) {
      return typeof inputData === 'string' ? inputData : JSON.stringify(inputData);
    }
  };
};

exports.default = _default;
