// @deity/falcon-ui-kit/dist/Filters/ColorFilter.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.ColorFilter = void 0;

var _extends2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("react"));

var _Color = /*#__PURE__*/require("../Color");

var _FilterItemLayout = /*#__PURE__*/require("./FilterItemLayout");

var _SelectedFilterItemLayout = /*#__PURE__*/require("./SelectedFilterItemLayout");

var _FilterItemListLayout = /*#__PURE__*/require("./FilterItemListLayout");

var ColorFilter = function ColorFilter(_ref) {
  var options = _ref.options,
      selected = _ref.selected,
      onChange = _ref.onChange,
      rest = (0, _objectWithoutPropertiesLoose2.default)(_ref, ["options", "selected", "onChange"]);
  var selectedOption = selected !== undefined ? options.find(function (x) {
    return x.value === selected;
  }) : undefined;
  return /*#__PURE__*/_react.default.createElement(_FilterItemListLayout.FilterItemListLayout, (0, _extends2.default)({
    display: "flex",
    flexWrap: "wrap"
  }, rest), selectedOption && /*#__PURE__*/_react.default.createElement(_SelectedFilterItemLayout.SelectedFilterItemLayout, {
    onClick: function onClick() {
      return onChange();
    }
  }, /*#__PURE__*/_react.default.createElement(_Color.ColorTile, {
    size: "lg",
    color: selectedOption.value,
    title: selectedOption.title
  })), !selectedOption && options.map(function (x) {
    return /*#__PURE__*/_react.default.createElement(_FilterItemLayout.FilterItemLayout, {
      key: x.value,
      onClick: function onClick() {
        return onChange(x.value);
      }
    }, /*#__PURE__*/_react.default.createElement(_Color.ColorTile, {
      size: "lg",
      color: x.value,
      title: x.title
    }));
  }));
};

exports.ColorFilter = ColorFilter;
