// @deity/falcon-ui-kit/dist/Filters/FilterDetails.js
"use strict";

var _interopRequireWildcard = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.FilterDetails = void 0;

var _extends2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("react"));

var _falconUi = /*#__PURE__*/require("@deity/falcon-ui");

var FilterDetails = function FilterDetails(_ref) {
  var initiallyOpen = _ref.initiallyOpen,
      children = _ref.children,
      rest = (0, _objectWithoutPropertiesLoose2.default)(_ref, ["initiallyOpen", "children"]);

  var _useState = (0, _react.useState)(initiallyOpen),
      isOpen = _useState[0],
      setIsOpen = _useState[1];

  return /*#__PURE__*/_react.default.createElement(_falconUi.Details, (0, _extends2.default)({
    open: isOpen
  }, rest, {
    defaultTheme: {
      filterDetails: {
        display: 'flex',
        flexDirection: 'column',
        css: function css(props) {
          return {
            '> :not(summary, style)': {
              display: props.open ? 'block' : 'none',
              flex: props.open ? '1' : 0
            },
            '> summary::-webkit-details-marker': {
              display: 'none'
            },
            '> summary:after': {
              display: 'block',
              content: props.open ? '"-"' : '"+"',
              marginLeft: props.theme.spacing.sm,
              fontSize: props.theme.fontSizes.md,
              lineHeight: 0.6,
              fontWeight: props.theme.fontWeights.bold,
              color: props.theme.colors.primary
            }
          };
        }
      }
    }
  }), children({
    set: setIsOpen,
    on: isOpen,
    toggle: function toggle(e) {
      if (e) {
        e.preventDefault();
      }

      setIsOpen(function (prev) {
        return !prev;
      });
    }
  }));
};

exports.FilterDetails = FilterDetails;
FilterDetails.defaultProps = {
  initiallyOpen: false
};
