// @deity/falcon-theme-editor/dist/ComponentsEditor.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.ComponentsEditor = void 0;

var _inheritsLoose2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/inheritsLoose"));

var _react = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("react"));

var _falconUi = /*#__PURE__*/require("@deity/falcon-ui");

var _thememeta = /*#__PURE__*/require("./thememeta");

function keys(object) {
  return Object.keys(object);
}

var ThemedPropEditor = function ThemedPropEditor(_ref) {
  var themeProp = _ref.themeProp,
      themeSection = _ref.themeSection,
      _onChange = _ref.onChange,
      propKey = _ref.propKey,
      propValue = _ref.propValue;
  var meta = _thememeta.themeMeta[themeProp];
  if (!meta) return null;
  return /*#__PURE__*/_react.default.createElement(_falconUi.Dropdown, {
    onChange: function onChange(value) {
      return _onChange(value.key, value.value);
    }
  }, /*#__PURE__*/_react.default.createElement(_falconUi.DropdownLabel, {
    css: {
      width: '100%'
    }
  }, /*#__PURE__*/_react.default.createElement(_falconUi.GridLayout, {
    gridTemplateColumns: "auto 1fr"
  }, /*#__PURE__*/_react.default.createElement(_falconUi.Box, null, propValue, /*#__PURE__*/_react.default.createElement(_falconUi.Box, {
    as: "span",
    fontSize: "xs",
    ml: "xs"
  }, "(", themeSection[propValue], meta.unit, ")")), meta.previewCss && /*#__PURE__*/_react.default.createElement(_falconUi.Box, {
    justifySelf: "end",
    bg: "secondaryLight",
    border: "regular",
    borderColor: "secondaryDark",
    size: "md",
    css: meta.previewCss(themeSection[propValue])
  }))), /*#__PURE__*/_react.default.createElement(_falconUi.DropdownMenu, null, keys(themeSection).map(function (key) {
    return /*#__PURE__*/_react.default.createElement(_falconUi.DropdownMenuItem, {
      key: key,
      value: {
        key: propKey,
        value: key
      },
      pl: "sm",
      pr: "xxl",
      color: key === propValue ? 'primaryText' : 'secondaryText',
      bg: key === propValue ? 'primary' : 'transparent'
    }, /*#__PURE__*/_react.default.createElement(_falconUi.GridLayout, {
      gridTemplateColumns: "auto 1fr"
    }, /*#__PURE__*/_react.default.createElement(_falconUi.Box, null, key, /*#__PURE__*/_react.default.createElement(_falconUi.Box, {
      as: "span",
      fontSize: "xs",
      ml: "xs"
    }, "(", themeSection[key], meta.unit, ")")), meta.previewCss && /*#__PURE__*/_react.default.createElement(_falconUi.Box, {
      justifySelf: "end",
      bg: "secondaryLight",
      border: "regular",
      borderColor: "secondaryDark",
      size: "md",
      css: meta.previewCss(themeSection[key])
    })));
  })));
};

var ComponentEditor = /*#__PURE__*/function (_React$Component) {
  (0, _inheritsLoose2.default)(ComponentEditor, _React$Component);

  function ComponentEditor() {
    return _React$Component.apply(this, arguments) || this;
  }

  var _proto = ComponentEditor.prototype;

  _proto.render = function render() {
    var _this$props = this.props,
        theme = _this$props.theme,
        themeProps = _this$props.themeProps,
        defaultThemeProps = _this$props.defaultThemeProps,
        _onChange3 = _this$props.onChange;
    var mergedProps = (0, _falconUi.mergeThemes)(defaultThemeProps, themeProps || {});
    return /*#__PURE__*/_react.default.createElement(_falconUi.Box, null, keys(_falconUi.mappings).filter(function (mapping) {
      return mergedProps[mapping] !== undefined;
    }).map(function (prop) {
      var _ref2 = _falconUi.mappings[prop],
          themeProp = _ref2.themeProp;
      var propValue = mergedProps[prop];
      return /*#__PURE__*/_react.default.createElement(_falconUi.GridLayout, {
        key: prop,
        gridTemplateColumns: "120px 1fr",
        mb: "md",
        alignItems: "center"
      }, /*#__PURE__*/_react.default.createElement(_falconUi.Text, {
        p: "none",
        ellipsis: true
      }, prop), typeof propValue === 'string' && (themeProp ? /*#__PURE__*/_react.default.createElement(ThemedPropEditor, {
        propKey: prop,
        propValue: propValue,
        themeProp: themeProp,
        themeSection: theme[themeProp],
        onChange: _onChange3
      }) : /*#__PURE__*/_react.default.createElement(_falconUi.Input, {
        border: "none",
        fontWeight: "bold",
        value: propValue,
        readOnly: true
      })), typeof propValue === 'object' && /*#__PURE__*/_react.default.createElement(_falconUi.GridLayout, {
        gridGap: "sm",
        border: "regular",
        borderColor: "secondary",
        borderRadius: "medium",
        p: "sm"
      }, keys(propValue).map(function (breakpointKey) {
        return /*#__PURE__*/_react.default.createElement(_falconUi.Box, null, /*#__PURE__*/_react.default.createElement(_falconUi.Text, {
          pb: "xs"
        }, breakpointKey, /*#__PURE__*/_react.default.createElement(_falconUi.Box, {
          as: "span",
          ml: "xs",
          fontSize: "xs"
        }, "(min-width: ", theme.breakpoints[breakpointKey], "px)")), themeProp ? /*#__PURE__*/_react.default.createElement(ThemedPropEditor, {
          propKey: prop,
          propValue: propValue[breakpointKey],
          themeProp: themeProp,
          themeSection: theme[themeProp],
          onChange: function onChange(key, value) {
            var _onChange2;

            _onChange3(key, (_onChange2 = {}, _onChange2[breakpointKey] = value, _onChange2));
          }
        }) : /*#__PURE__*/_react.default.createElement(_falconUi.Input, {
          border: "none",
          fontWeight: "bold",
          value: propValue[breakpointKey],
          readOnly: true
        }));
      })));
    }));
  };

  return ComponentEditor;
}(_react.default.Component);

var ComponentsEditor = function ComponentsEditor(_ref3) {
  var theme = _ref3.theme,
      selectedComponents = _ref3.selectedComponents,
      onComponentThemeChange = _ref3.onComponentThemeChange;
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, selectedComponents.map(function (selectedComponent) {
    var themeKey = Object.keys(selectedComponent.defaultTheme)[0];
    return /*#__PURE__*/_react.default.createElement(_falconUi.Box, {
      key: themeKey
    }, /*#__PURE__*/_react.default.createElement(_falconUi.H3, {
      mb: "md"
    }, themeKey, " theme"), /*#__PURE__*/_react.default.createElement(ComponentEditor, {
      defaultThemeProps: selectedComponent.defaultTheme[themeKey],
      themeProps: theme.components[themeKey],
      theme: theme,
      onChange: onComponentThemeChange(themeKey)
    }), selectedComponent.defaultTheme[themeKey].variants && /*#__PURE__*/_react.default.createElement(_falconUi.Box, null, Object.keys(selectedComponent.defaultTheme[themeKey].variants).map(function (variantKey) {
      return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, {
        key: variantKey
      }, /*#__PURE__*/_react.default.createElement(_falconUi.H3, {
        fontSize: "md",
        my: "md"
      }, "variant: ", variantKey), /*#__PURE__*/_react.default.createElement(ComponentEditor, {
        defaultThemeProps: selectedComponent.defaultTheme[themeKey].variants[variantKey],
        themeProps: theme.components[themeKey] && theme.components[themeKey].variants && theme.components[themeKey].variants[variantKey],
        theme: theme,
        onChange: onComponentThemeChange(themeKey, variantKey)
      }));
    })));
  }));
};

exports.ComponentsEditor = ComponentsEditor;
