// @deity/falcon-theme-editor/dist/Tabs.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.Tabs = void 0;

var _react = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("react"));

var _falconUi = /*#__PURE__*/require("@deity/falcon-ui");

var Tabs = function Tabs(props) {
  var items = props.items,
      active = props.active,
      onChange = props.onChange;
  return /*#__PURE__*/_react.default.createElement(_falconUi.FlexLayout, {
    pr: "sm",
    justifyContent: "space-between",
    css: {
      textAlign: 'center'
    }
  }, items.map(function (item) {
    return /*#__PURE__*/_react.default.createElement(_falconUi.Box, {
      key: item.key,
      css: {
        cursor: 'pointer'
      },
      onClick: function onClick() {
        return onChange(item.key);
      }
    }, /*#__PURE__*/_react.default.createElement(_falconUi.Icon, {
      src: item.key,
      stroke: item.key === active ? 'primary' : 'black',
      fill: item.key === active ? 'primary' : 'black',
      transitionTimingFunction: "easeIn",
      transitionDuration: "short",
      css: {
        strokeWidth: 2,
        transitionProperty: 'all'
      }
    }), /*#__PURE__*/_react.default.createElement(_falconUi.Text, {
      fontSize: "xs",
      color: item.key === active ? 'primary' : 'black'
    }, item.name));
  }));
};

exports.Tabs = Tabs;
