// @deity/falcon-theme-editor/dist/ThemeDownloader.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.ThemeDownloader = void 0;

var _react = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("react"));

var _deepObjectDiff = /*#__PURE__*/require("deep-object-diff");

var _falconUi = /*#__PURE__*/require("@deity/falcon-ui");

var CREATE_THEME_IMPORT = "import { createTheme } from '@deity/falcon-ui';";

function getThemeCustomizations(initialTheme, currentTheme) {
  var themeDiff = (0, _deepObjectDiff.diff)(initialTheme, currentTheme);
  return JSON.stringify(themeDiff, null, 2);
}

function withExports(customizationsSerialized) {
  return CREATE_THEME_IMPORT + "\n\nexport const theme = createTheme(" + customizationsSerialized + ");";
}

function copyToClipboard(serializedCustomizations) {
  window.navigator.clipboard.writeText(withExports(serializedCustomizations));
}

function download(serializedCustomizations) {
  var customizations = withExports(serializedCustomizations);
  var data = "data:text/json;charset=utf-8," + encodeURIComponent(customizations);
  var downloader = document.createElement('a');
  downloader.setAttribute('href', data);
  downloader.setAttribute('download', 'theme.js');
  downloader.click();
}

var ThemeDownloader = function ThemeDownloader(_ref) {
  var initialTheme = _ref.initialTheme,
      currentTheme = _ref.currentTheme;
  var serializedCustomizations = getThemeCustomizations(initialTheme, currentTheme);

  if (serializedCustomizations === '{}') {
    return /*#__PURE__*/_react.default.createElement(_falconUi.Text, {
      fontSize: "md"
    }, "Nothing to download yet, try making few customizations first!");
  }

  return /*#__PURE__*/_react.default.createElement(_falconUi.FlexLayout, {
    css: {
      height: '100%'
    },
    flexDirection: "column"
  }, /*#__PURE__*/_react.default.createElement(_falconUi.FlexLayout, {
    mb: "sm"
  }, /*#__PURE__*/_react.default.createElement(_falconUi.Button, {
    mr: "xs",
    flex: "1",
    onClick: function onClick() {
      return copyToClipboard(serializedCustomizations);
    }
  }, /*#__PURE__*/_react.default.createElement(_falconUi.Icon, {
    stroke: "white",
    src: "copy",
    size: "md",
    mr: "sm"
  }), "Copy"), /*#__PURE__*/_react.default.createElement(_falconUi.Button, {
    ml: "xs",
    flex: "1",
    onClick: function onClick() {
      return download(serializedCustomizations);
    }
  }, /*#__PURE__*/_react.default.createElement(_falconUi.Icon, {
    size: "md",
    mr: "sm",
    stroke: "white",
    css: {
      strokeWidth: '4'
    },
    src: "download"
  }), "Download")), /*#__PURE__*/_react.default.createElement(_falconUi.Input, {
    as: "textarea",
    flex: "1",
    spellCheck: false,
    readOnly: true,
    value: serializedCustomizations
  }));
};

exports.ThemeDownloader = ThemeDownloader;
