// @deity/falcon-ui-kit/dist/Blog/BlogPostExcerptLayout.js
"use strict";

exports.__esModule = true;
exports.BlogPostExcerptLayout = exports.BlogPostExcerptArea = void 0;

var _falconUi = /*#__PURE__*/require("@deity/falcon-ui");

var _helpers = /*#__PURE__*/require("../helpers");

var BlogPostExcerptArea = {
  image: 'image',
  title: 'title',
  date: 'date',
  excerpt: 'excerpt',
  readMore: 'readMore'
};
exports.BlogPostExcerptArea = BlogPostExcerptArea;
var BlogPostExcerptLayout = /*#__PURE__*/(0, _falconUi.themed)({
  tag: _falconUi.Link,
  defaultTheme: {
    blogPostExcerptLayout: {
      display: 'grid',
      gridRowGap: 'xs',
      gridColumnGap: 'lg',
      color: 'black',
      gridTemplate: {
        xs: /*#__PURE__*/(0, _helpers.toGridTemplate)([['1fr'], [BlogPostExcerptArea.image], [BlogPostExcerptArea.date], [BlogPostExcerptArea.title], [BlogPostExcerptArea.excerpt], [BlogPostExcerptArea.readMore]])
      },
      css: {
        textDecoration: 'none'
      }
    }
  }
});
exports.BlogPostExcerptLayout = BlogPostExcerptLayout;
