// @deity/falcon-ui-kit/dist/SEO/BlogPostSchema.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.BlogPostSchema = void 0;

var _react = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("react"));

var _JsonLd = /*#__PURE__*/require("./JsonLd");

function getWordCount(string) {
  return string.trim().split(/\s+/).length;
}

function addBackslash(slug) {
  return slug[0] === '/' ? slug : "/" + slug;
}

/**
 * Get the schema data for a blog.
 * @param  blog
 */
var getBlogSchemaData = function getBlogSchemaData(blog) {
  var _blog$image, _blog$author;

  if (!blog) {
    return null;
  }

  var schema = {
    '@context': 'https://schema.org',
    '@type': 'BlogPosting',
    headline: blog.title,
    dateCreated: blog.date,
    dateModified: blog.modified,
    description: blog.excerpt
  };

  if (blog != null && blog.content) {
    schema.articleBody = blog.content;
    schema.wordcount = getWordCount(blog.content);
  }

  if (blog != null && (_blog$image = blog.image) != null && _blog$image.url) {
    schema.image = blog.image.url;
  }

  if (blog != null && blog.slug) {
    schema.url = addBackslash(blog.slug);
  }

  if (blog != null && (_blog$author = blog.author) != null && _blog$author.name) {
    schema.author = {
      '@type': 'Person',
      name: blog.author.name
    };
  }

  return schema;
};

var BlogPostSchema = function BlogPostSchema(_ref) {
  var blog = _ref.blog;
  var schemaData = getBlogSchemaData(blog);
  return /*#__PURE__*/_react.default.createElement(_JsonLd.JsonLd, {
    item: schemaData
  });
};

exports.BlogPostSchema = BlogPostSchema;
