// @deity/falcon-data/dist/Url/UrlQuery.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.useUrlLazyQuery = exports.useUrlQuery = exports.UrlQuery = exports.GET_URL = void 0;

var _inheritsLoose2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/inheritsLoose"));

var _propTypes = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("prop-types"));

var _client = /*#__PURE__*/require("@apollo/client");

var _Query2 = /*#__PURE__*/require("../Query/Query");

var GET_URL = {
  "kind": "Document",
  "definitions": [{
    "kind": "OperationDefinition",
    "operation": "query",
    "name": {
      "kind": "Name",
      "value": "Url"
    },
    "variableDefinitions": [{
      "kind": "VariableDefinition",
      "variable": {
        "kind": "Variable",
        "name": {
          "kind": "Name",
          "value": "path"
        }
      },
      "type": {
        "kind": "NonNullType",
        "type": {
          "kind": "NamedType",
          "name": {
            "kind": "Name",
            "value": "String"
          }
        }
      },
      "directives": []
    }],
    "directives": [],
    "selectionSet": {
      "kind": "SelectionSet",
      "selections": [{
        "kind": "Field",
        "name": {
          "kind": "Name",
          "value": "url"
        },
        "arguments": [{
          "kind": "Argument",
          "name": {
            "kind": "Name",
            "value": "path"
          },
          "value": {
            "kind": "Variable",
            "name": {
              "kind": "Name",
              "value": "path"
            }
          }
        }],
        "directives": [],
        "selectionSet": {
          "kind": "SelectionSet",
          "selections": [{
            "kind": "Field",
            "name": {
              "kind": "Name",
              "value": "__typename"
            },
            "arguments": [],
            "directives": []
          }, {
            "kind": "InlineFragment",
            "typeCondition": {
              "kind": "NamedType",
              "name": {
                "kind": "Name",
                "value": "ResourceMeta"
              }
            },
            "directives": [],
            "selectionSet": {
              "kind": "SelectionSet",
              "selections": [{
                "kind": "Field",
                "name": {
                  "kind": "Name",
                  "value": "path"
                },
                "arguments": [],
                "directives": []
              }, {
                "kind": "Field",
                "name": {
                  "kind": "Name",
                  "value": "type"
                },
                "arguments": [],
                "directives": []
              }, {
                "kind": "Field",
                "name": {
                  "kind": "Name",
                  "value": "id"
                },
                "arguments": [],
                "directives": []
              }]
            }
          }, {
            "kind": "InlineFragment",
            "typeCondition": {
              "kind": "NamedType",
              "name": {
                "kind": "Name",
                "value": "RedirectMeta"
              }
            },
            "directives": [],
            "selectionSet": {
              "kind": "SelectionSet",
              "selections": [{
                "kind": "Field",
                "name": {
                  "kind": "Name",
                  "value": "path"
                },
                "arguments": [],
                "directives": []
              }, {
                "kind": "Field",
                "name": {
                  "kind": "Name",
                  "value": "status"
                },
                "arguments": [],
                "directives": []
              }]
            }
          }]
        }
      }]
    }
  }],
  "loc": {
    "start": 0,
    "end": 218,
    "source": {
      "body": "\n  query Url($path: String!) {\n    url(path: $path) {\n      __typename\n      ... on ResourceMeta {\n        path\n        type\n        id\n      }\n      ... on RedirectMeta {\n        path\n        status\n      }\n    }\n  }\n",
      "name": "GraphQL request",
      "locationOffset": {
        "line": 1,
        "column": 1
      }
    }
  }
};
exports.GET_URL = GET_URL;

var UrlQuery = /*#__PURE__*/function (_Query) {
  (0, _inheritsLoose2.default)(UrlQuery, _Query);

  function UrlQuery() {
    return _Query.apply(this, arguments) || this;
  }

  return UrlQuery;
}(_Query2.Query);

exports.UrlQuery = UrlQuery;
UrlQuery.defaultProps = {
  query: GET_URL
};
UrlQuery.propTypes = /*#__PURE__*/Object.assign({}, _Query2.Query.propTypes, {
  variables: /*#__PURE__*/_propTypes.default.shape({
    path: _propTypes.default.string.isRequired
  })
});

var useUrlQuery = function useUrlQuery(options) {
  return (0, _client.useQuery)(GET_URL, options);
};

exports.useUrlQuery = useUrlQuery;

var useUrlLazyQuery = function useUrlLazyQuery(options) {
  return (0, _client.useLazyQuery)(GET_URL, options);
};

exports.useUrlLazyQuery = useUrlLazyQuery;
