// @deity/falcon-shop-data/dist/Checkout/PlaceOrderMutation.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.usePlaceOrderMutation = exports.PlaceOrderMutation = exports.PLACE_ORDER = void 0;

var _inheritsLoose2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/inheritsLoose"));

var _client = /*#__PURE__*/require("@apollo/client");

var _falconData = /*#__PURE__*/require("@deity/falcon-data");

var PLACE_ORDER = {
  "kind": "Document",
  "definitions": [{
    "kind": "OperationDefinition",
    "operation": "mutation",
    "name": {
      "kind": "Name",
      "value": "PlaceOrder"
    },
    "variableDefinitions": [{
      "kind": "VariableDefinition",
      "variable": {
        "kind": "Variable",
        "name": {
          "kind": "Name",
          "value": "input"
        }
      },
      "type": {
        "kind": "NamedType",
        "name": {
          "kind": "Name",
          "value": "PlaceOrderInput"
        }
      },
      "directives": []
    }],
    "directives": [],
    "selectionSet": {
      "kind": "SelectionSet",
      "selections": [{
        "kind": "Field",
        "name": {
          "kind": "Name",
          "value": "placeOrder"
        },
        "arguments": [{
          "kind": "Argument",
          "name": {
            "kind": "Name",
            "value": "input"
          },
          "value": {
            "kind": "Variable",
            "name": {
              "kind": "Name",
              "value": "input"
            }
          }
        }],
        "directives": [],
        "selectionSet": {
          "kind": "SelectionSet",
          "selections": [{
            "kind": "Field",
            "name": {
              "kind": "Name",
              "value": "__typename"
            },
            "arguments": [],
            "directives": []
          }, {
            "kind": "InlineFragment",
            "typeCondition": {
              "kind": "NamedType",
              "name": {
                "kind": "Name",
                "value": "Order"
              }
            },
            "directives": [],
            "selectionSet": {
              "kind": "SelectionSet",
              "selections": [{
                "kind": "Field",
                "name": {
                  "kind": "Name",
                  "value": "id"
                },
                "arguments": [],
                "directives": []
              }, {
                "kind": "Field",
                "name": {
                  "kind": "Name",
                  "value": "referenceNo"
                },
                "arguments": [],
                "directives": []
              }]
            }
          }, {
            "kind": "InlineFragment",
            "typeCondition": {
              "kind": "NamedType",
              "name": {
                "kind": "Name",
                "value": "PlaceOrderIntermediateResult"
              }
            },
            "directives": [],
            "selectionSet": {
              "kind": "SelectionSet",
              "selections": [{
                "kind": "Field",
                "name": {
                  "kind": "Name",
                  "value": "url"
                },
                "arguments": [],
                "directives": []
              }, {
                "kind": "Field",
                "name": {
                  "kind": "Name",
                  "value": "method"
                },
                "arguments": [],
                "directives": []
              }, {
                "kind": "Field",
                "name": {
                  "kind": "Name",
                  "value": "fields"
                },
                "arguments": [],
                "directives": [],
                "selectionSet": {
                  "kind": "SelectionSet",
                  "selections": [{
                    "kind": "Field",
                    "name": {
                      "kind": "Name",
                      "value": "name"
                    },
                    "arguments": [],
                    "directives": []
                  }, {
                    "kind": "Field",
                    "name": {
                      "kind": "Name",
                      "value": "value"
                    },
                    "arguments": [],
                    "directives": []
                  }]
                }
              }]
            }
          }]
        }
      }]
    }
  }],
  "loc": {
    "start": 0,
    "end": 306,
    "source": {
      "body": "\n  mutation PlaceOrder($input: PlaceOrderInput) {\n    placeOrder(input: $input) {\n      __typename\n      ... on Order {\n        id\n        referenceNo\n      }\n      ... on PlaceOrderIntermediateResult {\n        url\n        method\n        fields {\n          name\n          value\n        }\n      }\n    }\n  }\n",
      "name": "GraphQL request",
      "locationOffset": {
        "line": 1,
        "column": 1
      }
    }
  }
};
exports.PLACE_ORDER = PLACE_ORDER;

var PlaceOrderMutation = /*#__PURE__*/function (_Mutation) {
  (0, _inheritsLoose2.default)(PlaceOrderMutation, _Mutation);

  function PlaceOrderMutation() {
    return _Mutation.apply(this, arguments) || this;
  }

  return PlaceOrderMutation;
}(_falconData.Mutation);

exports.PlaceOrderMutation = PlaceOrderMutation;
PlaceOrderMutation.defaultProps = {
  mutation: PLACE_ORDER,
  refetchQueries: ['Cart', 'MiniCart', 'OrderList'],
  awaitRefetchQueries: true
};

var usePlaceOrderMutation = function usePlaceOrderMutation(options) {
  return (0, _client.useMutation)(PLACE_ORDER, Object.assign({
    refetchQueries: ['Cart', 'MiniCart', 'OrderList'],
    awaitRefetchQueries: true
  }, options || {}));
};

exports.usePlaceOrderMutation = usePlaceOrderMutation;
