// @deity/falcon-shop-data/dist/Checkout/SetPaymentMethodMutation.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.useSetPaymentMethodMutation = exports.SetPaymentMethodMutation = exports.SET_PAYMENT_METHOD = void 0;

var _inheritsLoose2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/inheritsLoose"));

var _client = /*#__PURE__*/require("@apollo/client");

var _falconData = /*#__PURE__*/require("@deity/falcon-data");

var SET_PAYMENT_METHOD = {
  "kind": "Document",
  "definitions": [{
    "kind": "OperationDefinition",
    "operation": "mutation",
    "name": {
      "kind": "Name",
      "value": "SetPaymentMethod"
    },
    "variableDefinitions": [{
      "kind": "VariableDefinition",
      "variable": {
        "kind": "Variable",
        "name": {
          "kind": "Name",
          "value": "input"
        }
      },
      "type": {
        "kind": "NonNullType",
        "type": {
          "kind": "NamedType",
          "name": {
            "kind": "Name",
            "value": "CheckoutDetailsInput"
          }
        }
      },
      "directives": []
    }],
    "directives": [],
    "selectionSet": {
      "kind": "SelectionSet",
      "selections": [{
        "kind": "Field",
        "name": {
          "kind": "Name",
          "value": "setPaymentMethod"
        },
        "arguments": [{
          "kind": "Argument",
          "name": {
            "kind": "Name",
            "value": "input"
          },
          "value": {
            "kind": "Variable",
            "name": {
              "kind": "Name",
              "value": "input"
            }
          }
        }],
        "directives": [],
        "selectionSet": {
          "kind": "SelectionSet",
          "selections": [{
            "kind": "Field",
            "name": {
              "kind": "Name",
              "value": "provider"
            },
            "arguments": [],
            "directives": []
          }, {
            "kind": "Field",
            "name": {
              "kind": "Name",
              "value": "method"
            },
            "arguments": [],
            "directives": []
          }, {
            "kind": "Field",
            "name": {
              "kind": "Name",
              "value": "config"
            },
            "arguments": [],
            "directives": []
          }, {
            "kind": "Field",
            "name": {
              "kind": "Name",
              "value": "surcharge"
            },
            "arguments": [],
            "directives": [],
            "selectionSet": {
              "kind": "SelectionSet",
              "selections": [{
                "kind": "Field",
                "name": {
                  "kind": "Name",
                  "value": "amount"
                },
                "arguments": [],
                "directives": []
              }, {
                "kind": "Field",
                "name": {
                  "kind": "Name",
                  "value": "percentage"
                },
                "arguments": [],
                "directives": []
              }, {
                "kind": "Field",
                "name": {
                  "kind": "Name",
                  "value": "fixed"
                },
                "arguments": [],
                "directives": []
              }]
            }
          }]
        }
      }]
    }
  }],
  "loc": {
    "start": 0,
    "end": 225,
    "source": {
      "body": "\n  mutation SetPaymentMethod($input: CheckoutDetailsInput!) {\n    setPaymentMethod(input: $input) {\n      provider\n      method\n      config\n      surcharge {\n        amount\n        percentage\n        fixed\n      }\n    }\n  }\n",
      "name": "GraphQL request",
      "locationOffset": {
        "line": 1,
        "column": 1
      }
    }
  }
};
exports.SET_PAYMENT_METHOD = SET_PAYMENT_METHOD;

var SetPaymentMethodMutation = /*#__PURE__*/function (_Mutation) {
  (0, _inheritsLoose2.default)(SetPaymentMethodMutation, _Mutation);

  function SetPaymentMethodMutation() {
    return _Mutation.apply(this, arguments) || this;
  }

  return SetPaymentMethodMutation;
}(_falconData.Mutation);

exports.SetPaymentMethodMutation = SetPaymentMethodMutation;
SetPaymentMethodMutation.defaultProps = {
  mutation: SET_PAYMENT_METHOD,
  awaitRefetchQueries: true,
  refetchQueries: ['Cart', 'CheckoutState']
};

var useSetPaymentMethodMutation = function useSetPaymentMethodMutation(options) {
  return (0, _client.useMutation)(SET_PAYMENT_METHOD, Object.assign({
    refetchQueries: ['Cart', 'CheckoutState'],
    awaitRefetchQueries: true
  }, options || {}));
};

exports.useSetPaymentMethodMutation = useSetPaymentMethodMutation;
