// @deity/falcon-shop-data/dist/Checkout/SetShippingMethodMutation.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.useSetShippingMethodMutation = exports.SetShippingMethodMutation = exports.SET_SHIPPING_METHOD = void 0;

var _inheritsLoose2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/inheritsLoose"));

var _client = /*#__PURE__*/require("@apollo/client");

var _falconData = /*#__PURE__*/require("@deity/falcon-data");

var SET_SHIPPING_METHOD = {
  "kind": "Document",
  "definitions": [{
    "kind": "OperationDefinition",
    "operation": "mutation",
    "name": {
      "kind": "Name",
      "value": "SetShippingMethod"
    },
    "variableDefinitions": [{
      "kind": "VariableDefinition",
      "variable": {
        "kind": "Variable",
        "name": {
          "kind": "Name",
          "value": "input"
        }
      },
      "type": {
        "kind": "NonNullType",
        "type": {
          "kind": "NamedType",
          "name": {
            "kind": "Name",
            "value": "CheckoutDetailsInput"
          }
        }
      },
      "directives": []
    }],
    "directives": [],
    "selectionSet": {
      "kind": "SelectionSet",
      "selections": [{
        "kind": "Field",
        "name": {
          "kind": "Name",
          "value": "setShippingMethod"
        },
        "arguments": [{
          "kind": "Argument",
          "name": {
            "kind": "Name",
            "value": "input"
          },
          "value": {
            "kind": "Variable",
            "name": {
              "kind": "Name",
              "value": "input"
            }
          }
        }],
        "directives": []
      }]
    }
  }],
  "loc": {
    "start": 0,
    "end": 104,
    "source": {
      "body": "\n  mutation SetShippingMethod($input: CheckoutDetailsInput!) {\n    setShippingMethod(input: $input)\n  }\n",
      "name": "GraphQL request",
      "locationOffset": {
        "line": 1,
        "column": 1
      }
    }
  }
};
exports.SET_SHIPPING_METHOD = SET_SHIPPING_METHOD;

var SetShippingMethodMutation = /*#__PURE__*/function (_Mutation) {
  (0, _inheritsLoose2.default)(SetShippingMethodMutation, _Mutation);

  function SetShippingMethodMutation() {
    return _Mutation.apply(this, arguments) || this;
  }

  return SetShippingMethodMutation;
}(_falconData.Mutation);

exports.SetShippingMethodMutation = SetShippingMethodMutation;
SetShippingMethodMutation.defaultProps = {
  mutation: SET_SHIPPING_METHOD,
  refetchQueries: ['Cart', 'CheckoutState'],
  awaitRefetchQueries: true
};

var useSetShippingMethodMutation = function useSetShippingMethodMutation(options) {
  return (0, _client.useMutation)(SET_SHIPPING_METHOD, Object.assign({
    refetchQueries: ['Cart', 'CheckoutState'],
    awaitRefetchQueries: true
  }, options || {}));
};

exports.useSetShippingMethodMutation = useSetShippingMethodMutation;
