// @deity/falcon-ui-kit/dist/Filters/MultipleFilter.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.MultipleFilter = void 0;

var _react = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("react"));

var _falconUi = /*#__PURE__*/require("@deity/falcon-ui");

var _FilterItemLayout = /*#__PURE__*/require("./FilterItemLayout");

var _FilterItemListLayout = /*#__PURE__*/require("./FilterItemListLayout");

var MultipleFilter = function MultipleFilter(_ref) {
  var options = _ref.options,
      _ref$selected = _ref.selected,
      selected = _ref$selected === void 0 ? [] : _ref$selected,
      onChange = _ref.onChange;

  var handleOnChange = function handleOnChange(option, isSelected) {
    if (isSelected) {
      onChange([].concat(selected, [option.value]));
    } else {
      onChange(selected.filter(function (value) {
        return value !== option.value;
      }));
    }
  };

  return /*#__PURE__*/_react.default.createElement(_FilterItemListLayout.FilterItemListLayout, null, options.map(function (x) {
    return /*#__PURE__*/_react.default.createElement(_FilterItemLayout.FilterItemLayout, {
      key: x.value
    }, /*#__PURE__*/_react.default.createElement(_falconUi.Checkbox, {
      id: x.title + "-" + x.value,
      checked: selected.some(function (value) {
        return value === x.value;
      }),
      onChange: function onChange(e) {
        return handleOnChange(x, e.target.checked);
      }
    }), /*#__PURE__*/_react.default.createElement(_falconUi.Label, {
      ml: "sm",
      htmlFor: x.title + "-" + x.value
    }, x.title, " (", x.count, ")"));
  }));
};

exports.MultipleFilter = MultipleFilter;
