// @deity/falcon-ui-kit/dist/Filters/SelectedFilterItemLayout.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.SelectedFilterItemLayout = void 0;

var _objectWithoutPropertiesLoose2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("react"));

var _falconUi = /*#__PURE__*/require("@deity/falcon-ui");

var SelectedFilterItemLayoutInnerDOM = function SelectedFilterItemLayoutInnerDOM(_ref) {
  var _onClick = _ref.onClick,
      children = _ref.children,
      rest = (0, _objectWithoutPropertiesLoose2.default)(_ref, ["onClick", "children"]);
  return /*#__PURE__*/_react.default.createElement(_falconUi.ListItem, rest, /*#__PURE__*/_react.default.createElement(_falconUi.Icon, {
    src: "close",
    size: "md",
    mr: "xs",
    onClick: function onClick() {
      return _onClick && _onClick();
    },
    css: function css(_ref2) {
      var theme = _ref2.theme;
      return {
        cursor: _onClick ? 'pointer' : undefined,
        ':hover': {
          stroke: theme.colors.primaryLight
        }
      };
    }
  }), children);
};

var SelectedFilterItemLayout = /*#__PURE__*/(0, _falconUi.themed)({
  tag: SelectedFilterItemLayoutInnerDOM,
  defaultTheme: {
    selectedFilterItemLayout: {
      display: 'flex',
      alignItems: 'center',
      flexWrap: 'wrap',
      pr: 'xs',
      pb: 'xs'
    }
  }
});
exports.SelectedFilterItemLayout = SelectedFilterItemLayout;
