// @deity/falcon-ui-kit/dist/Error/ErrorSummary.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.ErrorSummary = void 0;

var _objectWithoutPropertiesLoose2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("react"));

var _propTypes = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("prop-types"));

var _falconUi = /*#__PURE__*/require("@deity/falcon-ui");

var _ErrorListLayout = /*#__PURE__*/require("./ErrorListLayout");

var _Error = /*#__PURE__*/require("./Error");

var ErrorSummary = function ErrorSummary(_ref) {
  var errors = _ref.errors,
      rest = (0, _objectWithoutPropertiesLoose2.default)(_ref, ["errors"]);

  if (!errors) {
    return null;
  }

  errors = Array.isArray(errors) ? errors : [errors];
  return /*#__PURE__*/_react.default.createElement(_ErrorListLayout.ErrorListLayout, rest, errors.map(function (error) {
    return /*#__PURE__*/_react.default.createElement(_Error.Error, {
      as: _falconUi.ListItem,
      key: error.message,
      insights: error
    }, error.message);
  }));
};

exports.ErrorSummary = ErrorSummary;

var errorType = /*#__PURE__*/_propTypes.default.shape({
  message: _propTypes.default.string.isRequired,
  code: _propTypes.default.string.isRequired
}).isRequired;

ErrorSummary.propTypes = {
  // @ts-ignore https://github.com/DefinitelyTyped/DefinitelyTyped/issues/34692
  errors: /*#__PURE__*/_propTypes.default.oneOfType([errorType, /*#__PURE__*/_propTypes.default.arrayOf(errorType).isRequired])
};
