// @deity/falcon-ui-kit/dist/Error/NotFound.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.NotFound = void 0;

var _react = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("react"));

var _reactRouterDom = /*#__PURE__*/require("react-router-dom");

var _propTypes = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("prop-types"));

var _falconI18n = /*#__PURE__*/require("@deity/falcon-i18n");

var _falconUi = /*#__PURE__*/require("@deity/falcon-ui");

var _Layouts = /*#__PURE__*/require("../Layouts");

var NotFound = function NotFound(_ref) {
  var location = _ref.location;
  var pathname = location.pathname;
  var path = pathname.startsWith('/') ? pathname.substring(1) : pathname;
  return /*#__PURE__*/_react.default.createElement(_Layouts.PageLayout, null, /*#__PURE__*/_react.default.createElement(_falconUi.H1, null, /*#__PURE__*/_react.default.createElement(_falconI18n.T, {
    id: "notFound.title"
  })), /*#__PURE__*/_react.default.createElement(_falconUi.FlexLayout, {
    flexDirection: "column",
    alignItems: "center",
    p: "sm"
  }, /*#__PURE__*/_react.default.createElement(_falconUi.Text, {
    fontSize: "md",
    mb: "xs"
  }, /*#__PURE__*/_react.default.createElement(_falconI18n.T, {
    id: "notFound.text",
    path: path
  })), /*#__PURE__*/_react.default.createElement(_falconUi.Button, {
    as: _reactRouterDom.Link,
    to: "/",
    p: "xs"
  }, /*#__PURE__*/_react.default.createElement(_falconI18n.T, {
    id: "notFound.goHomeButton"
  }))));
};

exports.NotFound = NotFound;
NotFound.propTypes = {
  location: _propTypes.default.any.isRequired
};
