// @deity/falcon-stripe-plugin/dist/StripePlugin.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.StripePlugin = void 0;

var _regenerator = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/asyncToGenerator"));

var _Loader = /*#__PURE__*/require("./Loader");

/**
 * Testing docs - https://stripe.com/docs/testing#cards
 */
var StripeCard = function StripeCard(props) {
  var stripe = props.stripe,
      config = props.config,
      children = props.children;
  var clientSecret = config.clientSecret;
  var cardElement = props.elements.getElement('card');

  var payFn = /*#__PURE__*/function () {
    var _ref = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee() {
      return _regenerator.default.wrap(function _callee$(_context) {
        while (1) {
          switch (_context.prev = _context.next) {
            case 0:
              return _context.abrupt("return", stripe.confirmCardPayment(clientSecret, {
                payment_method: {
                  card: cardElement
                }
              }).then(function (_ref2) {
                var paymentIntent = _ref2.paymentIntent,
                    error = _ref2.error;

                if (error) {
                  return Promise.reject(new Error(error.message));
                }

                if (paymentIntent && paymentIntent.status === 'succeeded') {
                  return {
                    intentId: paymentIntent.id
                  };
                }

                throw new Error("Payment failed. Status: " + paymentIntent.status);
              }));

            case 1:
            case "end":
              return _context.stop();
          }
        }
      }, _callee);
    }));

    return function payFn() {
      return _ref.apply(this, arguments);
    };
  }();

  return children(payFn);
};

var StripePlugin = /*#__PURE__*/(0, _Loader.withStripe)(StripeCard);
exports.StripePlugin = StripePlugin;
