// @deity/falcon-shop-data/dist/SignIn/SignOutMutation.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.useSignOutMutation = exports.SignOutMutation = exports.SIGN_OUT = void 0;

var _inheritsLoose2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/inheritsLoose"));

var _client = /*#__PURE__*/require("@apollo/client");

var _falconData = /*#__PURE__*/require("@deity/falcon-data");

var _Customer = /*#__PURE__*/require("../Customer");

var SIGN_OUT = {
  "kind": "Document",
  "definitions": [{
    "kind": "OperationDefinition",
    "operation": "mutation",
    "name": {
      "kind": "Name",
      "value": "SignOut"
    },
    "variableDefinitions": [],
    "directives": [],
    "selectionSet": {
      "kind": "SelectionSet",
      "selections": [{
        "kind": "Field",
        "name": {
          "kind": "Name",
          "value": "signOut"
        },
        "arguments": [],
        "directives": []
      }]
    }
  }],
  "loc": {
    "start": 0,
    "end": 38,
    "source": {
      "body": "\n  mutation SignOut {\n    signOut\n  }\n",
      "name": "GraphQL request",
      "locationOffset": {
        "line": 1,
        "column": 1
      }
    }
  }
};
exports.SIGN_OUT = SIGN_OUT;
var defaultOptions = {
  refetchQueries: ['Customer', 'Cart', 'MiniCart'],
  awaitRefetchQueries: false,
  update: function update(store) {
    var data = store.readQuery({
      query: _Customer.GET_IS_AUTHENTICATED
    });
    store.writeQuery({
      query: _Customer.GET_IS_AUTHENTICATED,
      data: Object.assign({}, data, {
        customer: null
      })
    });
  }
};

var SignOutMutation = /*#__PURE__*/function (_Mutation) {
  (0, _inheritsLoose2.default)(SignOutMutation, _Mutation);

  function SignOutMutation() {
    return _Mutation.apply(this, arguments) || this;
  }

  return SignOutMutation;
}(_falconData.Mutation);

exports.SignOutMutation = SignOutMutation;
SignOutMutation.defaultProps = /*#__PURE__*/Object.assign({
  mutation: SIGN_OUT
}, defaultOptions);

var useSignOutMutation = function useSignOutMutation(options) {
  if (options === void 0) {
    options = {};
  }

  return (0, _client.useMutation)(SIGN_OUT, Object.assign({}, defaultOptions, options));
};

exports.useSignOutMutation = useSignOutMutation;
