// @deity/falcon-shop-data/dist/SignIn/SignInMutation.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.useSignInMutation = exports.SignInMutation = exports.SIGN_IN = void 0;

var _inheritsLoose2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/inheritsLoose"));

var _client = /*#__PURE__*/require("@apollo/client");

var _falconData = /*#__PURE__*/require("@deity/falcon-data");

var _Customer = /*#__PURE__*/require("../Customer");

var SIGN_IN = {
  "kind": "Document",
  "definitions": [{
    "kind": "OperationDefinition",
    "operation": "mutation",
    "name": {
      "kind": "Name",
      "value": "SignIn"
    },
    "variableDefinitions": [{
      "kind": "VariableDefinition",
      "variable": {
        "kind": "Variable",
        "name": {
          "kind": "Name",
          "value": "input"
        }
      },
      "type": {
        "kind": "NonNullType",
        "type": {
          "kind": "NamedType",
          "name": {
            "kind": "Name",
            "value": "SignInInput"
          }
        }
      },
      "directives": []
    }],
    "directives": [],
    "selectionSet": {
      "kind": "SelectionSet",
      "selections": [{
        "kind": "Field",
        "name": {
          "kind": "Name",
          "value": "signIn"
        },
        "arguments": [{
          "kind": "Argument",
          "name": {
            "kind": "Name",
            "value": "input"
          },
          "value": {
            "kind": "Variable",
            "name": {
              "kind": "Name",
              "value": "input"
            }
          }
        }],
        "directives": [],
        "selectionSet": {
          "kind": "SelectionSet",
          "selections": [{
            "kind": "Field",
            "name": {
              "kind": "Name",
              "value": "id"
            },
            "arguments": [],
            "directives": []
          }]
        }
      }]
    }
  }],
  "loc": {
    "start": 0,
    "end": 90,
    "source": {
      "body": "\n  mutation SignIn($input: SignInInput!) {\n    signIn(input: $input) {\n      id\n    }\n  }\n",
      "name": "GraphQL request",
      "locationOffset": {
        "line": 1,
        "column": 1
      }
    }
  }
};
exports.SIGN_IN = SIGN_IN;
var defaultOptions = {
  refetchQueries: ['Cart', 'Customer', 'MiniCart'],
  awaitRefetchQueries: false,
  update: function update(store, _ref) {
    var customer = _ref.data.signIn;
    var data = store.readQuery({
      query: _Customer.GET_IS_AUTHENTICATED
    });
    store.writeQuery({
      query: _Customer.GET_IS_AUTHENTICATED,
      data: Object.assign({}, data, {
        customer: customer
      })
    });
  }
};

var SignInMutation = /*#__PURE__*/function (_Mutation) {
  (0, _inheritsLoose2.default)(SignInMutation, _Mutation);

  function SignInMutation() {
    return _Mutation.apply(this, arguments) || this;
  }

  return SignInMutation;
}(_falconData.Mutation);

exports.SignInMutation = SignInMutation;
SignInMutation.defaultProps = /*#__PURE__*/Object.assign({
  mutation: SIGN_IN
}, defaultOptions);

var useSignInMutation = function useSignInMutation(options) {
  if (options === void 0) {
    options = {};
  }

  return (0, _client.useMutation)(SIGN_IN, Object.assign({}, defaultOptions, options));
};

exports.useSignInMutation = useSignInMutation;
