// @deity/falcon-front-kit/dist/SidebarContainer/SidebarContainerProvider.js
"use strict";

var _interopRequireWildcard = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireWildcard");

exports.__esModule = true;
exports.SidebarContainer = exports.useSidebarContainer = exports.SidebarContainerProvider = void 0;

var _react = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("react"));

var _SidebarContainerContext = /*#__PURE__*/require("./SidebarContainerContext");

var SidebarContainerProvider = function SidebarContainerProvider(props) {
  var children = props.children;

  var _useState = (0, _react.useState)(false),
      isOpen = _useState[0],
      setIsOpen = _useState[1];

  var _useState2 = (0, _react.useState)(undefined),
      content = _useState2[0],
      setContent = _useState2[1];

  var open = function open(type, params, side) {
    setContent({
      type: type,
      params: params,
      side: side
    });
    setIsOpen(true);
  };

  var close = function close() {
    setIsOpen(false);
  };

  return /*#__PURE__*/_react.default.createElement(_SidebarContainerContext.SidebarContainerContext.Provider, {
    value: {
      isOpen: isOpen,
      content: content,
      open: open,
      close: close
    }
  }, children);
};

exports.SidebarContainerProvider = SidebarContainerProvider;

var useSidebarContainer = function useSidebarContainer() {
  return (0, _react.useContext)(_SidebarContainerContext.SidebarContainerContext);
};

exports.useSidebarContainer = useSidebarContainer;

var SidebarContainer = function SidebarContainer(_ref) {
  var children = _ref.children;
  return children(Object.assign({}, useSidebarContainer()));
};

exports.SidebarContainer = SidebarContainer;
