// @deity/falcon-adyen-plugin/dist/AdyenCCPlugin.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.AdyenCCPlugin = void 0;

var _inheritsLoose2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/inheritsLoose"));

var _falconPaymentPlugin = /*#__PURE__*/require("@deity/falcon-payment-plugin");

// eslint-disable-next-line import/order
var adyen = /*#__PURE__*/require('adyen-cse-web');

var AdyenCCPlugin = /*#__PURE__*/function (_PaymentPluginModel) {
  (0, _inheritsLoose2.default)(AdyenCCPlugin, _PaymentPluginModel);

  function AdyenCCPlugin(props) {
    var _this;

    _this = _PaymentPluginModel.call(this, props) || this;
    _this.cseInstance = void 0;

    _this.setCreditCardData = function (data) {
      _this.encryptCreditCard(data);
    };

    var key = _this.props.config.key;
    _this.cseInstance = adyen.createEncryption(key, {});
    return _this;
  }

  var _proto = AdyenCCPlugin.prototype;

  _proto.encryptCreditCard = function encryptCreditCard(_ref) {
    var number = _ref.number,
        expiry = _ref.expiry,
        cvc = _ref.cvc,
        name = _ref.name;
    var expiryMonth = expiry.split('/')[0].trim();
    var expiryYear = expiry.split('/')[1].trim();

    if (expiryYear.length === 2) {
      // Adding `20xx` prefix for 2-digit year
      expiryYear = "20" + expiryYear;
    }

    var creditCard = {
      holderName: name,
      number: number.replace(/ /g, ''),
      cvc: cvc,
      expiryMonth: expiryMonth,
      expiryYear: expiryYear,
      generationtime: new Date().toISOString()
    };
    var encryptedData = this.cseInstance.encrypt(creditCard);

    if (encryptedData) {
      this.props.onPaymentDetailsReady({
        cc_type: '',
        encrypted_data: encryptedData,
        store_cc: false
      });
    }
  };

  _proto.render = function render() {
    return this.props.children({
      setCreditCardData: this.setCreditCardData
    });
  };

  return AdyenCCPlugin;
}(_falconPaymentPlugin.PaymentPluginModel);

exports.AdyenCCPlugin = AdyenCCPlugin;
AdyenCCPlugin.icon = 'https://raw.githubusercontent.com/Adyen/adyen-magento2/develop/view/base/web/images/logos/creditcard.png';
