// @deity/falcon-adyen-plugin/dist/falcon-adyen-plugin.cjs.js
'use strict';

var falconPaymentPlugin = require('@deity/falcon-payment-plugin');

// eslint-disable-next-line import/order

const adyen = require('adyen-cse-web');

class AdyenCCPlugin extends falconPaymentPlugin.PaymentPluginModel {
  constructor(props) {
    super(props);
    this.cseInstance = void 0;

    this.setCreditCardData = data => {
      this.encryptCreditCard(data);
    };

    const {
      key
    } = this.props.config;
    this.cseInstance = adyen.createEncryption(key, {});
  }

  encryptCreditCard({
    number,
    expiry,
    cvc,
    name
  }) {
    const expiryMonth = expiry.split('/')[0].trim();
    let expiryYear = expiry.split('/')[1].trim();

    if (expiryYear.length === 2) {
      // Adding `20xx` prefix for 2-digit year
      expiryYear = `20${expiryYear}`;
    }

    const creditCard = {
      holderName: name,
      number: number.replace(/ /g, ''),
      cvc,
      expiryMonth,
      expiryYear,
      generationtime: new Date().toISOString()
    };
    const encryptedData = this.cseInstance.encrypt(creditCard);

    if (encryptedData) {
      this.props.onPaymentDetailsReady({
        cc_type: '',
        encrypted_data: encryptedData,
        store_cc: false
      });
    }
  }

  render() {
    return this.props.children({
      setCreditCardData: this.setCreditCardData
    });
  }

}
AdyenCCPlugin.icon = 'https://raw.githubusercontent.com/Adyen/adyen-magento2/develop/view/base/web/images/logos/creditcard.png';

module.exports = AdyenCCPlugin;
