// @deity/falcon-shop-data/dist/Product/ProductDetailsFromOptionsQuery.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.useProductDetailsFromOptionsLazyQuery = exports.useProductDetailsFromOptionsQuery = exports.ProductDetailsFromOptionsQuery = exports.GET_PRODUCT_DETAILS_FROM_OPTIONS = void 0;

var _inheritsLoose2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/inheritsLoose"));

var _client = /*#__PURE__*/require("@apollo/client");

var _falconData = /*#__PURE__*/require("@deity/falcon-data");

var GET_PRODUCT_DETAILS_FROM_OPTIONS = {
  "kind": "Document",
  "definitions": [{
    "kind": "OperationDefinition",
    "operation": "query",
    "name": {
      "kind": "Name",
      "value": "ProductDetailsFromOptions"
    },
    "variableDefinitions": [{
      "kind": "VariableDefinition",
      "variable": {
        "kind": "Variable",
        "name": {
          "kind": "Name",
          "value": "input"
        }
      },
      "type": {
        "kind": "NonNullType",
        "type": {
          "kind": "NamedType",
          "name": {
            "kind": "Name",
            "value": "ProductDetailsInput"
          }
        }
      },
      "directives": []
    }],
    "directives": [],
    "selectionSet": {
      "kind": "SelectionSet",
      "selections": [{
        "kind": "Field",
        "name": {
          "kind": "Name",
          "value": "productDetailsFromOptions"
        },
        "arguments": [{
          "kind": "Argument",
          "name": {
            "kind": "Name",
            "value": "input"
          },
          "value": {
            "kind": "Variable",
            "name": {
              "kind": "Name",
              "value": "input"
            }
          }
        }],
        "directives": [],
        "selectionSet": {
          "kind": "SelectionSet",
          "selections": [{
            "kind": "Field",
            "name": {
              "kind": "Name",
              "value": "id"
            },
            "arguments": [],
            "directives": []
          }, {
            "kind": "Field",
            "name": {
              "kind": "Name",
              "value": "sku"
            },
            "arguments": [],
            "directives": []
          }, {
            "kind": "Field",
            "name": {
              "kind": "Name",
              "value": "gallery"
            },
            "arguments": [],
            "directives": [],
            "selectionSet": {
              "kind": "SelectionSet",
              "selections": [{
                "kind": "Field",
                "name": {
                  "kind": "Name",
                  "value": "full"
                },
                "arguments": [],
                "directives": []
              }, {
                "kind": "Field",
                "name": {
                  "kind": "Name",
                  "value": "thumbnail"
                },
                "arguments": [],
                "directives": []
              }, {
                "kind": "Field",
                "name": {
                  "kind": "Name",
                  "value": "isDefault"
                },
                "arguments": [],
                "directives": []
              }]
            }
          }, {
            "kind": "Field",
            "name": {
              "kind": "Name",
              "value": "price"
            },
            "arguments": [],
            "directives": [],
            "selectionSet": {
              "kind": "SelectionSet",
              "selections": [{
                "kind": "Field",
                "name": {
                  "kind": "Name",
                  "value": "regular"
                },
                "arguments": [],
                "directives": []
              }, {
                "kind": "Field",
                "name": {
                  "kind": "Name",
                  "value": "special"
                },
                "arguments": [],
                "directives": []
              }, {
                "kind": "Field",
                "name": {
                  "kind": "Name",
                  "value": "min"
                },
                "arguments": [],
                "directives": []
              }]
            }
          }]
        }
      }]
    }
  }],
  "loc": {
    "start": 0,
    "end": 282,
    "source": {
      "body": "\n  query ProductDetailsFromOptions($input: ProductDetailsInput!) {\n    productDetailsFromOptions(input: $input) {\n      id\n      sku\n      gallery {\n        full\n        thumbnail\n        isDefault\n      }\n      price {\n        regular\n        special\n        min\n      }\n    }\n  }\n",
      "name": "GraphQL request",
      "locationOffset": {
        "line": 1,
        "column": 1
      }
    }
  }
};
exports.GET_PRODUCT_DETAILS_FROM_OPTIONS = GET_PRODUCT_DETAILS_FROM_OPTIONS;

var ProductDetailsFromOptionsQuery = /*#__PURE__*/function (_Query) {
  (0, _inheritsLoose2.default)(ProductDetailsFromOptionsQuery, _Query);

  function ProductDetailsFromOptionsQuery() {
    return _Query.apply(this, arguments) || this;
  }

  return ProductDetailsFromOptionsQuery;
}(_falconData.Query);

exports.ProductDetailsFromOptionsQuery = ProductDetailsFromOptionsQuery;
ProductDetailsFromOptionsQuery.defaultProps = {
  query: GET_PRODUCT_DETAILS_FROM_OPTIONS,
  fetchPolicy: 'cache-and-network'
};

var useProductDetailsFromOptionsQuery = function useProductDetailsFromOptionsQuery(options) {
  if (options === void 0) {
    options = {};
  }

  return (0, _client.useQuery)(GET_PRODUCT_DETAILS_FROM_OPTIONS, Object.assign({
    fetchPolicy: 'cache-and-network'
  }, options));
};

exports.useProductDetailsFromOptionsQuery = useProductDetailsFromOptionsQuery;

var useProductDetailsFromOptionsLazyQuery = function useProductDetailsFromOptionsLazyQuery(options) {
  if (options === void 0) {
    options = {};
  }

  return (0, _client.useLazyQuery)(GET_PRODUCT_DETAILS_FROM_OPTIONS, Object.assign({
    fetchPolicy: 'cache-and-network'
  }, options));
};

exports.useProductDetailsFromOptionsLazyQuery = useProductDetailsFromOptionsLazyQuery;
