// @deity/falcon-data/dist/Query/Query.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.Query = void 0;

var _objectWithoutPropertiesLoose2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _inheritsLoose2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/inheritsLoose"));

var _react = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("react"));

var _propTypes = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("prop-types"));

var _client = /*#__PURE__*/require("@apollo/client");

var _components = /*#__PURE__*/require("@apollo/client/react/components");

var _Loader = /*#__PURE__*/require("./Loader");

var _OperationError = /*#__PURE__*/require("./OperationError");

var Query = /*#__PURE__*/function (_React$Component) {
  (0, _inheritsLoose2.default)(Query, _React$Component);

  function Query() {
    return _React$Component.apply(this, arguments) || this;
  }

  var _proto = Query.prototype;

  _proto.render = function render() {
    var _this$props = this.props,
        children = _this$props.children,
        fetchMore = _this$props.fetchMore,
        passLoading = _this$props.passLoading,
        passError = _this$props.passError,
        restProps = (0, _objectWithoutPropertiesLoose2.default)(_this$props, ["children", "fetchMore", "passLoading", "passError"]);
    return /*#__PURE__*/_react.default.createElement(_components.Query, restProps, function (result) {
      var networkStatus = result.networkStatus,
          error = result.error,
          data = result.data,
          apolloFetchMore = result.fetchMore;

      if (!passError && error) {
        return /*#__PURE__*/_react.default.createElement(_OperationError.OperationError, error);
      }

      var loading = networkStatus === _client.NetworkStatus.loading || networkStatus !== _client.NetworkStatus.fetchMore && result.loading;

      if (!passLoading && loading) {
        if (data) {
          return children(Object.assign({}, result, {
            data: data,
            fetchMore: fetchMore ? function () {
              return fetchMore(data, apolloFetchMore);
            } : apolloFetchMore
          }));
        }

        return /*#__PURE__*/_react.default.createElement(_Loader.Loader, null);
      }

      return children(Object.assign({}, result, {
        data: data || {},
        fetchMore: fetchMore ? function () {
          return fetchMore(data, apolloFetchMore);
        } : apolloFetchMore
      }));
    });
  };

  return Query;
}(_react.default.Component);

exports.Query = Query;
Query.propTypes = /*#__PURE__*/Object.assign({}, _components.Query.propTypes, {
  passLoading: _propTypes.default.bool,
  passError: _propTypes.default.bool
});
