// @deity/falcon-ui-kit/dist/MiniCart/MiniCartProduct.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.MiniCartProduct = exports.MiniCartProductArea = void 0;

var _react = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("react"));

var _falconUi = /*#__PURE__*/require("@deity/falcon-ui");

var _falconI18n = /*#__PURE__*/require("@deity/falcon-i18n");

var _falconShopData = /*#__PURE__*/require("@deity/falcon-shop-data");

var _helpers = /*#__PURE__*/require("../helpers");

var _Price = /*#__PURE__*/require("../Price");

var MiniCartProductArea = {
  empty: '.',
  thumb: 'thumb',
  price: 'price',
  productName: 'productName',
  modify: 'modify',
  remove: 'remove'
};
exports.MiniCartProductArea = MiniCartProductArea;
var miniCartProductTheme = {
  miniCartProduct: {
    display: 'grid',
    gridGap: 'xs',
    // prettier-ignore
    gridTemplate: /*#__PURE__*/(0, _helpers.toGridTemplate)([['30px', '1fr', '2fr'], [MiniCartProductArea.empty, MiniCartProductArea.thumb, MiniCartProductArea.productName], [MiniCartProductArea.remove, MiniCartProductArea.thumb, MiniCartProductArea.price, '1fr'], [MiniCartProductArea.empty, MiniCartProductArea.thumb, MiniCartProductArea.modify]])
  }
};

var MiniCartProduct = function MiniCartProduct(_ref) {
  var product = _ref.product;
  return /*#__PURE__*/_react.default.createElement(_falconUi.Box, {
    defaultTheme: miniCartProductTheme
  }, /*#__PURE__*/_react.default.createElement(_falconUi.Image, {
    gridArea: MiniCartProductArea.thumb,
    src: product.thumbnailUrl
  }), /*#__PURE__*/_react.default.createElement(_falconUi.H4, {
    gridArea: MiniCartProductArea.productName
  }, product.name), /*#__PURE__*/_react.default.createElement(_Price.Price, {
    gridArea: MiniCartProductArea.price,
    value: product.price,
    fontSize: "md",
    fontWeight: "bold"
  }), /*#__PURE__*/_react.default.createElement(_falconShopData.RemoveCartItemMutation, null, function (removeCartItem) {
    return /*#__PURE__*/_react.default.createElement(_falconUi.Link, {
      gridArea: MiniCartProductArea.remove,
      display: "flex",
      alignItems: "center",
      onClick: function onClick() {
        return removeCartItem({
          variables: {
            input: {
              id: product.id
            }
          },
          optimisticResponse: {
            removeCartItem: {
              __typename: 'RemoveCartItemPayload',
              id: product.id
            }
          }
        });
      }
    }, /*#__PURE__*/_react.default.createElement(_falconUi.Icon, {
      size: "lg",
      stroke: "secondaryDark",
      src: "remove",
      mr: "xs",
      css: function css(_ref2) {
        var theme = _ref2.theme;
        return {
          ':hover': {
            stroke: theme.colors.primary
          }
        };
      }
    }));
  }), /*#__PURE__*/_react.default.createElement(_falconShopData.UpdateCartItemMutation, null, function (updateCartItem, _ref3) {
    var loading = _ref3.loading,
        error = _ref3.error;
    return /*#__PURE__*/_react.default.createElement(_falconUi.Box, {
      gridArea: MiniCartProductArea.modify
    }, /*#__PURE__*/_react.default.createElement(_falconI18n.I18n, null, function (t) {
      return /*#__PURE__*/_react.default.createElement(_falconUi.NumberInput, {
        disabled: loading,
        name: "qty",
        value: product.qty,
        "aria-label": t('product.qtyFieldLabel'),
        onChange: function onChange(ev) {
          return updateCartItem({
            variables: {
              input: {
                id: product.id,
                sku: product.sku,
                qty: parseInt(ev.target.value, 10)
              }
            }
          });
        }
      });
    }), !!error && /*#__PURE__*/_react.default.createElement(_falconUi.Text, {
      color: "error"
    }, error.message));
  }));
};

exports.MiniCartProduct = MiniCartProduct;
