// @deity/falcon-ui-kit/dist/MiniCart/MiniCart.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.MiniCart = void 0;

var _react = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("react"));

var _falconUi = /*#__PURE__*/require("@deity/falcon-ui");

var _falconI18n = /*#__PURE__*/require("@deity/falcon-i18n");

var _helpers = /*#__PURE__*/require("../helpers");

var _MiniCartProduct = /*#__PURE__*/require("./MiniCartProduct");

var _MiniCartLayout = /*#__PURE__*/require("./MiniCartLayout");

var MiniCart = function MiniCart(_ref) {
  var onCheckout = _ref.onCheckout,
      items = _ref.items;
  return /*#__PURE__*/_react.default.createElement(_MiniCartLayout.MiniCartLayout, null, /*#__PURE__*/_react.default.createElement(_falconUi.Box, {
    gridArea: _MiniCartLayout.MiniCartLayoutArea.items,
    css: function css(props) {
      return Object.assign({}, (0, _helpers.prettyScrollbars)(props.theme));
    }
  }, /*#__PURE__*/_react.default.createElement(_falconUi.List, null, items.map(function (item, index) {
    var isLastItem = index === items.length - 1;
    return /*#__PURE__*/_react.default.createElement(_falconUi.ListItem, {
      key: item.sku,
      pb: "none"
    }, /*#__PURE__*/_react.default.createElement(_MiniCartProduct.MiniCartProduct, {
      product: item
    }), !isLastItem && /*#__PURE__*/_react.default.createElement(_falconUi.Divider, {
      my: "md"
    }));
  }))), /*#__PURE__*/_react.default.createElement(_falconUi.Box, {
    gridArea: _MiniCartLayout.MiniCartLayoutArea.cta,
    py: "sm",
    bgFullWidth: "secondaryLight"
  }, /*#__PURE__*/_react.default.createElement(_falconUi.Button, {
    onClick: function onClick() {
      return onCheckout();
    },
    css: {
      width: '100%'
    },
    height: "xl"
  }, /*#__PURE__*/_react.default.createElement(_falconUi.Icon, {
    stroke: "white",
    size: "md",
    mr: "xs",
    src: "lock"
  }), /*#__PURE__*/_react.default.createElement(_falconI18n.T, {
    id: "miniCart.checkout"
  }))));
};

exports.MiniCart = MiniCart;
