// @deity/falcon-ui-kit/dist/helpers.js
"use strict";

exports.__esModule = true;
exports.placeholder = exports.lineClamp = exports.noScrollbars = exports.prettyScrollbars = exports.toGridTemplate = void 0;

var toGridTemplate = function toGridTemplate(items) {
  var columnTemplate = items.shift();

  if (!columnTemplate) {
    return '';
  }

  var gridAreas = items.map(function (item) {
    var rowTemplate = item.length > columnTemplate.length ? item.pop() : '';
    return "\"" + item.join(' ') + "\" " + rowTemplate;
  }).join(' ');
  return gridAreas + " / " + columnTemplate.join(' ');
};

exports.toGridTemplate = toGridTemplate;

var prettyScrollbars = function prettyScrollbars(theme) {
  return {
    overflowY: ['auto', 'overlay'],
    WebkitOverflowScrolling: 'touch',
    paddingRight: 10,
    marginRight: 6,
    '::-webkit-scrollbar': {
      width: 3,
      backgroundColor: theme.colors.secondaryLight,
      borderRadius: theme.borderRadius.medium
    },
    '::-webkit-scrollbar-thumb': {
      borderRadius: theme.borderRadius.medium,
      backgroundColor: theme.colors.secondaryDark
    }
  };
};

exports.prettyScrollbars = prettyScrollbars;

var noScrollbars = function noScrollbars() {
  return {
    overflow: ['auto', 'overlay'],
    WebkitOverflowScrolling: 'touch',
    '::-webkit-scrollbar': {
      display: 'none'
    },
    msOverflowStyle: 'none',
    scrollbarWidth: 'none'
  };
};

exports.noScrollbars = noScrollbars;

var lineClamp = function lineClamp(numLines) {
  return {
    display: '-webkit-box',
    WebkitBoxOrient: 'vertical',
    overflow: 'hidden',
    WebkitLineClamp: numLines
  };
};

exports.lineClamp = lineClamp;

var placeholder = function placeholder(styles) {
  return {
    '::-webkit-input-placeholder': styles,
    '::-moz-placeholder': styles,
    ':-ms-input-placeholder`': styles,
    '::-ms-input-placeholder`': styles
  };
};

exports.placeholder = placeholder;
