// @deity/falcon-stripe-plugin/dist/Loader.js
"use strict";

var _interopRequireWildcard = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireWildcard");

exports.__esModule = true;
exports.withStripe = void 0;

var _react = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("react"));

var _reactHelmetAsync = /*#__PURE__*/require("react-helmet-async");

var _reactStripeElements = /*#__PURE__*/require("react-stripe-elements");

var withStripe = function withStripe(WrappedComponent) {
  var WrappedComponentWithStripe = (0, _reactStripeElements.injectStripe)(WrappedComponent);
  return function (props) {
    var _useState = (0, _react.useState)(null),
        stripe = _useState[0],
        setStripe = _useState[1];

    var errMessage = 'Stripe Public Key was not returned from the Server!';
    var errorComponent = props.errorComponent,
        key = props.config.key;

    if (!key) {
      return errorComponent ? errorComponent(errMessage) : /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, errMessage);
    }

    return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_reactHelmetAsync.Helmet, {
      onChangeClientState: function onChangeClientState() {
        document.querySelector('#stripe-js').addEventListener('load', function () {
          setStripe(window.Stripe(key));
        });
      }
    }, /*#__PURE__*/_react.default.createElement("script", {
      id: "stripe-js",
      src: "https://js.stripe.com/v3/",
      async: true
    })), stripe && /*#__PURE__*/_react.default.createElement(_reactStripeElements.StripeProvider, {
      stripe: stripe
    }, /*#__PURE__*/_react.default.createElement(_reactStripeElements.Elements, null, /*#__PURE__*/_react.default.createElement(WrappedComponentWithStripe, props))));
  };
};

exports.withStripe = withStripe;
