// @deity/falcon-ui/dist/components/Button.js
import { themed } from '../theme';
export var Button = /*#__PURE__*/themed({
  tag: 'button',
  defaultTheme: {
    button: {
      color: 'primaryText',
      bg: 'primary',
      borderRadius: 'medium',
      fontSize: 'sm',
      lineHeight: 'small',
      height: 'lg',
      px: 'sm',
      transitionTimingFunction: 'easeIn',
      transitionDuration: 'short',
      border: 'none',
      css: function css(_ref) {
        var theme = _ref.theme;
        return {
          // basic reset styles
          fontFamily: 'inherit',
          WebkitFontSmoothing: 'antialiased',
          display: 'inline-flex',
          justifyContent: 'center',
          alignItems: 'center',
          textAlign: 'center',
          textDecoration: 'none',
          appearance: 'none',
          cursor: 'pointer',
          // define transform that scales on active
          transitionProperty: 'transform',
          ':active': {
            transform: 'scale(0.95)'
          },
          ':hover:enabled': {
            backgroundColor: theme.colors.primaryLight
          },
          ':disabled': {
            opacity: 0.6,
            cursor: 'default'
          }
        };
      },
      variants: {
        secondary: {
          bg: 'secondary',
          color: 'secondaryText',
          css: function css(_ref2) {
            var theme = _ref2.theme;
            return {
              ':hover': {
                backgroundColor: theme.colors.secondaryLight
              }
            };
          }
        }
      }
    }
  }
});
