// @deity/falcon-ui/dist/components/Menu.js
import { themed } from '../theme';
import { List, ListItem } from './List';
export var Menu = /*#__PURE__*/themed({
  tag: List,
  defaultTheme: {
    menu: {
      p: 'none',
      m: 'none',
      bg: 'primary',
      color: 'primaryText',
      css: {
        display: 'flex',
        flexDirection: 'column',
        listStyle: 'none',
        position: 'relative'
      }
    }
  }
});
export var MenuItem = /*#__PURE__*/themed({
  tag: ListItem,
  defaultTheme: {
    menuItem: {
      fontSize: 'sm',
      fontWeight: 'bold',
      display: 'flex',
      color: 'primaryText',
      css: function css(_ref) {
        var theme = _ref.theme;
        return {
          cursor: 'pointer',
          userSelect: 'none',
          listStyle: 'none',
          ':not(:last-child)': {
            borderBottom: theme.borders.regular,
            borderColor: theme.colors.secondaryDark
          },
          ':hover': {
            background: theme.colors.primaryLight,
            borderColor: theme.colors.secondaryLight
          }
        };
      }
    }
  }
});
