// @deity/falcon-ui/dist/components/RangeInput.js
import { themed } from '../theme';

var rangeInputTrack = function rangeInputTrack(styles) {
  return {
    '::-webkit-slider-runnable-track': styles,
    '::-moz-range-track': styles,
    '::-ms-track': styles
  };
};

var rangeInputThumb = function rangeInputThumb(styles) {
  return {
    '::-webkit-slider-thumb ': styles,
    '::-moz-range-thumb': styles,
    '::-ms-thumb': styles
  };
};

export var RangeInput = /*#__PURE__*/themed({
  tag: 'input',
  defaultProps: {
    type: 'range'
  },
  defaultTheme: {
    rangeInput: {
      height: 'xs',
      css: function css(_ref) {
        var theme = _ref.theme;
        return Object.assign({
          background: 'transparent',
          width: '100%',
          WebkitAppearance: 'none',
          ':focus': {
            outline: 'none'
          }
        }, rangeInputTrack({
          width: '100%',
          cursor: 'pointer',
          display: 'flex',
          alignItems: 'center',
          border: 'none',
          borderRadius: theme.borderRadius.medium,
          height: '50%',
          background: theme.colors.secondary
        }), rangeInputThumb({
          WebkitAppearance: 'none',
          cursor: 'pointer',
          border: 'none',
          borderRadius: theme.borderRadius.round,
          background: theme.colors.primary,
          height: theme.spacing.sm,
          width: theme.spacing.sm,
          transition: 'transform',
          transitionTimingFunction: theme.easingFunctions.easeIn,
          transitionDuration: theme.transitionDurations.short
        }), {
          ':active': Object.assign({}, rangeInputThumb({
            transform: 'scale(1.1)'
          }))
        });
      }
    }
  }
});
