// @deity/falcon-ui/dist/components/Select.js
import { themed } from '../theme';
export var Select = /*#__PURE__*/themed({
  tag: 'select',
  defaultProps: {
    invalid: false
  },
  defaultTheme: {
    select: {
      py: 'xs',
      px: 'sm',
      border: 'regular',
      borderRadius: 'medium',
      css: function css(_ref) {
        var theme = _ref.theme,
            invalid = _ref.invalid;
        return {
          display: 'block',
          fontFamily: 'inherit',
          lineHeight: 'inherit',
          color: 'inherit',
          borderColor: invalid ? theme.colors.error : theme.colors.secondaryDark,
          width: '100%',
          position: 'relative',
          background: "url(data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHdpZHRoPSIyNCIgaGVpZ2h0PSIyNCIgdmlld0JveD0iMCAwIDI0IDI0IiBmaWxsPSJub25lIiBzdHJva2U9ImN1cnJlbnRDb2xvciIgc3Ryb2tlLXdpZHRoPSIyIiBzdHJva2UtbGluZWNhcD0icm91bmQiIHN0cm9rZS1saW5lam9pbj0icm91bmQiIGNsYXNzPSJmZWF0aGVyIGZlYXRoZXItY2hldnJvbi1kb3duIj48cG9seWxpbmUgcG9pbnRzPSI2IDkgMTIgMTUgMTggOSI+PC9wb2x5bGluZT48L3N2Zz4=) 99% 50% no-repeat " + theme.colors.white + ";",
          WebkitAppearance: 'none',
          MozAppearance: 'none',
          appearance: 'none',
          cursor: 'pointer',
          ':disabled': {
            opacity: 0.5,
            cursor: 'not-allowed'
          },
          '::-ms-expand': {
            display: 'none'
          }
        };
      }
    }
  }
});
export var Option = /*#__PURE__*/themed({
  tag: 'option',
  defaultTheme: {
    option: {
      fontSize: 'sm',
      css: {
        fontFamily: 'inherit'
      }
    }
  }
});
