// @deity/falcon-payment-plugin/dist/TestAdditionalPaymentStep.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.TestAdditionalPaymentStep = void 0;

var _inheritsLoose2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/inheritsLoose"));

var _react = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("react"));

var TestAdditionalPaymentStep = /*#__PURE__*/function (_React$Component) {
  (0, _inheritsLoose2.default)(TestAdditionalPaymentStep, _React$Component);

  function TestAdditionalPaymentStep(props) {
    var _this;

    _this = _React$Component.call(this, props) || this;
    _this.form = void 0;
    _this.form = /*#__PURE__*/_react.default.createRef();
    return _this;
  }

  var _proto = TestAdditionalPaymentStep.prototype;

  _proto.componentDidMount = function componentDidMount() {
    var form = this.form.current;
    var method = form.method,
        action = form.action;

    if (method.toLowerCase() === 'get') {
      // Delay the redirect a little bit, to give the ConfirmLeavePrompt time to remove the beforeunload handler
      setTimeout(function () {
        window.location.href = action;
      }, 1500);
    } else {
      form.submit();
    }
  };

  _proto.render = function render() {
    var _this$props = this.props,
        url = _this$props.url,
        _this$props$method = _this$props.method,
        method = _this$props$method === void 0 ? 'POST' : _this$props$method,
        fields = _this$props.fields;
    return /*#__PURE__*/_react.default.createElement("form", {
      ref: this.form,
      method: method,
      action: url
    }, (fields || []).map(function (field) {
      return /*#__PURE__*/_react.default.createElement("input", {
        key: field.name,
        type: "hidden",
        name: field.name,
        value: field.value
      });
    }));
  };

  return TestAdditionalPaymentStep;
}(_react.default.Component);

exports.TestAdditionalPaymentStep = TestAdditionalPaymentStep;
