// @deity/falcon-logger/dist/format/index.js
"use strict";

var _interopRequireWildcard = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.CONSTANTS = exports.falconPrettyFactory = void 0;

var _chalk = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("chalk"));

var _jmespath = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("jmespath"));

var CONSTANTS = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./constants"));

exports.CONSTANTS = CONSTANTS;

var _colors = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("./colors"));

exports.colors = _colors.default;

var _minimal = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("./minimal"));

var _utils = /*#__PURE__*/require("./utils");

exports.isObject = _utils.isObject;
exports.jsonParser = _utils.jsonParser;
exports.prettifyErrorLog = _utils.prettifyErrorLog;
exports.prettifyGraphQLErrorLog = _utils.prettifyGraphQLErrorLog;
exports.prettifyLevel = _utils.prettifyLevel;
exports.prettifyMessage = _utils.prettifyMessage;
exports.prettifyMetadata = _utils.prettifyMetadata;
exports.prettifyObject = _utils.prettifyObject;
exports.prettifyModule = _utils.prettifyModule;
exports.prettifyTime = _utils.prettifyTime;
var defaultOptions = {
  colorize: _chalk.default.supportsColor,
  crlf: false,
  errorLikeObjectKeys: CONSTANTS.ERROR_LIKE_KEYS,
  errorProps: '',
  levelFirst: false,
  messageKey: CONSTANTS.MESSAGE_KEY,
  timestampKey: CONSTANTS.TIMESTAMP_KEY,
  translateTime: 'SYS:yyyy-mm-dd HH:MM:ss.l',
  useMetadata: false,
  outputStream: process.stdout
};

var falconPrettyFactory = function falconPrettyFactory(options) {
  var opts = Object.assign({}, defaultOptions, options);
  var EOL = opts.crlf ? '\r\n' : '\n';
  var IDENT = '    ';
  var messageKey = opts.messageKey;
  var timestampKey = opts.timestampKey;
  var errorLikeObjectKeys = opts.errorLikeObjectKeys;
  var errorProps = opts.errorProps.split(',');
  var ignoreKeys = opts.ignore ? new Set(opts.ignore.split(',')) : undefined;
  var colorizer = (0, _colors.default)(opts.colorize);
  var search = opts.search;

  if (opts.minimal) {
    return (0, _minimal.default)({
      IDENT: IDENT,
      EOL: EOL
    });
  }

  var formatMessage = function formatMessage(inputData) {
    var log;

    if (!(0, _utils.isObject)(inputData)) {
      var parsed = (0, _utils.jsonParser)(inputData);
      log = parsed.value;

      if (parsed.err || !(0, _utils.isObject)(log)) {
        // pass through
        return inputData + EOL;
      }
    } else {
      log = inputData;
    } // Short-circuit for spec allowed primitive values.


    if ([null, true, false].includes(log)) {
      return log + "\n";
    }

    if (search && !_jmespath.default.search(log, search)) {
      return undefined;
    }

    if (ignoreKeys) {
      log = Object.keys(log).filter(function (key) {
        return !ignoreKeys.has(key);
      }).reduce(function (res, key) {
        res[key] = log[key];
        return res;
      }, {});
    }

    var prettifiedLevel = (0, _utils.prettifyLevel)({
      log: log,
      colorizer: colorizer
    });
    var prettifiedMessage = (0, _utils.prettifyMessage)({
      log: log,
      messageKey: messageKey,
      colorizer: colorizer
    });
    var prettifiedMetadata = (0, _utils.prettifyMetadata)({
      log: log
    });
    var prettifiedTime = (0, _utils.prettifyTime)({
      log: log,
      translateFormat: opts.translateTime,
      timestampKey: timestampKey
    });
    var prettifiedModule = (0, _utils.prettifyModule)({
      log: log,
      colorizer: colorizer
    });
    var line = '';

    if (opts.levelFirst && prettifiedLevel) {
      line = "" + prettifiedLevel;
    }

    if (prettifiedTime && line === '') {
      line = "" + prettifiedTime;
    } else if (prettifiedTime) {
      line = line + " " + prettifiedTime;
    }

    if (!opts.levelFirst && prettifiedLevel) {
      if (line.length > 0) {
        line = line + " " + prettifiedLevel;
      } else {
        line = prettifiedLevel;
      }
    }

    if (opts.useMetadata && prettifiedMetadata) {
      line = line + " " + prettifiedMetadata;
    }

    if (prettifiedModule) {
      line = line + " " + prettifiedModule;
    }

    if (line.endsWith(':') === false && line !== '') {
      line += ':';
    }

    if (prettifiedMessage) {
      line = line + " " + prettifiedMessage;
    }

    if (line.length > 0) {
      line += EOL;
    }

    if (log.type === 'Error' && (log.stack || log.extensions || log.body)) {
      var prettifiedErrorLog = '';

      if (log.body && log.locations) {
        prettifiedErrorLog = (0, _utils.prettifyGraphQLParsingErrorLog)({
          log: log,
          colorizer: colorizer,
          ident: IDENT,
          eol: EOL
        });
      } else if (log.stack) {
        prettifiedErrorLog = (0, _utils.prettifyErrorLog)({
          log: log,
          messageKey: messageKey,
          colorizer: colorizer,
          errorLikeKeys: errorLikeObjectKeys,
          errorProperties: errorProps,
          ident: IDENT,
          eol: EOL
        });
      } else if (log.extensions) {
        prettifiedErrorLog = (0, _utils.prettifyGraphQLErrorLog)({
          log: log,
          colorizer: colorizer,
          ident: IDENT,
          eol: EOL
        });
      }

      line += prettifiedErrorLog;
    } else {
      var skipKeys = typeof log[messageKey] === 'string' ? [messageKey] : undefined;
      var prettifiedObject = (0, _utils.prettifyObject)({
        log: log,
        skipKeys: skipKeys,
        errorLikeKeys: errorLikeObjectKeys,
        eol: EOL,
        ident: IDENT
      });
      line += prettifiedObject;
    }

    return line;
  };

  return function (inputData) {
    try {
      return formatMessage(inputData);
    } catch (_unused) {
      return typeof inputData === 'string' ? inputData : JSON.stringify(inputData);
    }
  };
};

exports.falconPrettyFactory = falconPrettyFactory;
