// @deity/falcon-shop-data/dist/Order/OrderListQuery.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.OrderListQuery = exports.GET_ORDER_LIST = void 0;

var _inheritsLoose2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/inheritsLoose"));

var _falconData = /*#__PURE__*/require("@deity/falcon-data");

var GET_ORDER_LIST = {
  "kind": "Document",
  "definitions": [{
    "kind": "OperationDefinition",
    "operation": "query",
    "name": {
      "kind": "Name",
      "value": "OrderList"
    },
    "variableDefinitions": [{
      "kind": "VariableDefinition",
      "variable": {
        "kind": "Variable",
        "name": {
          "kind": "Name",
          "value": "pagination"
        }
      },
      "type": {
        "kind": "NamedType",
        "name": {
          "kind": "Name",
          "value": "PaginationInput"
        }
      },
      "directives": []
    }],
    "directives": [],
    "selectionSet": {
      "kind": "SelectionSet",
      "selections": [{
        "kind": "Field",
        "name": {
          "kind": "Name",
          "value": "orderList"
        },
        "arguments": [{
          "kind": "Argument",
          "name": {
            "kind": "Name",
            "value": "pagination"
          },
          "value": {
            "kind": "Variable",
            "name": {
              "kind": "Name",
              "value": "pagination"
            }
          }
        }],
        "directives": [],
        "selectionSet": {
          "kind": "SelectionSet",
          "selections": [{
            "kind": "Field",
            "name": {
              "kind": "Name",
              "value": "items"
            },
            "arguments": [],
            "directives": [],
            "selectionSet": {
              "kind": "SelectionSet",
              "selections": [{
                "kind": "Field",
                "name": {
                  "kind": "Name",
                  "value": "id"
                },
                "arguments": [],
                "directives": []
              }, {
                "kind": "Field",
                "name": {
                  "kind": "Name",
                  "value": "referenceNo"
                },
                "arguments": [],
                "directives": []
              }, {
                "kind": "Field",
                "name": {
                  "kind": "Name",
                  "value": "createdAt"
                },
                "arguments": [],
                "directives": []
              }, {
                "kind": "Field",
                "name": {
                  "kind": "Name",
                  "value": "customerFirstname"
                },
                "arguments": [],
                "directives": []
              }, {
                "kind": "Field",
                "name": {
                  "kind": "Name",
                  "value": "customerLastname"
                },
                "arguments": [],
                "directives": []
              }, {
                "kind": "Field",
                "name": {
                  "kind": "Name",
                  "value": "status"
                },
                "arguments": [],
                "directives": []
              }, {
                "kind": "Field",
                "name": {
                  "kind": "Name",
                  "value": "currency"
                },
                "arguments": [],
                "directives": []
              }, {
                "kind": "Field",
                "name": {
                  "kind": "Name",
                  "value": "totals"
                },
                "arguments": [],
                "directives": [],
                "selectionSet": {
                  "kind": "SelectionSet",
                  "selections": [{
                    "kind": "Field",
                    "name": {
                      "kind": "Name",
                      "value": "code"
                    },
                    "arguments": [],
                    "directives": []
                  }, {
                    "kind": "Field",
                    "name": {
                      "kind": "Name",
                      "value": "title"
                    },
                    "arguments": [],
                    "directives": []
                  }, {
                    "kind": "Field",
                    "name": {
                      "kind": "Name",
                      "value": "value"
                    },
                    "arguments": [],
                    "directives": []
                  }]
                }
              }]
            }
          }, {
            "kind": "Field",
            "name": {
              "kind": "Name",
              "value": "pagination"
            },
            "arguments": [],
            "directives": [],
            "selectionSet": {
              "kind": "SelectionSet",
              "selections": [{
                "kind": "Field",
                "name": {
                  "kind": "Name",
                  "value": "currentPage"
                },
                "arguments": [],
                "directives": []
              }, {
                "kind": "Field",
                "name": {
                  "kind": "Name",
                  "value": "totalItems"
                },
                "arguments": [],
                "directives": []
              }, {
                "kind": "Field",
                "name": {
                  "kind": "Name",
                  "value": "nextPage"
                },
                "arguments": [],
                "directives": []
              }]
            }
          }]
        }
      }]
    }
  }],
  "loc": {
    "start": 0,
    "end": 413,
    "source": {
      "body": "\n  query OrderList($pagination: PaginationInput) {\n    orderList(pagination: $pagination) {\n      items {\n        id\n        referenceNo\n        createdAt\n        customerFirstname\n        customerLastname\n        status\n        currency\n        totals {\n          code\n          title\n          value\n        }\n      }\n      pagination {\n        currentPage\n        totalItems\n        nextPage\n      }\n    }\n  }\n",
      "name": "GraphQL request",
      "locationOffset": {
        "line": 1,
        "column": 1
      }
    }
  }
};
exports.GET_ORDER_LIST = GET_ORDER_LIST;

var fetchMore = function fetchMore(data, apolloFetchMore) {
  return apolloFetchMore({
    variables: {
      pagination: {
        page: data.orderList.pagination.nextPage
      }
    },
    updateQuery: function updateQuery(prev, _ref) {
      var fetchMoreResult = _ref.fetchMoreResult;

      if (!fetchMoreResult) {
        return prev;
      }

      return Object.assign({}, prev, {
        orderList: Object.assign({}, prev.orderList, {
          items: [].concat(prev.orderList.items, fetchMoreResult.orderList.items),
          pagination: Object.assign({}, fetchMoreResult.orderList.pagination)
        })
      });
    }
  });
};

var OrderListQuery = /*#__PURE__*/function (_Query) {
  (0, _inheritsLoose2.default)(OrderListQuery, _Query);

  function OrderListQuery() {
    return _Query.apply(this, arguments) || this;
  }

  return OrderListQuery;
}(_falconData.Query);

exports.OrderListQuery = OrderListQuery;
OrderListQuery.defaultProps = {
  query: GET_ORDER_LIST,
  fetchMore: fetchMore,
  notifyOnNetworkStatusChange: true,
  fetchPolicy: 'cache-and-network'
};
