// @deity/falcon-ui-kit/dist/Forms/FormField.js
"use strict";

var _interopRequireWildcard = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.FormField = void 0;

var _extends2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("react"));

var _falconFrontKit = /*#__PURE__*/require("@deity/falcon-front-kit");

var _falconUi = /*#__PURE__*/require("@deity/falcon-ui");

var _FormFieldLabel = /*#__PURE__*/require("./FormFieldLabel");

var _FormFieldError = /*#__PURE__*/require("./FormFieldError");

var _FormFieldLayout = /*#__PURE__*/require("./FormFieldLayout");

var FormField = /*#__PURE__*/_react.default.memo(function (props) {
  var name = props.name,
      validate = props.validate,
      required = props.required,
      children = props.children,
      restProps = (0, _objectWithoutPropertiesLoose2.default)(props, ["name", "validate", "required", "children"]);

  var _extractThemableProps = (0, _falconUi.extractThemableProps)(restProps),
      themableProps = _extractThemableProps.themableProps,
      rest = _extractThemableProps.rest;

  var validators = (0, _react.useCallback)(function () {
    return (0, _falconFrontKit.getDefaultInputValidators)(props);
  }, [props]);
  return /*#__PURE__*/_react.default.createElement(_falconFrontKit.Field, (0, _extends2.default)({
    name: name,
    validate: validators()
  }, rest), function (_ref) {
    var form = _ref.form,
        field = _ref.field,
        label = _ref.label,
        error = _ref.error;
    return /*#__PURE__*/_react.default.createElement(_FormFieldLayout.FormFieldLayout, themableProps, label && /*#__PURE__*/_react.default.createElement(_FormFieldLabel.FormFieldLabel, {
      htmlFor: field.id
    }, label), children ? children({
      form: form,
      field: Object.assign({}, field, {
        gridArea: _FormFieldLayout.FormFieldArea.input
      })
    }) : /*#__PURE__*/_react.default.createElement(_falconUi.Input, (0, _extends2.default)({}, field, {
      gridArea: _FormFieldLayout.FormFieldArea.input
    })), /*#__PURE__*/_react.default.createElement(_FormFieldError.FormFieldError, null, field.invalid ? error : null));
  });
});

exports.FormField = FormField;
