// @deity/falcon-ui-kit/dist/Forms/PasswordRevealInput.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.PasswordRevealInput = void 0;

var _extends2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/extends"));

var _inheritsLoose2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/inheritsLoose"));

var _react = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("react"));

var _falconI18n = /*#__PURE__*/require("@deity/falcon-i18n");

var _falconUi = /*#__PURE__*/require("@deity/falcon-ui");

var PasswordRevealInput = /*#__PURE__*/function (_React$Component) {
  (0, _inheritsLoose2.default)(PasswordRevealInput, _React$Component);

  function PasswordRevealInput() {
    var _this;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _React$Component.call.apply(_React$Component, [this].concat(args)) || this;
    _this.state = {
      isRevealed: false
    };

    _this.toggleInputType = function () {
      _this.setState(function (state) {
        return {
          isRevealed: !state.isRevealed
        };
      });
    };

    return _this;
  }

  var _proto = PasswordRevealInput.prototype;

  _proto.render = function render() {
    var _this2 = this;

    var isRevealed = this.state.isRevealed;
    var inputType = isRevealed ? 'text' : 'password';
    return /*#__PURE__*/_react.default.createElement(_falconI18n.I18n, null, function (t) {
      return /*#__PURE__*/_react.default.createElement(_falconUi.Box, {
        position: "relative"
      }, /*#__PURE__*/_react.default.createElement(_falconUi.Input, (0, _extends2.default)({}, _this2.props, {
        type: inputType
      })), /*#__PURE__*/_react.default.createElement(_falconUi.Box, {
        position: "absolute",
        top: 0,
        bottom: 0,
        right: 0,
        title: t('form.passwordRevealTitle', {
          context: isRevealed ? 'hide' : 'show'
        })
      }, /*#__PURE__*/_react.default.createElement(_falconUi.Icon, {
        onClick: _this2.toggleInputType,
        stroke: isRevealed ? 'primary' : 'black',
        cursor: "pointer",
        size: "md",
        src: isRevealed ? 'eye' : 'eyeOff',
        bg: "white",
        m: "xs"
      })));
    });
  };

  return PasswordRevealInput;
}(_react.default.Component);

exports.PasswordRevealInput = PasswordRevealInput;
