// @deity/falcon-ui-kit/dist/Forms/RadioInput.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.RadioInput = void 0;

var _objectWithoutPropertiesLoose2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("react"));

var _falconUi = /*#__PURE__*/require("@deity/falcon-ui");

var RadioInput = /*#__PURE__*/_react.default.memo(function (_ref) {
  var children = _ref.children,
      label = _ref.label,
      rest = (0, _objectWithoutPropertiesLoose2.default)(_ref, ["children", "label"]);
  return /*#__PURE__*/_react.default.createElement(_falconUi.Label, {
    htmlFor: rest.id,
    css: {
      display: 'inline-flex',
      alignItems: 'center'
    }
  }, /*#__PURE__*/_react.default.createElement(_falconUi.Radio, rest), children || /*#__PURE__*/_react.default.createElement(_falconUi.Text, {
    as: "span",
    p: "xs"
  }, label));
});

exports.RadioInput = RadioInput;
