// @deity/falcon-ui-kit/dist/Forms/RadioFormField.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.RadioFormField = exports.RadioFormFieldLayout = void 0;

var _extends2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("react"));

var _falconFrontKit = /*#__PURE__*/require("@deity/falcon-front-kit");

var _falconUi = /*#__PURE__*/require("@deity/falcon-ui");

var _helpers = /*#__PURE__*/require("../helpers");

var _FormFieldLayout = /*#__PURE__*/require("./FormFieldLayout");

var RadioFormFieldLayout = /*#__PURE__*/(0, _falconUi.themed)({
  tag: _falconUi.Box,
  defaultTheme: {
    radioFormFieldLayout: {
      display: 'grid',
      gridGap: 'xs',
      // prettier-ignore
      gridTemplate: /*#__PURE__*/(0, _helpers.toGridTemplate)([['auto', '1fr'], [_FormFieldLayout.FormFieldArea.input, _FormFieldLayout.FormFieldArea.label], [_FormFieldLayout.FormFieldArea.error, _FormFieldLayout.FormFieldArea.error, '0px']])
    }
  }
});
exports.RadioFormFieldLayout = RadioFormFieldLayout;

var RadioFormField = function RadioFormField(props) {
  var name = props.name,
      value = props.value,
      validate = props.validate,
      required = props.required,
      children = props.children,
      restProps = (0, _objectWithoutPropertiesLoose2.default)(props, ["name", "value", "validate", "required", "children"]);

  var _extractThemableProps = (0, _falconUi.extractThemableProps)(restProps),
      themableProps = _extractThemableProps.themableProps,
      rest = _extractThemableProps.rest;

  return /*#__PURE__*/_react.default.createElement(_falconFrontKit.Field, (0, _extends2.default)({
    name: name,
    validate: (0, _falconFrontKit.getDefaultInputValidators)(props)
  }, rest), function (_ref) {
    var field = _ref.field,
        label = _ref.label;
    return /*#__PURE__*/_react.default.createElement(RadioFormFieldLayout, themableProps, /*#__PURE__*/_react.default.createElement(_falconUi.Radio, (0, _extends2.default)({}, field, {
      value: value,
      checked: value === field.value,
      gridArea: _FormFieldLayout.FormFieldArea.input
    })), /*#__PURE__*/_react.default.createElement(_falconUi.FlexLayout, {
      alignItems: "center",
      gridArea: _FormFieldLayout.FormFieldArea.label
    }, /*#__PURE__*/_react.default.createElement(_falconUi.Label, {
      htmlFor: field.id
    }, children || label)));
  });
};

exports.RadioFormField = RadioFormField;
