// @deity/falcon-ui-kit/dist/Price/Price.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.Price = void 0;

var _extends2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("react"));

var _propTypes = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("prop-types"));

var _falconUi = /*#__PURE__*/require("@deity/falcon-ui");

var _falconFrontKit = /*#__PURE__*/require("@deity/falcon-front-kit");

var PriceInnerDom = function PriceInnerDom(_ref) {
  var value = _ref.value,
      formatOptions = _ref.formatOptions,
      rest = (0, _objectWithoutPropertiesLoose2.default)(_ref, ["value", "formatOptions"]);

  var _useCurrency = (0, _falconFrontKit.useCurrency)(),
      priceFormat = _useCurrency.priceFormat;

  return /*#__PURE__*/_react.default.createElement(_falconUi.Text, (0, _extends2.default)({
    m: "lg"
  }, rest), priceFormat(value, formatOptions));
};

PriceInnerDom.propTypes = {
  value: _propTypes.default.number.isRequired
};
var Price = /*#__PURE__*/(0, _falconUi.themed)({
  tag: PriceInnerDom,
  defaultProps: {
    value: 0.0,
    formatOptions: undefined,
    ellipsis: false
  },
  defaultTheme: {
    price: {
      display: 'block',
      m: 'none',
      css: {
        whiteSpace: 'nowrap',
        overflow: 'hidden'
      },
      variants: {
        old: {
          css: {
            textDecorationStyle: 'solid',
            textDecorationLine: 'line-through'
          }
        },
        special: {
          fontWeight: 'bold',
          css: {
            color: 'red'
          }
        }
      }
    }
  }
});
exports.Price = Price;
