// @deity/falcon-ui-kit/dist/Product/ProductLayout.js
"use strict";

exports.__esModule = true;
exports.ProductLayout = exports.ProductLayoutArea = void 0;

var _falconUi = /*#__PURE__*/require("@deity/falcon-ui");

var _helpers = /*#__PURE__*/require("../helpers");

var ProductLayoutArea = {
  back: 'back',
  gallery: 'gallery',
  sku: 'sku',
  title: 'title',
  description: 'description',
  cta: 'cta',
  price: 'price',
  meta: 'meta',
  empty: 'empty',
  options: 'options',
  error: 'error'
};
exports.ProductLayoutArea = ProductLayoutArea;
var ProductLayout = /*#__PURE__*/(0, _falconUi.themed)({
  tag: 'article',
  defaultTheme: {
    productLayout: {
      display: 'grid',
      gridGap: 'sm',
      gridTemplate: {
        // prettier-ignore
        xs: /*#__PURE__*/(0, _helpers.toGridTemplate)([['1fr'], [ProductLayoutArea.title], [ProductLayoutArea.sku], [ProductLayoutArea.gallery], [ProductLayoutArea.price], [ProductLayoutArea.error], [ProductLayoutArea.options], [ProductLayoutArea.cta], [ProductLayoutArea.description], [ProductLayoutArea.meta]]),
        // prettier-ignore
        md: /*#__PURE__*/(0, _helpers.toGridTemplate)([['1.5fr', '1fr'], [ProductLayoutArea.gallery, ProductLayoutArea.sku], [ProductLayoutArea.gallery, ProductLayoutArea.title], [ProductLayoutArea.gallery, ProductLayoutArea.price], [ProductLayoutArea.gallery, ProductLayoutArea.options], [ProductLayoutArea.gallery, ProductLayoutArea.cta], [ProductLayoutArea.gallery, ProductLayoutArea.error], [ProductLayoutArea.gallery, ProductLayoutArea.description, '1fr'], [ProductLayoutArea.gallery, ProductLayoutArea.meta]])
      }
    }
  }
});
exports.ProductLayout = ProductLayout;
