// @deity/falcon-ui-kit/dist/Product/ProductOptionRadioGroupField.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.ProductOptionRadioGroupField = void 0;

var _extends2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/extends"));

var _react = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("react"));

var _falconUi = /*#__PURE__*/require("@deity/falcon-ui");

var _falconFrontKit = /*#__PURE__*/require("@deity/falcon-front-kit");

var _Forms = /*#__PURE__*/require("../Forms");

var _ProductOptionRadio = /*#__PURE__*/require("./ProductOptionRadio");

var ProductOptionRadioGroupField = function ProductOptionRadioGroupField(_ref) {
  var label = _ref.label,
      name = _ref.name,
      values = _ref.values,
      disabled = _ref.disabled;
  return /*#__PURE__*/_react.default.createElement(_falconFrontKit.Field, {
    name: name,
    label: label,
    validate: [_falconFrontKit.requiredValidator]
  }, function (_ref2) {
    var field = _ref2.field,
        error = _ref2.error;
    return /*#__PURE__*/_react.default.createElement(_falconUi.Box, {
      as: "fieldset",
      p: "none",
      border: "none"
    }, label && /*#__PURE__*/_react.default.createElement(_falconUi.Box, {
      as: "legend",
      mb: "xs"
    }, label), values.map(function (value) {
      return /*#__PURE__*/_react.default.createElement(_falconUi.Label, {
        htmlFor: field.id + "-" + value.id,
        key: value.id,
        mr: "xs",
        style: {
          display: 'inline-flex'
        }
      }, /*#__PURE__*/_react.default.createElement(_ProductOptionRadio.ProductOptionRadio, (0, _extends2.default)({}, field, {
        id: field.id + "-" + value.id,
        value: value.id,
        checked: value.id === field.value,
        disabled: disabled,
        icon: /*#__PURE__*/_react.default.createElement("div", null, value.label)
      })));
    }), /*#__PURE__*/_react.default.createElement(_Forms.FormFieldError, null, field.invalid ? error : null));
  });
};

exports.ProductOptionRadioGroupField = ProductOptionRadioGroupField;
