// @deity/falcon-ui-kit/dist/Product/ProductPrice.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.ProductPrice = void 0;

var _extends2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("react"));

var _propTypes = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("prop-types"));

var _falconUi = /*#__PURE__*/require("@deity/falcon-ui");

var _falconI18n = /*#__PURE__*/require("@deity/falcon-i18n");

var _Price = /*#__PURE__*/require("../Price");

var ProductPrice = function ProductPrice(_ref) {
  var regular = _ref.regular,
      special = _ref.special,
      min = _ref.min,
      rest = (0, _objectWithoutPropertiesLoose2.default)(_ref, ["regular", "special", "min"]);

  var _useI18n = (0, _falconI18n.useI18n)(),
      t = _useI18n.t;

  if (min) {
    return /*#__PURE__*/_react.default.createElement(_falconUi.Box, {
      display: "flex",
      alignItems: "center"
    }, t('price.from'), " ", /*#__PURE__*/_react.default.createElement(_Price.Price, (0, _extends2.default)({
      ml: "xs",
      value: min
    }, rest)));
  }

  return special ? /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_Price.Price, (0, _extends2.default)({
    value: regular,
    fontSize: "md"
  }, rest, {
    variant: "old",
    mr: "xs"
  })), /*#__PURE__*/_react.default.createElement(_Price.Price, (0, _extends2.default)({
    value: special,
    fontSize: "md"
  }, rest, {
    variant: "special"
  }))) : /*#__PURE__*/_react.default.createElement(_Price.Price, (0, _extends2.default)({
    value: regular,
    fontSize: "md"
  }, rest));
};

exports.ProductPrice = ProductPrice;
ProductPrice.propTypes = {
  regular: _propTypes.default.number.isRequired,
  special: _propTypes.default.number
};
