// @deity/falcon-ui-kit/dist/Account/NewAccount.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.NewAccount = void 0;

var _react = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("react"));

var _falconUi = /*#__PURE__*/require("@deity/falcon-ui");

var _falconI18n = /*#__PURE__*/require("@deity/falcon-i18n");

var NewAccount = function NewAccount(_ref) {
  var onCreateNewAccount = _ref.onCreateNewAccount;
  return /*#__PURE__*/_react.default.createElement(_falconUi.GridLayout, null, /*#__PURE__*/_react.default.createElement(_falconUi.H4, {
    mb: "xs"
  }, /*#__PURE__*/_react.default.createElement(_falconI18n.T, {
    id: "newAccount.title"
  })), /*#__PURE__*/_react.default.createElement(_falconUi.Text, null, /*#__PURE__*/_react.default.createElement(_falconI18n.T, {
    id: "newAccount.benefitsHeader"
  })), /*#__PURE__*/_react.default.createElement(_falconUi.List, null, /*#__PURE__*/_react.default.createElement(_falconI18n.I18n, null, function (t) {
    return [].concat(t('newAccount.benefitList', {
      returnObjects: true
    })).map(function (x) {
      return /*#__PURE__*/_react.default.createElement(_falconUi.ListItem, {
        key: x,
        display: "flex",
        mb: "sm"
      }, /*#__PURE__*/_react.default.createElement(_falconUi.Icon, {
        src: "check",
        size: "md",
        mr: "xs",
        stroke: "primaryLight"
      }), x);
    });
  })), /*#__PURE__*/_react.default.createElement(_falconUi.Button, {
    justifySelf: "end",
    variant: "secondary",
    onClick: function onClick() {
      return onCreateNewAccount();
    }
  }, /*#__PURE__*/_react.default.createElement(_falconI18n.T, {
    id: "newAccount.createAnAccount"
  })));
};

exports.NewAccount = NewAccount;
