// @deity/falcon-ui-kit/dist/Account/SignUpForm.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.SignUpForm = void 0;

var _extends2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("react"));

var _falconFrontKit = /*#__PURE__*/require("@deity/falcon-front-kit");

var _Error = /*#__PURE__*/require("../Error");

var _Forms = /*#__PURE__*/require("../Forms");

var SignUpForm = function SignUpForm(_ref) {
  var onSuccess = _ref.onSuccess,
      formProps = (0, _objectWithoutPropertiesLoose2.default)(_ref, ["onSuccess"]);
  return /*#__PURE__*/_react.default.createElement(_falconFrontKit.SignUpFormProvider, {
    onSuccess: onSuccess
  }, function (_ref2) {
    var status = _ref2.status,
        values = _ref2.values;
    return /*#__PURE__*/_react.default.createElement(_Forms.Form, (0, _extends2.default)({
      id: "sign-up",
      i18nId: "signUp"
    }, formProps), /*#__PURE__*/_react.default.createElement(_Forms.FormField, {
      name: "firstname",
      type: "text",
      required: true,
      autoComplete: "given-name"
    }), /*#__PURE__*/_react.default.createElement(_Forms.FormField, {
      name: "lastname",
      type: "text",
      required: true,
      autoComplete: "family-name"
    }), /*#__PURE__*/_react.default.createElement(_Forms.FormField, {
      name: "email",
      type: "email",
      required: true,
      autoComplete: "email"
    }), /*#__PURE__*/_react.default.createElement(_Forms.FormField, {
      name: "password",
      required: true,
      type: "password"
    }, function (_ref3) {
      var field = _ref3.field;
      return /*#__PURE__*/_react.default.createElement(_Forms.PasswordRevealInput, field);
    }), /*#__PURE__*/_react.default.createElement(_Forms.FormField, {
      name: "confirmPassword",
      type: "password",
      validate: [_falconFrontKit.requiredValidator, (0, _falconFrontKit.confirmPasswordValidator)(values.password)]
    }, function (_ref4) {
      var field = _ref4.field;
      return /*#__PURE__*/_react.default.createElement(_Forms.PasswordRevealInput, field);
    }), /*#__PURE__*/_react.default.createElement(_Forms.FormSubmit, null), /*#__PURE__*/_react.default.createElement(_Error.ErrorSummary, {
      errors: status.error
    }));
  });
};

exports.SignUpForm = SignUpForm;
