// @deity/falcon-front-kit/dist/Product/AddToCartFormProvider.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.AddToCartFormProvider = void 0;

var _extends2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("react"));

var _formik = /*#__PURE__*/require("formik");

var _falconData = /*#__PURE__*/require("@deity/falcon-data");

var _falconShopData = /*#__PURE__*/require("@deity/falcon-shop-data");

var _React = /*#__PURE__*/require("../React");

var _productOptionMappers = /*#__PURE__*/require("./productOptionMappers");

var AddToCartFormProvider = function AddToCartFormProvider(props) {
  var onSuccess = props.onSuccess,
      initialValues = props.initialValues,
      quantity = props.quantity,
      product = props.product,
      mutationOptions = props.mutationOptions,
      formikProps = (0, _objectWithoutPropertiesLoose2.default)(props, ["onSuccess", "initialValues", "quantity", "product", "mutationOptions"]);
  var defaultInitialValues = {
    qty: quantity,
    options: (0, _productOptionMappers.productOptionsToForm)(product.options || []),
    bundleOptions: product.bundleOptions || []
  };

  var _useAddToCartMutation = (0, _falconShopData.useAddToCartMutation)(),
      addToCart = _useAddToCartMutation[0];

  var getUserError = (0, _falconData.useGetUserError)();
  var isMounted = (0, _React.useMounted)();
  return /*#__PURE__*/_react.default.createElement(_formik.Formik, (0, _extends2.default)({
    initialStatus: {},
    initialValues: initialValues || defaultInitialValues,
    onSubmit: function onSubmit(values, _ref) {
      var setSubmitting = _ref.setSubmitting,
          setStatus = _ref.setStatus;
      return addToCart(Object.assign({
        variables: {
          input: {
            sku: product.sku,
            qty: parseInt(values.qty.toString(), 10),
            options: (0, _productOptionMappers.formProductOptionsToInput)(values.options),
            bundleOptions: undefined // values.bundleOptions as any - TODO: add appropriate mapper

          }
        }
      }, mutationOptions || {})).then(function (_ref2) {
        var data = _ref2.data;

        if (isMounted.current) {
          setSubmitting(false);
          setStatus({
            data: data
          });
          return onSuccess && onSuccess(data);
        }
      }).catch(function (e) {
        var error = getUserError(e);

        if (error.length && isMounted.current) {
          setStatus({
            error: error
          });
          setSubmitting(false);
        }
      });
    }
  }, formikProps));
};

exports.AddToCartFormProvider = AddToCartFormProvider;
