// @deity/falcon-shop-data/dist/Checkout/SetShippingAddressMutation.js
"use strict";

var _interopRequireDefault = /*#__PURE__*/require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.useSetShippingAddressMutation = exports.SetShippingAddressMutation = exports.SET_SHIPPING_ADDRESS = void 0;

var _inheritsLoose2 = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("@babel/runtime/helpers/inheritsLoose"));

var _client = /*#__PURE__*/require("@apollo/client");

var _falconData = /*#__PURE__*/require("@deity/falcon-data");

var SET_SHIPPING_ADDRESS = {
  "kind": "Document",
  "definitions": [{
    "kind": "OperationDefinition",
    "operation": "mutation",
    "name": {
      "kind": "Name",
      "value": "SetShippingAddress"
    },
    "variableDefinitions": [{
      "kind": "VariableDefinition",
      "variable": {
        "kind": "Variable",
        "name": {
          "kind": "Name",
          "value": "input"
        }
      },
      "type": {
        "kind": "NonNullType",
        "type": {
          "kind": "NamedType",
          "name": {
            "kind": "Name",
            "value": "SetCheckoutAddressInput"
          }
        }
      },
      "directives": []
    }],
    "directives": [],
    "selectionSet": {
      "kind": "SelectionSet",
      "selections": [{
        "kind": "Field",
        "name": {
          "kind": "Name",
          "value": "setShippingAddress"
        },
        "arguments": [{
          "kind": "Argument",
          "name": {
            "kind": "Name",
            "value": "input"
          },
          "value": {
            "kind": "Variable",
            "name": {
              "kind": "Name",
              "value": "input"
            }
          }
        }],
        "directives": []
      }]
    }
  }],
  "loc": {
    "start": 0,
    "end": 109,
    "source": {
      "body": "\n  mutation SetShippingAddress($input: SetCheckoutAddressInput!) {\n    setShippingAddress(input: $input)\n  }\n",
      "name": "GraphQL request",
      "locationOffset": {
        "line": 1,
        "column": 1
      }
    }
  }
};
exports.SET_SHIPPING_ADDRESS = SET_SHIPPING_ADDRESS;

var SetShippingAddressMutation = /*#__PURE__*/function (_Mutation) {
  (0, _inheritsLoose2.default)(SetShippingAddressMutation, _Mutation);

  function SetShippingAddressMutation() {
    return _Mutation.apply(this, arguments) || this;
  }

  return SetShippingAddressMutation;
}(_falconData.Mutation);

exports.SetShippingAddressMutation = SetShippingAddressMutation;
SetShippingAddressMutation.defaultProps = {
  mutation: SET_SHIPPING_ADDRESS,
  refetchQueries: ['CheckoutState'],
  awaitRefetchQueries: true
};

var useSetShippingAddressMutation = function useSetShippingAddressMutation(options) {
  return (0, _client.useMutation)(SET_SHIPPING_ADDRESS, Object.assign({
    refetchQueries: ['CheckoutState'],
    awaitRefetchQueries: true
  }, options || {}));
};

exports.useSetShippingAddressMutation = useSetShippingAddressMutation;
